<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="../../../../">
<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>pygenalgo.operators.crossover.mutli_point_crossover &mdash; PyGenAlgo 01.70.50 documentation</title>
      <link rel="stylesheet" type="text/css" href="../../../../_static/pygments.css?v=92fd9be5" />
      <link rel="stylesheet" type="text/css" href="../../../../_static/css/theme.css?v=19f00094" />

  
  <!--[if lt IE 9]>
    <script src="../../../../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script src="../../../../_static/jquery.js?v=5d32c60e"></script>
        <script src="../../../../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
        <script src="../../../../_static/documentation_options.js?v=d0c4109e"></script>
        <script src="../../../../_static/doctools.js?v=9a2dae69"></script>
        <script src="../../../../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../../../../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../../../../genindex.html" />
    <link rel="search" title="Search" href="../../../../search.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../../../../index.html" class="icon icon-home">
            PyGenAlgo
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../../modules.html">pygenalgo</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../../../index.html">PyGenAlgo</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../../../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="../../../index.html">Module code</a></li>
      <li class="breadcrumb-item active">pygenalgo.operators.crossover.mutli_point_crossover</li>
      <li class="wy-breadcrumbs-aside">
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <h1>Source code for pygenalgo.operators.crossover.mutli_point_crossover</h1><div class="highlight"><pre>
<span></span><span class="kn">from</span> <span class="nn">pygenalgo.genome.chromosome</span> <span class="kn">import</span> <span class="n">Chromosome</span>
<span class="kn">from</span> <span class="nn">pygenalgo.operators.crossover.crossover_operator</span> <span class="kn">import</span> <span class="n">CrossoverOperator</span>


<div class="viewcode-block" id="MultiPointCrossover">
<a class="viewcode-back" href="../../../../pygenalgo.operators.crossover.html#pygenalgo.operators.crossover.mutli_point_crossover.MultiPointCrossover">[docs]</a>
<span class="k">class</span> <span class="nc">MultiPointCrossover</span><span class="p">(</span><span class="n">CrossoverOperator</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Description:</span>

<span class="sd">        Multipoint crossover creates two children chromosomes (offsprings),</span>
<span class="sd">        by taking two parent chromosomes and cutting them at randomly chosen,</span>
<span class="sd">        sites (loci).</span>

<span class="sd">        It produces faster mixing, compared with single-point crossover.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">crossover_probability</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">0.9</span><span class="p">,</span> <span class="n">num_loci</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">2</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Construct a &#39;MultiPointCrossover&#39; object with a given</span>
<span class="sd">        probability value.</span>

<span class="sd">        :param crossover_probability: (float).</span>

<span class="sd">        :param num_loci: (int).</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="c1"># Call the super constructor with the provided</span>
        <span class="c1"># probability value.</span>
        <span class="nb">super</span><span class="p">()</span><span class="o">.</span><span class="fm">__init__</span><span class="p">(</span><span class="n">crossover_probability</span><span class="p">)</span>

        <span class="c1"># Make sure number of loci are at least 2.</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_items</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="nb">int</span><span class="p">(</span><span class="n">num_loci</span><span class="p">),</span> <span class="mi">2</span><span class="p">)</span>
    <span class="c1"># _end_def_</span>

<div class="viewcode-block" id="MultiPointCrossover.crossover">
<a class="viewcode-back" href="../../../../pygenalgo.operators.crossover.html#pygenalgo.operators.crossover.mutli_point_crossover.MultiPointCrossover.crossover">[docs]</a>
    <span class="k">def</span> <span class="nf">crossover</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">parent1</span><span class="p">:</span> <span class="n">Chromosome</span><span class="p">,</span> <span class="n">parent2</span><span class="p">:</span> <span class="n">Chromosome</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Perform the crossover operation on the two input parent</span>
<span class="sd">        chromosomes, using multiple cutting points (num_loci).</span>

<span class="sd">        NOTE: the number of loci is held in the &#39;_items&#39; variable.</span>

<span class="sd">        :param parent1: (Chromosome).</span>

<span class="sd">        :param parent2: (Chromosome).</span>

<span class="sd">        :return: child1 and child2 (as Chromosomes).</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="c1"># Get the number of genes from the first parent (chromosome).</span>
        <span class="c1"># N.B.: It is assumed that both parents have the same size.</span>
        <span class="n">num_genes</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">parent1</span><span class="p">)</span>

        <span class="c1"># Extract the number of cut points.</span>
        <span class="n">num_points</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_items</span>

        <span class="c1"># Ensure the number of requested cutting points</span>
        <span class="c1"># do not exceed the length of the chromosomes.</span>
        <span class="k">if</span> <span class="n">num_points</span> <span class="o">&gt;=</span> <span class="n">num_genes</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="vm">__class__</span><span class="o">.</span><span class="vm">__name__</span><span class="si">}</span><span class="s2">:&quot;</span>
                             <span class="s2">&quot; Number of requested crossover points&quot;</span>
                             <span class="s2">&quot; exceeds the length of the chromosome.&quot;</span><span class="p">)</span>
        <span class="c1"># _end_def_</span>

        <span class="c1"># If the crossover probability is higher than</span>
        <span class="c1"># a uniformly random value, make the changes.</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">is_operator_applicable</span><span class="p">():</span>

            <span class="c1"># Select randomly the crossover points and sort them.</span>
            <span class="n">loci</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">rng</span><span class="o">.</span><span class="n">choice</span><span class="p">(</span><span class="n">num_genes</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="n">num_points</span><span class="p">,</span>
                                          <span class="n">replace</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">shuffle</span><span class="o">=</span><span class="kc">False</span><span class="p">))</span>

            <span class="c1"># Initialize the offspring genomes to None.</span>
            <span class="n">genome_1</span> <span class="o">=</span> <span class="n">num_genes</span> <span class="o">*</span> <span class="p">[</span><span class="kc">None</span><span class="p">]</span>
            <span class="n">genome_2</span> <span class="o">=</span> <span class="n">num_genes</span> <span class="o">*</span> <span class="p">[</span><span class="kc">None</span><span class="p">]</span>

            <span class="c1"># Initialize a set of hyperparameters.</span>
            <span class="n">reset_flag</span><span class="p">,</span> <span class="n">upper_lim</span><span class="p">,</span> <span class="n">j</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span> <span class="n">loci</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="mi">0</span>

            <span class="c1"># Scan the whole list of genes.</span>
            <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">num_genes</span><span class="p">):</span>

                <span class="c1"># Once we surpass the upper limit (in loci)</span>
                <span class="c1"># we reset the  flag value to allow changes</span>
                <span class="c1"># to take place within that range.</span>
                <span class="k">if</span> <span class="n">i</span> <span class="o">&gt;=</span> <span class="n">upper_lim</span><span class="p">:</span>

                    <span class="c1"># Swap the reset flag.</span>
                    <span class="n">reset_flag</span> <span class="o">=</span> <span class="ow">not</span> <span class="n">reset_flag</span>

                    <span class="c1"># Increase the index of the loci.</span>
                    <span class="n">j</span> <span class="o">+=</span> <span class="mi">1</span>

                    <span class="c1"># We make sure the upper limit value does not exceed</span>
                    <span class="c1"># the number of genes. Also, this avoids the out of</span>
                    <span class="c1"># bound IndexError.</span>
                    <span class="n">upper_lim</span> <span class="o">=</span> <span class="n">loci</span><span class="p">[</span><span class="n">j</span><span class="p">]</span> <span class="k">if</span> <span class="n">j</span> <span class="o">&lt;</span> <span class="n">num_points</span> <span class="k">else</span> <span class="n">num_genes</span>
                <span class="c1"># _end_if_</span>

                <span class="c1"># Check the flag value.</span>
                <span class="k">if</span> <span class="n">reset_flag</span><span class="p">:</span>
                    <span class="n">genome_1</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="n">parent1</span><span class="o">.</span><span class="n">genome</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">clone</span><span class="p">()</span>
                    <span class="n">genome_2</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="n">parent2</span><span class="o">.</span><span class="n">genome</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">clone</span><span class="p">()</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">genome_1</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="n">parent2</span><span class="o">.</span><span class="n">genome</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">clone</span><span class="p">()</span>
                    <span class="n">genome_2</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="n">parent1</span><span class="o">.</span><span class="n">genome</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">clone</span><span class="p">()</span>
                <span class="c1"># _end_if_</span>

            <span class="c1"># _end_for_</span>

            <span class="c1"># Create the two NEW offsprings.</span>
            <span class="n">child1</span> <span class="o">=</span> <span class="n">Chromosome</span><span class="p">(</span><span class="n">genome_1</span><span class="p">)</span>
            <span class="n">child2</span> <span class="o">=</span> <span class="n">Chromosome</span><span class="p">(</span><span class="n">genome_2</span><span class="p">)</span>

            <span class="c1"># Increase the crossover counter.</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">inc_counter</span><span class="p">()</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="c1"># Each child points to a clone of a single parent.</span>
            <span class="n">child1</span> <span class="o">=</span> <span class="n">parent1</span><span class="o">.</span><span class="n">clone</span><span class="p">()</span>
            <span class="n">child2</span> <span class="o">=</span> <span class="n">parent2</span><span class="o">.</span><span class="n">clone</span><span class="p">()</span>
        <span class="c1"># _end_if_</span>

        <span class="c1"># Return the two offsprings.</span>
        <span class="k">return</span> <span class="n">child1</span><span class="p">,</span> <span class="n">child2</span></div>
</div>

    <span class="c1"># _end_def_</span>

<span class="c1"># _end_class_</span>
</pre></div>

           </div>
          </div>
          <footer>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2024, Michalis Vrettas, PhD.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>