<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="../../../">
<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>pygenalgo.genome.gene &mdash; PyGenAlgo 01.70.00 documentation</title>
      <link rel="stylesheet" type="text/css" href="../../../_static/pygments.css?v=92fd9be5" />
      <link rel="stylesheet" type="text/css" href="../../../_static/css/theme.css?v=19f00094" />

  
  <!--[if lt IE 9]>
    <script src="../../../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script src="../../../_static/jquery.js?v=5d32c60e"></script>
        <script src="../../../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
        <script src="../../../_static/documentation_options.js?v=fc45878b"></script>
        <script src="../../../_static/doctools.js?v=9a2dae69"></script>
        <script src="../../../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../../../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../../../index.html" class="icon icon-home">
            PyGenAlgo
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../modules.html">pygenalgo</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../../index.html">PyGenAlgo</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="../../index.html">Module code</a></li>
      <li class="breadcrumb-item active">pygenalgo.genome.gene</li>
      <li class="wy-breadcrumbs-aside">
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <h1>Source code for pygenalgo.genome.gene</h1><div class="highlight"><pre>
<span></span><span class="kn">from</span> <span class="nn">typing</span> <span class="kn">import</span> <span class="n">Any</span><span class="p">,</span> <span class="n">Callable</span>
<span class="kn">from</span> <span class="nn">copy</span> <span class="kn">import</span> <span class="n">deepcopy</span>
<span class="kn">from</span> <span class="nn">numpy.random</span> <span class="kn">import</span> <span class="n">default_rng</span><span class="p">,</span> <span class="n">Generator</span>


<div class="viewcode-block" id="Gene">
<a class="viewcode-back" href="../../../pygenalgo.genome.html#pygenalgo.genome.gene.Gene">[docs]</a>
<span class="k">class</span> <span class="nc">Gene</span><span class="p">(</span><span class="nb">object</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Description:</span>

<span class="sd">        This is the main class that encodes the data of a single Gene in the chromosome.</span>
<span class="sd">        The class encapsulates not only the data, but also the way that this gene can be</span>
<span class="sd">        mutated using a random function. This Gene can be from a single &#39;bit&#39; to a whole</span>
<span class="sd">        image. This way provides us with flexibility to parameterize the chromosome with</span>
<span class="sd">        different &quot;kinds of genes&quot; each one responsible for a specific function.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="c1"># Random number generator.</span>
    <span class="n">_rng</span><span class="p">:</span> <span class="n">Generator</span> <span class="o">=</span> <span class="n">default_rng</span><span class="p">()</span>

    <span class="c1"># Object variables.</span>
    <span class="vm">__slots__</span> <span class="o">=</span> <span class="p">(</span><span class="s2">&quot;_datum&quot;</span><span class="p">,</span> <span class="s2">&quot;_func&quot;</span><span class="p">,</span> <span class="s2">&quot;_valid&quot;</span><span class="p">)</span>

    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">datum</span><span class="p">:</span> <span class="n">Any</span><span class="p">,</span> <span class="n">func</span><span class="p">:</span> <span class="n">Callable</span><span class="p">,</span> <span class="n">valid</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Initialize a Gene object.</span>
<span class="sd">        </span>
<span class="sd">        :param datum: Datum holds a reference of the gene-data structure.</span>
<span class="sd">        </span>
<span class="sd">        :param func: This &#39;private&#39; function is used in the &#39;random()&#39; method to be used by the mutation operators.</span>
<span class="sd">        </span>
<span class="sd">        :param valid: This flag is used to set the Gene as valid (True) or invalid (False).</span>
<span class="sd">        &quot;&quot;&quot;</span>
        
        <span class="c1"># Copy the data reference.</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_datum</span> <span class="o">=</span> <span class="n">datum</span>
        
        <span class="c1"># Make sure the random function is callable.</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">callable</span><span class="p">(</span><span class="n">func</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="vm">__class__</span><span class="o">.</span><span class="vm">__name__</span><span class="si">}</span><span class="s2">: Random function is not callable.&quot;</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="c1"># Get the random function.</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_func</span> <span class="o">=</span> <span class="n">func</span>
        <span class="c1"># _end_if_</span>
        
        <span class="c1"># Copy the valid flag. Note that if the _datum field</span>
        <span class="c1"># is set to None, the Gene is automatically invalid.</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_valid</span> <span class="o">=</span> <span class="kc">False</span> <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">_datum</span> <span class="ow">is</span> <span class="kc">None</span> <span class="k">else</span> <span class="n">valid</span>
    <span class="c1"># _end_def_</span>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">value</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Any</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Accessor (getter) of the data reference.</span>

<span class="sd">        :return: the datum value.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_datum</span>
    <span class="c1"># _end_def_</span>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">is_valid</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">bool</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Accessor (getter) of the validity parameter.</span>

<span class="sd">        :return: the valid value.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_valid</span>
    <span class="c1"># _end_def_</span>

    <span class="nd">@is_valid</span><span class="o">.</span><span class="n">setter</span>
    <span class="k">def</span> <span class="nf">is_valid</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">new_value</span><span class="p">:</span> <span class="nb">bool</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Accessor (setter) of the validity flag.</span>

<span class="sd">        :param new_value: (bool).</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># Check for correct type.</span>
        <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">new_value</span><span class="p">,</span> <span class="nb">bool</span><span class="p">):</span>

            <span class="c1"># Update the flag value.</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_valid</span> <span class="o">=</span> <span class="n">new_value</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="vm">__class__</span><span class="o">.</span><span class="vm">__name__</span><span class="si">}</span><span class="s2">: &quot;</span>
                            <span class="sa">f</span><span class="s2">&quot;Validity flag should be bool: </span><span class="si">{</span><span class="nb">type</span><span class="p">(</span><span class="n">new_value</span><span class="p">)</span><span class="si">}</span><span class="s2">.&quot;</span><span class="p">)</span>
        <span class="c1"># _end_if_</span>
    <span class="c1"># _end_def_</span>

    <span class="k">def</span> <span class="fm">__eq__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">other</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">bool</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        When we compare two Genes we care only about the data they hold.</span>

<span class="sd">        :param other: the second object we want to compare to.</span>

<span class="sd">        :return: true or false.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="c1"># Make sure both items are of type &#39;Gene&#39;.</span>
        <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">other</span><span class="p">,</span> <span class="n">Gene</span><span class="p">):</span>
            <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_datum</span> <span class="o">==</span> <span class="n">other</span><span class="o">.</span><span class="n">_datum</span>
        <span class="c1"># _end_if_</span>

        <span class="k">return</span> <span class="kc">False</span>
    <span class="c1"># _end_def_</span>

    <span class="k">def</span> <span class="fm">__hash__</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">int</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Auxiliary method to hash the Gene object.</span>

<span class="sd">        :return: the hash value of the datum.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="nb">hash</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_datum</span><span class="p">)</span>
    <span class="c1"># _end_def_</span>

<div class="viewcode-block" id="Gene.random">
<a class="viewcode-back" href="../../../pygenalgo.genome.html#pygenalgo.genome.gene.Gene.random">[docs]</a>
    <span class="k">def</span> <span class="nf">random</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        This method should be different for each type of Gene. It describes</span>
<span class="sd">        how a specific type of Gene creates a random version of itself. The</span>
<span class="sd">        main  idea is that inside the Chromosome, each Gene can represent a</span>
<span class="sd">        very different concept of the  problem solution, so its Gene should</span>
<span class="sd">        have its own way to perform random mutation.</span>

<span class="sd">        This way by calling on the random() method, each Gene will know how</span>
<span class="sd">        to mutate itself without breaking any rules/constraints.</span>

<span class="sd">        :return: None.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="c1"># Use the random function to set a new value at the data.</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_datum</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_func</span><span class="p">()</span></div>

    <span class="c1"># _end_def_</span>

<div class="viewcode-block" id="Gene.flip">
<a class="viewcode-back" href="../../../pygenalgo.genome.html#pygenalgo.genome.gene.Gene.flip">[docs]</a>
    <span class="k">def</span> <span class="nf">flip</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        This method flips the value of the gene data. It is used only dy the</span>
<span class="sd">        FlipMutator operator for problems where the chromosome is represented</span>
<span class="sd">        by a list of bits.</span>

<span class="sd">         i)  1 -&gt; 0</span>
<span class="sd">        ii)  0 -&gt; 1</span>

<span class="sd">        :return: None.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="c1"># Flip the current gene value.</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_datum</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">_datum</span><span class="p">)</span></div>

    <span class="c1"># _end_def_</span>

<div class="viewcode-block" id="Gene.gaussian">
<a class="viewcode-back" href="../../../pygenalgo.genome.html#pygenalgo.genome.gene.Gene.gaussian">[docs]</a>
    <span class="k">def</span> <span class="nf">gaussian</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">sigma</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">1.0</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        This method adds a random value drawn from a standard normal</span>
<span class="sd">        distribution x ~ N(0, sigma) to the current gene data value.</span>
<span class="sd">        It is used mostly from the GaussianMutator method.</span>

<span class="sd">        :param sigma: (float) standard deviation of the Gaussian.</span>

<span class="sd">        :return: None.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="c1"># Add N(0, sigma) to the current gene value.</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_datum</span> <span class="o">+=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_rng</span><span class="o">.</span><span class="n">normal</span><span class="p">(</span><span class="n">scale</span><span class="o">=</span><span class="n">sigma</span><span class="p">)</span></div>

    <span class="c1"># _end_def_</span>

    <span class="k">def</span> <span class="fm">__str__</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Override to print a readable string presentation of the object.</span>

<span class="sd">        :return: a string representation of a Gene object.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="vm">__class__</span><span class="o">.</span><span class="vm">__name__</span><span class="si">}</span><span class="s2">: datum=</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">_datum</span><span class="si">}</span><span class="s2">&quot;</span>
    <span class="c1"># _end_def_</span>

    <span class="k">def</span> <span class="fm">__repr__</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Repr operator is called when a string representation is needed that can be evaluated.</span>

<span class="sd">        :return: Gene().</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="vm">__class__</span><span class="o">.</span><span class="vm">__name__</span><span class="si">}</span><span class="s2">(datum=</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">_datum</span><span class="si">}</span><span class="s2">, func=</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">_func</span><span class="si">}</span><span class="s2">, valid=</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">_valid</span><span class="si">}</span><span class="s2">)&quot;</span>
    <span class="c1"># _end_def_</span>

    <span class="k">def</span> <span class="nf">__copy__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        This custom method overrides the default copy method</span>
<span class="sd">        and is used when we call the copy() method on a class</span>
<span class="sd">        object.</span>

<span class="sd">        :return: a (shallow) copy of the self object.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="c1"># Return the new copy.</span>
        <span class="k">return</span> <span class="n">Gene</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_datum</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">_func</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">_valid</span><span class="p">)</span>
    <span class="c1"># _end_def_</span>

    <span class="k">def</span> <span class="nf">__deepcopy__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">memo</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        This custom method overrides the default deepcopy method</span>
<span class="sd">        and is used when we call the &quot;clone&quot; method of the class.</span>

<span class="sd">        :param memo: dictionary of objects already copied during</span>
<span class="sd">        the current copying pass.</span>

<span class="sd">        :return: a new identical &quot;clone&quot; of the self object.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="c1"># Create a new instance.</span>
        <span class="n">new_object</span> <span class="o">=</span> <span class="n">Gene</span><span class="o">.</span><span class="fm">__new__</span><span class="p">(</span><span class="n">Gene</span><span class="p">)</span>

        <span class="c1"># Don&#39;t copy self reference.</span>
        <span class="n">memo</span><span class="p">[</span><span class="nb">id</span><span class="p">(</span><span class="bp">self</span><span class="p">)]</span> <span class="o">=</span> <span class="n">new_object</span>

        <span class="c1"># Don&#39;t copy the cache.</span>
        <span class="k">if</span> <span class="nb">hasattr</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="s2">&quot;_cache&quot;</span><span class="p">):</span>
            <span class="n">memo</span><span class="p">[</span><span class="nb">id</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_cache</span><span class="p">)]</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_cache</span><span class="o">.</span><span class="fm">__new__</span><span class="p">(</span><span class="nb">dict</span><span class="p">)</span>
        <span class="c1"># _end_if_</span>

        <span class="c1"># Deep copy all other attributes.</span>
        <span class="k">for</span> <span class="n">attr</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="vm">__slots__</span><span class="p">:</span>
            <span class="nb">setattr</span><span class="p">(</span><span class="n">new_object</span><span class="p">,</span> <span class="n">attr</span><span class="p">,</span>
                    <span class="n">deepcopy</span><span class="p">(</span><span class="nb">getattr</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">attr</span><span class="p">),</span> <span class="n">memo</span><span class="p">))</span>
        <span class="c1"># _end_for_</span>

        <span class="c1"># Return identical instance.</span>
        <span class="k">return</span> <span class="n">new_object</span>
    <span class="c1"># _end_def_</span>

<div class="viewcode-block" id="Gene.clone">
<a class="viewcode-back" href="../../../pygenalgo.genome.html#pygenalgo.genome.gene.Gene.clone">[docs]</a>
    <span class="k">def</span> <span class="nf">clone</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Makes a duplicate of the self object.</span>

<span class="sd">        :return: a &quot;deep-copy&quot; of the object.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="n">deepcopy</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span></div>
</div>

    <span class="c1"># _end_def_</span>

<span class="c1"># _end_class_</span>
</pre></div>

           </div>
          </div>
          <footer>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2024, Michalis Vrettas, PhD.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>