

<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="../../../">
<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>pygenalgo.engines.auxiliary &mdash; PyGenAlgo 01.70.50 documentation</title>
      <link rel="stylesheet" type="text/css" href="../../../_static/pygments.css?v=92fd9be5" />
      <link rel="stylesheet" type="text/css" href="../../../_static/css/theme.css?v=e59714d7" />

  
      <script src="../../../_static/jquery.js?v=5d32c60e"></script>
      <script src="../../../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
      <script src="../../../_static/documentation_options.js?v=d0c4109e"></script>
      <script src="../../../_static/doctools.js?v=9bcbadda"></script>
      <script src="../../../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../../../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../../../index.html" class="icon icon-home">
            PyGenAlgo
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../modules.html">pygenalgo</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../../index.html">PyGenAlgo</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="../../index.html">Module code</a></li>
      <li class="breadcrumb-item active">pygenalgo.engines.auxiliary</li>
      <li class="wy-breadcrumbs-aside">
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <h1>Source code for pygenalgo.engines.auxiliary</h1><div class="highlight"><pre>
<span></span><span class="kn">from</span> <span class="nn">typing</span> <span class="kn">import</span> <span class="n">Callable</span>
<span class="kn">from</span> <span class="nn">dataclasses</span> <span class="kn">import</span> <span class="n">dataclass</span><span class="p">,</span> <span class="n">field</span>
<span class="kn">from</span> <span class="nn">pygenalgo.genome.chromosome</span> <span class="kn">import</span> <span class="n">Chromosome</span>

<span class="c1"># Public interface.</span>
<span class="n">__all__</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;hamming_distance&quot;</span><span class="p">,</span> <span class="s2">&quot;average_hamming_distance&quot;</span><span class="p">,</span>
           <span class="s2">&quot;apply_corrections&quot;</span><span class="p">,</span> <span class="s2">&quot;SubPopulation&quot;</span><span class="p">]</span>


<div class="viewcode-block" id="hamming_distance">
<a class="viewcode-back" href="../../../pygenalgo.engines.html#pygenalgo.engines.auxiliary.hamming_distance">[docs]</a>
<span class="k">def</span> <span class="nf">hamming_distance</span><span class="p">(</span><span class="n">item1</span><span class="p">:</span> <span class="n">Chromosome</span><span class="p">,</span> <span class="n">item2</span><span class="p">:</span> <span class="n">Chromosome</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">int</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Compute the Hamming distance of the &quot;item1&quot; object, with the</span>
<span class="sd">    &quot;item2&quot; chromosome. In practice it&#39;s the number of positions</span>
<span class="sd">    at which the corresponding genes are different.</span>

<span class="sd">    :param item1: (Chromosome) to compare the Hamming distance.</span>

<span class="sd">    :param item2: (Chromosome) to compare the Hamming distance.</span>

<span class="sd">    :return: (int) the number of dissimilarities between the two</span>
<span class="sd">    input chromosomes.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="c1"># Check for identical types.</span>
    <span class="k">if</span> <span class="nb">type</span><span class="p">(</span><span class="n">item1</span><span class="p">)</span> <span class="ow">is</span> <span class="nb">type</span><span class="p">(</span><span class="n">item2</span><span class="p">):</span>

        <span class="c1"># Quick exit if both objects are the</span>
        <span class="c1"># same or equal.</span>
        <span class="k">if</span> <span class="n">item1</span> <span class="ow">is</span> <span class="n">item2</span> <span class="ow">or</span> <span class="n">item1</span> <span class="o">==</span> <span class="n">item2</span><span class="p">:</span>
            <span class="k">return</span> <span class="mi">0</span>
        <span class="c1"># _end_if_</span>

        <span class="c1"># Exit with an error if both objects</span>
        <span class="c1"># are not the same length.</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">item1</span><span class="p">)</span> <span class="o">!=</span> <span class="nb">len</span><span class="p">(</span><span class="n">item2</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Input chromosomes must be of the same length.&quot;</span><span class="p">)</span>
        <span class="c1"># _end_if_</span>

        <span class="c1"># Compute the dissimilarities in their genomes.</span>
        <span class="k">return</span> <span class="p">[</span><span class="n">k</span> <span class="o">!=</span> <span class="n">l</span> <span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">l</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">item1</span><span class="o">.</span><span class="n">genome</span><span class="p">,</span> <span class="n">item2</span><span class="o">.</span><span class="n">genome</span><span class="p">)]</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="kc">True</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">RuntimeError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Can&#39;t compute Hamming distance in different type objects.&quot;</span><span class="p">)</span></div>

    <span class="c1"># _end_if_</span>

<span class="c1"># _end_def_</span>

<div class="viewcode-block" id="average_hamming_distance">
<a class="viewcode-back" href="../../../pygenalgo.engines.html#pygenalgo.engines.auxiliary.average_hamming_distance">[docs]</a>
<span class="k">def</span> <span class="nf">average_hamming_distance</span><span class="p">(</span><span class="n">population</span><span class="p">:</span> <span class="nb">list</span><span class="p">[</span><span class="n">Chromosome</span><span class="p">])</span> <span class="o">-&gt;</span> <span class="nb">float</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Computes the average Hamming distance of a population. We use this</span>
<span class="sd">    to measure the similarity in the whole population of chromosomes.</span>

<span class="sd">    :param population: List(Chromosome) the population we want to compute</span>
<span class="sd">    the average Hamming distance.</span>

<span class="sd">    :return: (float) the total number of differences, in the genes,</span>
<span class="sd">    divided by the total number of genes compared.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="c1"># Initialize the counters.</span>
    <span class="n">total_diffs</span><span class="p">,</span> <span class="n">total_genes</span> <span class="o">=</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span>

    <span class="c1"># Get the size of the chromosome. It is</span>
    <span class="c1"># assumed that all chromosomes have the</span>
    <span class="c1"># same size.</span>
    <span class="n">number_of_genes</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">population</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>

    <span class="c1"># Iterate through all the population.</span>
    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">item1</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">population</span><span class="p">):</span>

        <span class="c1"># Local copy of the 1st genome to</span>
        <span class="c1"># avoid recalling in the 2nd loop.</span>
        <span class="n">item1_genome</span> <span class="o">=</span> <span class="n">item1</span><span class="o">.</span><span class="n">genome</span>

        <span class="c1"># Compare the i-th chromosome with the rest of the population.</span>
        <span class="c1"># NOTE: Since the distances are symmetrical we don&#39;t check the</span>
        <span class="c1"># same pair of chromosomes twice.</span>
        <span class="k">for</span> <span class="n">item2</span> <span class="ow">in</span> <span class="n">population</span><span class="p">[</span><span class="n">i</span><span class="o">+</span><span class="mi">1</span><span class="p">:]:</span>
            <span class="c1"># Get the total number of different genes.</span>
            <span class="n">total_diffs</span> <span class="o">+=</span> <span class="p">[</span><span class="n">k</span> <span class="o">!=</span> <span class="n">l</span> <span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">l</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">item1_genome</span><span class="p">,</span>
                                                   <span class="n">item2</span><span class="o">.</span><span class="n">genome</span><span class="p">)]</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="kc">True</span><span class="p">)</span>
            <span class="c1"># We add the number of genes we tested.</span>
            <span class="n">total_genes</span> <span class="o">+=</span> <span class="n">number_of_genes</span>
        <span class="c1"># _end_for_</span>

    <span class="c1"># _end_for_</span>

    <span class="c1"># Sanity check.</span>
    <span class="k">if</span> <span class="n">total_genes</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">RuntimeError</span><span class="p">(</span><span class="s2">&quot;Average Humming Distance: Total number of Genes is zero.&quot;</span><span class="p">)</span>
    <span class="c1"># _end_if_</span>

    <span class="k">return</span> <span class="nb">float</span><span class="p">(</span><span class="n">total_diffs</span> <span class="o">/</span> <span class="n">total_genes</span><span class="p">)</span></div>

<span class="c1"># _end_def_</span>

<div class="viewcode-block" id="apply_corrections">
<a class="viewcode-back" href="../../../pygenalgo.engines.html#pygenalgo.engines.auxiliary.apply_corrections">[docs]</a>
<span class="k">def</span> <span class="nf">apply_corrections</span><span class="p">(</span><span class="n">input_population</span><span class="p">:</span> <span class="nb">list</span><span class="p">[</span><span class="n">Chromosome</span><span class="p">],</span>
                      <span class="n">fit_func</span><span class="p">:</span> <span class="n">Callable</span> <span class="o">=</span> <span class="kc">None</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">int</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Check the population  for invalid genes and correct them by applying directly</span>
<span class="sd">    the random method. It is assumed that the random method of the Gene is always</span>
<span class="sd">    returning a &#39;valid&#39; value for the Gene. After that, we need to reevaluate the</span>
<span class="sd">    chromosome to update its fitness.</span>

<span class="sd">    :param input_population: List(Chromosome) the population</span>
<span class="sd">    we want to apply corrections (if applicable).</span>

<span class="sd">    :param fit_func: callable fitness function.</span>

<span class="sd">    :return: the total number of corrected genes in the population.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="c1"># Holds the number of the corrected chromosomes.</span>
    <span class="n">corrections_counter</span> <span class="o">=</span> <span class="mi">0</span>

    <span class="c1"># Go through all the chromosomes of the input population.</span>
    <span class="k">for</span> <span class="n">chromosome</span> <span class="ow">in</span> <span class="n">input_population</span><span class="p">:</span>

        <span class="c1"># Holds the corrected genes.</span>
        <span class="n">corrected_genes</span> <span class="o">=</span> <span class="mi">0</span>

        <span class="c1"># Go through every Gene in the chromosome.</span>
        <span class="k">for</span> <span class="n">gene</span> <span class="ow">in</span> <span class="n">chromosome</span><span class="p">:</span>

            <span class="c1"># Check for validity.</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">gene</span><span class="o">.</span><span class="n">is_valid</span> <span class="ow">or</span> <span class="n">gene</span><span class="o">.</span><span class="n">value</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>

                <span class="c1"># Call the gene&#39;s random function.</span>
                <span class="n">gene</span><span class="o">.</span><span class="n">random</span><span class="p">()</span>

                <span class="c1"># Update the status of the gene.</span>
                <span class="n">gene</span><span class="o">.</span><span class="n">is_valid</span> <span class="o">=</span> <span class="kc">True</span>

                <span class="c1"># Update the counter.</span>
                <span class="n">corrected_genes</span> <span class="o">+=</span> <span class="mi">1</span>
            <span class="c1"># _end_if_</span>

        <span class="c1"># _end_for_</span>

        <span class="c1"># Check if there were any gene corrections.</span>
        <span class="k">if</span> <span class="n">corrected_genes</span><span class="p">:</span>

            <span class="c1"># Update the total corrections counter.</span>
            <span class="n">corrections_counter</span> <span class="o">+=</span> <span class="n">corrected_genes</span>

            <span class="c1"># Re-evaluate the fitness of the chromosome.</span>
            <span class="n">chromosome</span><span class="o">.</span><span class="n">fitness</span><span class="p">,</span> <span class="n">_</span> <span class="o">=</span> <span class="n">fit_func</span><span class="p">(</span><span class="n">chromosome</span><span class="p">)</span>
        <span class="c1"># _end_if_</span>

    <span class="c1"># _end_for_</span>

    <span class="c1"># Return the total number of corrected genes.</span>
    <span class="k">return</span> <span class="n">corrections_counter</span></div>

<span class="c1"># _end_def_</span>


<div class="viewcode-block" id="SubPopulation">
<a class="viewcode-back" href="../../../pygenalgo.engines.html#pygenalgo.engines.auxiliary.SubPopulation">[docs]</a>
<span class="nd">@dataclass</span><span class="p">(</span><span class="n">init</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="nb">repr</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="k">class</span> <span class="nc">SubPopulation</span><span class="p">(</span><span class="nb">object</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Auxiliary class container used in the IslandModelGA</span>
<span class="sd">    to hold all the subpopulations (one on each island).</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="c1"># SubPopulation ID.</span>
    <span class="n">pop_id</span><span class="p">:</span> <span class="nb">int</span>

    <span class="c1"># List of chromosomes.</span>
    <span class="n">population</span><span class="p">:</span> <span class="nb">list</span> <span class="o">=</span> <span class="n">field</span><span class="p">(</span><span class="n">default_factory</span><span class="o">=</span><span class="nb">list</span><span class="p">[</span><span class="n">Chromosome</span><span class="p">])</span>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">id</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">int</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Accessor (getter) of the id parameter.</span>

<span class="sd">        :return: the id value.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">pop_id</span>
    <span class="c1"># _end_def_</span>

    <span class="k">def</span> <span class="fm">__len__</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">int</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Accessor of the total length of the population.</span>

<span class="sd">        :return: the length (int) of the population.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">population</span><span class="p">)</span>
    <span class="c1"># _end_def_</span>

    <span class="k">def</span> <span class="fm">__getitem__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">index</span><span class="p">:</span> <span class="nb">int</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Chromosome</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Returns the reference of the Chromosome at position index.</span>

<span class="sd">        :param index: (int) position of chromosome to return.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">population</span><span class="p">[</span><span class="n">index</span><span class="p">]</span>
    <span class="c1"># _end_def_</span>

    <span class="k">def</span> <span class="fm">__setitem__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">index</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span> <span class="n">item</span><span class="p">:</span> <span class="n">Chromosome</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Sets the input Chromosome in the index position</span>
<span class="sd">        inside the (sub) population.</span>

<span class="sd">        :param index: (int) position in the population.</span>

<span class="sd">        :param item: Chromosome to attach to the new position.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">population</span><span class="p">[</span><span class="n">index</span><span class="p">]</span> <span class="o">=</span> <span class="n">item</span>
    <span class="c1"># _end_def_</span>

    <span class="k">def</span> <span class="fm">__contains__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">item</span><span class="p">:</span> <span class="n">Chromosome</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">bool</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Check for membership.</span>

<span class="sd">        :param item: an input Chromosome that we want to check</span>
<span class="sd">        if it exists in general population.</span>

<span class="sd">        :return: true if the &#39;item&#39; belongs in the population.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="n">item</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">population</span></div>

    <span class="c1"># _end_def_</span>

<span class="c1"># _end_class_</span>
</pre></div>

           </div>
          </div>
          <footer>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2024, Michalis Vrettas, PhD.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>