

<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="../../../">
<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>pygenalgo.engines.generic_ga &mdash; PyGenAlgo 01.70.50 documentation</title>
      <link rel="stylesheet" type="text/css" href="../../../_static/pygments.css?v=b86133f3" />
      <link rel="stylesheet" type="text/css" href="../../../_static/css/theme.css?v=e59714d7" />

  
      <script src="../../../_static/jquery.js?v=5d32c60e"></script>
      <script src="../../../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
      <script src="../../../_static/documentation_options.js?v=d0c4109e"></script>
      <script src="../../../_static/doctools.js?v=9a2dae69"></script>
      <script src="../../../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../../../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../../../index.html" class="icon icon-home">
            PyGenAlgo
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../modules.html">pygenalgo</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../../index.html">PyGenAlgo</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="../../index.html">Module code</a></li>
      <li class="breadcrumb-item active">pygenalgo.engines.generic_ga</li>
      <li class="wy-breadcrumbs-aside">
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <h1>Source code for pygenalgo.engines.generic_ga</h1><div class="highlight"><pre>
<span></span><span class="kn">from</span><span class="w"> </span><span class="nn">math</span><span class="w"> </span><span class="kn">import</span> <span class="n">isnan</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">os</span><span class="w"> </span><span class="kn">import</span> <span class="n">cpu_count</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">typing</span><span class="w"> </span><span class="kn">import</span> <span class="n">Callable</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">operator</span><span class="w"> </span><span class="kn">import</span> <span class="n">attrgetter</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">collections</span><span class="w"> </span><span class="kn">import</span> <span class="n">defaultdict</span>

<span class="kn">from</span><span class="w"> </span><span class="nn">joblib</span><span class="w"> </span><span class="kn">import</span> <span class="p">(</span><span class="n">Parallel</span><span class="p">,</span> <span class="n">delayed</span><span class="p">)</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">numpy.random</span><span class="w"> </span><span class="kn">import</span> <span class="n">default_rng</span><span class="p">,</span> <span class="n">Generator</span>

<span class="kn">from</span><span class="w"> </span><span class="nn">pygenalgo.genome.chromosome</span><span class="w"> </span><span class="kn">import</span> <span class="n">Chromosome</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">pygenalgo.operators.genetic_operator</span><span class="w"> </span><span class="kn">import</span> <span class="n">GeneticOperator</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">pygenalgo.operators.mutation.mutate_operator</span><span class="w"> </span><span class="kn">import</span> <span class="n">MutationOperator</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">pygenalgo.operators.selection.select_operator</span><span class="w"> </span><span class="kn">import</span> <span class="n">SelectionOperator</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">pygenalgo.operators.crossover.crossover_operator</span><span class="w"> </span><span class="kn">import</span> <span class="n">CrossoverOperator</span>

<span class="c1"># Public interface.</span>
<span class="n">__all__</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;GenericGA&quot;</span><span class="p">]</span>


<div class="viewcode-block" id="GenericGA">
<a class="viewcode-back" href="../../../pygenalgo.engines.html#pygenalgo.engines.generic_ga.GenericGA">[docs]</a>
<span class="k">class</span><span class="w"> </span><span class="nc">GenericGA</span><span class="p">(</span><span class="nb">object</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Description:</span>

<span class="sd">        Generic GA class models the interface of a specific genetic algorithm model (or engine).</span>
<span class="sd">        It provides the common variables and functionality that all GA models should share.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="c1"># Make a random number generator.</span>
    <span class="n">rng_GA</span><span class="p">:</span> <span class="n">Generator</span> <span class="o">=</span> <span class="n">default_rng</span><span class="p">()</span>

    <span class="c1"># Set the maximum number of CPUs (at least one).</span>
    <span class="n">MAX_CPUs</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">1</span> <span class="k">if</span> <span class="ow">not</span> <span class="n">cpu_count</span><span class="p">()</span> <span class="k">else</span> <span class="n">cpu_count</span><span class="p">()</span>

    <span class="c1"># Object variables.</span>
    <span class="vm">__slots__</span> <span class="o">=</span> <span class="p">(</span><span class="s2">&quot;population&quot;</span><span class="p">,</span> <span class="s2">&quot;fitness_func&quot;</span><span class="p">,</span> <span class="s2">&quot;_select_op&quot;</span><span class="p">,</span> <span class="s2">&quot;_crossx_op&quot;</span><span class="p">,</span> <span class="s2">&quot;_mutate_op&quot;</span><span class="p">,</span>
                 <span class="s2">&quot;_stats&quot;</span><span class="p">,</span> <span class="s2">&quot;_n_cpus&quot;</span><span class="p">,</span> <span class="s2">&quot;_f_eval&quot;</span><span class="p">,</span> <span class="s2">&quot;_iteration&quot;</span><span class="p">)</span>

    <span class="k">def</span><span class="w"> </span><span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">initial_pop</span><span class="p">:</span> <span class="nb">list</span><span class="p">[</span><span class="n">Chromosome</span><span class="p">],</span> <span class="n">fit_func</span><span class="p">:</span> <span class="n">Callable</span><span class="p">,</span> <span class="n">select_op</span><span class="p">:</span> <span class="n">SelectionOperator</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
                 <span class="n">mutate_op</span><span class="p">:</span> <span class="n">MutationOperator</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">crossx_op</span><span class="p">:</span> <span class="n">CrossoverOperator</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">n_cpus</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="kc">None</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Default constructor of GenericGA object.</span>

<span class="sd">        :param initial_pop: list of the initial population of (randomized) chromosomes.</span>

<span class="sd">        :param fit_func: callable fitness function.</span>

<span class="sd">        :param select_op: selection operator (must inherit from class SelectionOperator).</span>

<span class="sd">        :param mutate_op: mutation operator (must inherit from class MutationOperator).</span>

<span class="sd">        :param crossx_op: crossover operator (must inherit from class CrossoverOperator).</span>

<span class="sd">        :param n_cpus: Number of requested CPUs for the evolution process (Default=Max_CPU).</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="c1"># Copy the reference of the population.</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">population</span> <span class="o">=</span> <span class="n">initial_pop</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>

        <span class="c1"># Make sure the fitness function is indeed callable.</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">callable</span><span class="p">(</span><span class="n">fit_func</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="vm">__class__</span><span class="o">.</span><span class="vm">__name__</span><span class="si">}</span><span class="s2">: Fitness function is not callable.&quot;</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="c1"># Get the fitness function.</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">fitness_func</span> <span class="o">=</span> <span class="n">fit_func</span>
        <span class="c1"># _end_if_</span>

        <span class="c1"># Get Selection Operator.</span>
        <span class="k">if</span> <span class="n">select_op</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="vm">__class__</span><span class="o">.</span><span class="vm">__name__</span><span class="si">}</span><span class="s2">: Selection operator is missing.&quot;</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_select_op</span> <span class="o">=</span> <span class="n">select_op</span>
        <span class="c1"># _end_if_</span>

        <span class="c1"># Get Mutation Operator.</span>
        <span class="k">if</span> <span class="n">mutate_op</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="vm">__class__</span><span class="o">.</span><span class="vm">__name__</span><span class="si">}</span><span class="s2">: Mutation operator is missing.&quot;</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_mutate_op</span> <span class="o">=</span> <span class="n">mutate_op</span>
        <span class="c1"># _end_if_</span>

        <span class="c1"># Get Crossover Operator.</span>
        <span class="k">if</span> <span class="n">crossx_op</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="vm">__class__</span><span class="o">.</span><span class="vm">__name__</span><span class="si">}</span><span class="s2">: Crossover operator is missing.&quot;</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_crossx_op</span> <span class="o">=</span> <span class="n">crossx_op</span>
        <span class="c1"># _end_if_</span>

        <span class="c1"># Get the number of requested CPUs.</span>
        <span class="k">if</span> <span class="n">n_cpus</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>

            <span class="c1"># This is the default option.</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_n_cpus</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">GenericGA</span><span class="o">.</span><span class="n">MAX_CPUs</span><span class="o">-</span><span class="mi">1</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>

            <span class="c1"># Assign the  requested number, making sure we have</span>
            <span class="c1"># enough CPUs and the value entered has the correct</span>
            <span class="c1"># type.</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_n_cpus</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="nb">min</span><span class="p">(</span><span class="n">GenericGA</span><span class="o">.</span><span class="n">MAX_CPUs</span><span class="o">-</span><span class="mi">1</span><span class="p">,</span> <span class="nb">int</span><span class="p">(</span><span class="n">n_cpus</span><span class="p">)))</span>
        <span class="c1"># _end_if_</span>

        <span class="c1"># Dictionary with statistics.</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_stats</span> <span class="o">=</span> <span class="n">defaultdict</span><span class="p">(</span><span class="nb">list</span><span class="p">)</span>

        <span class="c1"># Set the function evaluation to zero.</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_f_eval</span> <span class="o">=</span> <span class="mi">0</span>

        <span class="c1"># Set the iterations counter to zero.</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_iteration</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="c1"># _end_def_</span>

    <span class="nd">@property</span>
    <span class="k">def</span><span class="w"> </span><span class="nf">iteration</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">int</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Accessor (getter) of the iteration parameter.</span>

<span class="sd">        :return: the iteration value.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_iteration</span>
    <span class="c1"># _end_def_</span>

    <span class="nd">@iteration</span><span class="o">.</span><span class="n">setter</span>
    <span class="k">def</span><span class="w"> </span><span class="nf">iteration</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">value</span><span class="p">:</span> <span class="nb">int</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Accessor (setter) of the iteration value.</span>

<span class="sd">        :param value: (int).</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># Check for correct type and allow only</span>
        <span class="c1"># the positive values.</span>
        <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">value</span><span class="p">,</span> <span class="nb">int</span><span class="p">)</span> <span class="ow">and</span> <span class="n">value</span> <span class="o">&gt;=</span> <span class="mi">0</span><span class="p">:</span>

            <span class="c1"># Update the iteration value.</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_iteration</span> <span class="o">=</span> <span class="n">value</span>

            <span class="c1"># Update the iteration value in the GeneticOperator Class.</span>
            <span class="n">GeneticOperator</span><span class="o">.</span><span class="n">set_iteration</span><span class="p">(</span><span class="n">value</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">RuntimeError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="vm">__class__</span><span class="o">.</span><span class="vm">__name__</span><span class="si">}</span><span class="s2">: &quot;</span>
                               <span class="sa">f</span><span class="s2">&quot;Iteration value should be positive int: </span><span class="si">{</span><span class="nb">type</span><span class="p">(</span><span class="n">value</span><span class="p">)</span><span class="si">}</span><span class="s2">.&quot;</span><span class="p">)</span>
    <span class="c1"># _end_def_</span>

<div class="viewcode-block" id="GenericGA.set_seed">
<a class="viewcode-back" href="../../../pygenalgo.engines.html#pygenalgo.engines.generic_ga.GenericGA.set_seed">[docs]</a>
    <span class="nd">@classmethod</span>
    <span class="k">def</span><span class="w"> </span><span class="nf">set_seed</span><span class="p">(</span><span class="bp">cls</span><span class="p">,</span> <span class="n">new_seed</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Sets a new seed for the random number generator.</span>

<span class="sd">        :param new_seed: New seed value (default=None).</span>

<span class="sd">        :return: None.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># Re-initialize the class variable.</span>
        <span class="bp">cls</span><span class="o">.</span><span class="n">rng_GA</span> <span class="o">=</span> <span class="n">default_rng</span><span class="p">(</span><span class="n">seed</span><span class="o">=</span><span class="n">new_seed</span><span class="p">)</span></div>

    <span class="c1"># _end_def_</span>

    <span class="nd">@property</span>
    <span class="k">def</span><span class="w"> </span><span class="nf">f_eval</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">int</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Accessor method that returns the value of the f_eval.</span>

<span class="sd">        :return: (int) the counted number of function evaluations.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_f_eval</span>
    <span class="c1"># _end_def_</span>

    <span class="nd">@property</span>
    <span class="k">def</span><span class="w"> </span><span class="nf">stats</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">dict</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Accessor method that returns the &#39;stats&#39; dictionary.</span>

<span class="sd">        :return: the dictionary with the statistics from the run.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_stats</span>
    <span class="c1"># _end_def_</span>

    <span class="nd">@property</span>
    <span class="k">def</span><span class="w"> </span><span class="nf">select_op</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">SelectionOperator</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Accessor method that returns the selection operator reference.</span>

<span class="sd">        :return: the SelectionOperator.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_select_op</span>
    <span class="c1"># _end_def_</span>

    <span class="nd">@property</span>
    <span class="k">def</span><span class="w"> </span><span class="nf">crossx_op</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">CrossoverOperator</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Accessor method that returns the crossover operator reference.</span>

<span class="sd">        :return: the CrossoverOperator.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_crossx_op</span>
    <span class="c1"># _end_def_</span>

    <span class="nd">@property</span>
    <span class="k">def</span><span class="w"> </span><span class="nf">mutate_op</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">MutationOperator</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Accessor method that returns the mutation operator reference.</span>

<span class="sd">        :return: the MutationOperator.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_mutate_op</span>
    <span class="c1"># _end_def_</span>

    <span class="nd">@property</span>
    <span class="k">def</span><span class="w"> </span><span class="nf">n_cpus</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">int</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Accessor method that returns the number of CPUs.</span>

<span class="sd">        :return: the n_cpus.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_n_cpus</span>
    <span class="c1"># _end_def_</span>

<div class="viewcode-block" id="GenericGA.f_eval_increase_by">
<a class="viewcode-back" href="../../../pygenalgo.engines.html#pygenalgo.engines.generic_ga.GenericGA.f_eval_increase_by">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">f_eval_increase_by</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">new_counts</span><span class="p">:</span> <span class="nb">int</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Utility method to allow the &#39;_f_eval&#39; to be updated with</span>
<span class="sd">        new counts outside the main class. This can happen during</span>
<span class="sd">        gene correction.</span>

<span class="sd">        :param new_counts: the new value of &#39;counts&#39; that we want</span>
<span class="sd">        to add on the current f_eval.</span>

<span class="sd">        :return: None.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="c1"># Sanity check.</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">new_counts</span><span class="p">,</span> <span class="nb">int</span><span class="p">)</span> <span class="ow">or</span> <span class="n">new_counts</span> <span class="o">&lt;</span> <span class="mi">0</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="vm">__class__</span><span class="o">.</span><span class="vm">__name__</span><span class="si">}</span><span class="s2">: &quot;</span>
                             <span class="sa">f</span><span class="s2">&quot;New counts must be positive integer.&quot;</span><span class="p">)</span>
        <span class="c1"># _end_if_</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">_f_eval</span> <span class="o">+=</span> <span class="n">new_counts</span></div>

    <span class="c1"># _end_def_</span>

<div class="viewcode-block" id="GenericGA.clear_all">
<a class="viewcode-back" href="../../../pygenalgo.engines.html#pygenalgo.engines.generic_ga.GenericGA.clear_all">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">clear_all</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Make sure all the genetic operator counters and the stats</span>
<span class="sd">        are cleared. This reset everything before each run().</span>

<span class="sd">        :return: None.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># Ensure the genetic operator counters are reset before.</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_crossx_op</span><span class="o">.</span><span class="n">reset_counter</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_mutate_op</span><span class="o">.</span><span class="n">reset_counter</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_select_op</span><span class="o">.</span><span class="n">reset_counter</span><span class="p">()</span>

        <span class="c1"># Reset stats dictionary.</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_stats</span><span class="o">.</span><span class="n">clear</span><span class="p">()</span>

        <span class="c1"># Reset f_eval counter.</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_f_eval</span> <span class="o">=</span> <span class="mi">0</span></div>

    <span class="c1"># _end_def_</span>

<div class="viewcode-block" id="GenericGA.best_chromosome">
<a class="viewcode-back" href="../../../pygenalgo.engines.html#pygenalgo.engines.generic_ga.GenericGA.best_chromosome">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">best_chromosome</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Chromosome</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Auxiliary method that returns the chromosome with the</span>
<span class="sd">        highest fitness value. Safeguarded with ignoring NaNs.</span>

<span class="sd">        :return: Return the chromosome with the highest fitness.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># Return the chromosome with the highest fitness.</span>
        <span class="k">return</span> <span class="nb">max</span><span class="p">((</span><span class="n">p</span> <span class="k">for</span> <span class="n">p</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">population</span> <span class="k">if</span> <span class="ow">not</span> <span class="n">isnan</span><span class="p">(</span><span class="n">p</span><span class="o">.</span><span class="n">fitness</span><span class="p">)),</span>
                   <span class="n">key</span><span class="o">=</span><span class="n">attrgetter</span><span class="p">(</span><span class="s2">&quot;fitness&quot;</span><span class="p">),</span> <span class="n">default</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span></div>

    <span class="c1"># _end_def_</span>

<div class="viewcode-block" id="GenericGA.best_n">
<a class="viewcode-back" href="../../../pygenalgo.engines.html#pygenalgo.engines.generic_ga.GenericGA.best_n">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">best_n</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">n</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">1</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">list</span><span class="p">[</span><span class="n">Chromosome</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Auxiliary method that returns the best &#39;n&#39; chromosomes</span>
<span class="sd">        with the highest fitness value.</span>

<span class="sd">        :param n: the number of the best chromosomes. Default = 1.</span>

<span class="sd">        :return: Return the &#39;n&#39; chromosomes with the highest fitness.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="c1"># Make sure &#39;n&#39; is positive integer.</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">n</span><span class="p">,</span> <span class="nb">int</span><span class="p">)</span> <span class="ow">or</span> <span class="n">n</span> <span class="o">&lt;=</span> <span class="mi">0</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="vm">__class__</span><span class="o">.</span><span class="vm">__name__</span><span class="si">}</span><span class="s2">: &quot;</span>
                             <span class="sa">f</span><span class="s2">&quot;Input must be a positive integer.&quot;</span><span class="p">)</span>
        <span class="c1"># _end_if_</span>

        <span class="c1"># Ensure the number of return chromosome</span>
        <span class="c1"># do not exceed the size of the population.</span>
        <span class="k">if</span> <span class="n">n</span> <span class="o">&gt;</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">population</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">RuntimeError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="vm">__class__</span><span class="o">.</span><span class="vm">__name__</span><span class="si">}</span><span class="s2">: &quot;</span>
                               <span class="sa">f</span><span class="s2">&quot;Best </span><span class="si">{</span><span class="n">n</span><span class="si">}</span><span class="s2"> exceeds population size.&quot;</span><span class="p">)</span>
        <span class="c1"># _end_if_</span>

        <span class="c1"># Sort the population in descending order.</span>
        <span class="n">sorted_population</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">([</span><span class="n">p</span> <span class="k">for</span> <span class="n">p</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">population</span> <span class="k">if</span> <span class="ow">not</span> <span class="n">isnan</span><span class="p">(</span><span class="n">p</span><span class="o">.</span><span class="n">fitness</span><span class="p">)],</span>
                                   <span class="n">key</span><span class="o">=</span><span class="n">attrgetter</span><span class="p">(</span><span class="s2">&quot;fitness&quot;</span><span class="p">),</span> <span class="n">reverse</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

        <span class="c1"># Return the best &#39;n&#39; chromosomes.</span>
        <span class="k">return</span> <span class="n">sorted_population</span><span class="p">[</span><span class="mi">0</span><span class="p">:</span><span class="n">n</span><span class="p">]</span></div>

    <span class="c1"># _end_def_</span>

<div class="viewcode-block" id="GenericGA.crossover_mutate">
<a class="viewcode-back" href="../../../pygenalgo.engines.html#pygenalgo.engines.generic_ga.GenericGA.crossover_mutate">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">crossover_mutate</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">input_population</span><span class="p">:</span> <span class="nb">list</span><span class="p">[</span><span class="n">Chromosome</span><span class="p">])</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        This is an auxiliary method that combines the crossover and mutation</span>
<span class="sd">        operations in one call. Since these operations happen in place the</span>
<span class="sd">        &#39;input_population&#39; will be modified directly.</span>

<span class="sd">        This method should be called AFTER the selection of the parents that</span>
<span class="sd">        have been selected for breeding.</span>

<span class="sd">        :param input_population: this is the population that we will apply</span>
<span class="sd">        the two genetic operators.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="c1"># Get the size of the input population.</span>
        <span class="n">pop_size</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">input_population</span><span class="p">)</span>

        <span class="c1"># CROSSOVER and MUTATE to produce the new offsprings.</span>
        <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">pop_size</span><span class="p">,</span> <span class="mi">2</span><span class="p">):</span>

            <span class="c1"># In case of &#39;odd sized&#39; populations.</span>
            <span class="k">if</span> <span class="n">j</span> <span class="o">==</span> <span class="n">pop_size</span><span class="o">-</span><span class="mi">1</span><span class="p">:</span>

                <span class="c1"># MUTATE in place the last offspring.</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">_mutate_op</span><span class="p">(</span><span class="n">input_population</span><span class="p">[</span><span class="n">j</span><span class="p">])</span>

                <span class="c1"># Exit the loop.</span>
                <span class="k">break</span>
            <span class="c1"># _end_if_</span>

            <span class="c1"># Get the index of the next chromosome.</span>
            <span class="n">k</span> <span class="o">=</span> <span class="n">j</span><span class="o">+</span><span class="mi">1</span>

            <span class="c1"># Replace directly the OLD parents with the NEW offsprings.</span>
            <span class="n">input_population</span><span class="p">[</span><span class="n">j</span><span class="p">],</span> <span class="n">input_population</span><span class="p">[</span><span class="n">k</span><span class="p">]</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_crossx_op</span><span class="p">(</span><span class="n">input_population</span><span class="p">[</span><span class="n">j</span><span class="p">],</span>
                                                                       <span class="n">input_population</span><span class="p">[</span><span class="n">k</span><span class="p">])</span>
            <span class="c1"># MUTATE in place the 1st offspring.</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_mutate_op</span><span class="p">(</span><span class="n">input_population</span><span class="p">[</span><span class="n">j</span><span class="p">])</span>

            <span class="c1"># MUTATE in place the 2nd offspring.</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_mutate_op</span><span class="p">(</span><span class="n">input_population</span><span class="p">[</span><span class="n">k</span><span class="p">])</span></div>

        <span class="c1"># _end_for_</span>
    <span class="c1"># _end_def_</span>

<div class="viewcode-block" id="GenericGA.adapt_probabilities">
<a class="viewcode-back" href="../../../pygenalgo.engines.html#pygenalgo.engines.generic_ga.GenericGA.adapt_probabilities">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">adapt_probabilities</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">threshold</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="kc">None</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">bool</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        This method is used (optionally) to adjust simultaneously the crossover</span>
<span class="sd">        and mutation parameters of the GenericGA object.</span>

<span class="sd">        :param threshold: (float) This parameter is used to determine whether we</span>
<span class="sd">        are going to increase or decrease the crossover and mutation parameters.</span>

<span class="sd">        :return: True if the parameters have changed, else False.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="c1"># Check if the threshold value is missing.</span>
        <span class="k">if</span> <span class="n">threshold</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">RuntimeError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="vm">__class__</span><span class="o">.</span><span class="vm">__name__</span><span class="si">}</span><span class="s2">: &quot;</span>
                               <span class="sa">f</span><span class="s2">&quot;Threshold parameter is missing.&quot;</span><span class="p">)</span>
        <span class="c1"># _end_if_</span>

        <span class="c1"># Sanity check.</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">threshold</span><span class="p">,</span> <span class="nb">float</span><span class="p">)</span> <span class="ow">or</span> <span class="n">threshold</span> <span class="o">&lt;</span> <span class="mf">0.0</span> <span class="ow">or</span> <span class="n">threshold</span> <span class="o">&gt;</span> <span class="mf">1.0</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="vm">__class__</span><span class="o">.</span><span class="vm">__name__</span><span class="si">}</span><span class="s2">: &quot;</span>
                             <span class="sa">f</span><span class="s2">&quot;Threshold value must be float in [0.0, 1.0].&quot;</span><span class="p">)</span>
        <span class="c1"># _end_if_</span>

        <span class="c1"># Initialize the trial values with the current</span>
        <span class="c1"># probabilities to avoid going out of limits.</span>
        <span class="n">trial_pc</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_crossx_op</span><span class="o">.</span><span class="n">probability</span>
        <span class="n">trial_pm</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_mutate_op</span><span class="o">.</span><span class="n">probability</span>

        <span class="c1"># Initialize the flag with &quot;False&quot;</span>
        <span class="c1"># to avoid unnecessary assignments.</span>
        <span class="n">have_changed</span> <span class="o">=</span> <span class="kc">False</span>

        <span class="c1"># Use the threshold value to adjust</span>
        <span class="c1"># the probabilities accordingly.</span>
        <span class="k">if</span> <span class="n">threshold</span> <span class="o">&lt;</span> <span class="mf">0.1</span><span class="p">:</span>

            <span class="n">trial_pc</span> <span class="o">*=</span> <span class="mf">0.9</span>
            <span class="n">trial_pm</span> <span class="o">*=</span> <span class="mf">1.1</span>
            <span class="n">have_changed</span> <span class="o">=</span> <span class="kc">True</span>

        <span class="k">elif</span> <span class="n">threshold</span> <span class="o">&gt;</span> <span class="mf">0.8</span><span class="p">:</span>

            <span class="n">trial_pc</span> <span class="o">*=</span> <span class="mf">1.1</span>
            <span class="n">trial_pm</span> <span class="o">*=</span> <span class="mf">0.9</span>
            <span class="n">have_changed</span> <span class="o">=</span> <span class="kc">True</span>
        <span class="c1"># _end_if_</span>

        <span class="c1"># Check if the probabilities have changed.</span>
        <span class="k">if</span> <span class="n">have_changed</span><span class="p">:</span>
            <span class="c1"># Ensure the probabilities stay within the range [0, 1].</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_crossx_op</span><span class="o">.</span><span class="n">probability</span> <span class="o">=</span> <span class="nb">min</span><span class="p">(</span><span class="nb">max</span><span class="p">(</span><span class="n">trial_pc</span><span class="p">,</span> <span class="mf">0.0</span><span class="p">),</span> <span class="mf">1.0</span><span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_mutate_op</span><span class="o">.</span><span class="n">probability</span> <span class="o">=</span> <span class="nb">min</span><span class="p">(</span><span class="nb">max</span><span class="p">(</span><span class="n">trial_pm</span><span class="p">,</span> <span class="mf">0.0</span><span class="p">),</span> <span class="mf">1.0</span><span class="p">)</span>
        <span class="c1"># _end_if_</span>

        <span class="k">return</span> <span class="n">have_changed</span></div>

    <span class="c1"># _end_def_</span>

<div class="viewcode-block" id="GenericGA.population_fitness">
<a class="viewcode-back" href="../../../pygenalgo.engines.html#pygenalgo.engines.generic_ga.GenericGA.population_fitness">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">population_fitness</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">list</span><span class="p">[</span><span class="nb">float</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Get the fitness values of all the population.</span>

<span class="sd">        :return: A list with all the fitness values.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="p">[</span><span class="n">p</span><span class="o">.</span><span class="n">fitness</span> <span class="k">for</span> <span class="n">p</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">population</span><span class="p">]</span></div>

    <span class="c1"># _end_def_</span>

<div class="viewcode-block" id="GenericGA.individual_fitness">
<a class="viewcode-back" href="../../../pygenalgo.engines.html#pygenalgo.engines.generic_ga.GenericGA.individual_fitness">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">individual_fitness</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">index</span><span class="p">:</span> <span class="nb">int</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">float</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Get the fitness value of an individual member of the population.</span>

<span class="sd">        :param index: Position of the individual in the population.</span>

<span class="sd">        :return: The fitness value (float).</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">population</span><span class="p">[</span><span class="n">index</span><span class="p">]</span><span class="o">.</span><span class="n">fitness</span></div>

    <span class="c1"># _end_def_</span>

<div class="viewcode-block" id="GenericGA.evaluate_fitness">
<a class="viewcode-back" href="../../../pygenalgo.engines.html#pygenalgo.engines.generic_ga.GenericGA.evaluate_fitness">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">evaluate_fitness</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">input_population</span><span class="p">:</span> <span class="nb">list</span><span class="p">[</span><span class="n">Chromosome</span><span class="p">],</span>
                         <span class="n">parallel_mode</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
                         <span class="n">backend</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;threading&quot;</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">tuple</span><span class="p">[</span><span class="nb">list</span><span class="p">[</span><span class="nb">float</span><span class="p">],</span> <span class="nb">bool</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Evaluate all the chromosomes of the input list with the custom</span>
<span class="sd">        fitness function. The parallel_mode is optional. Moreover, the</span>
<span class="sd">        default backend is &quot;threading&quot;, but in the IslandModelGA it is</span>
<span class="sd">        better to select &quot;loky&quot;.</span>

<span class="sd">        :param input_population: (list) The population of Chromosomes</span>
<span class="sd">        that we want to evaluate their fitness.</span>

<span class="sd">        :param parallel_mode: (bool) Enables parallel computation of</span>
<span class="sd">        the fitness function.</span>

<span class="sd">        :param backend: Backend for the parallel Joblib framework.</span>

<span class="sd">        :return: a list with the fitness values and the found solution flag.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="c1"># Get a local copy of the fitness function.</span>
        <span class="n">fit_func</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">fitness_func</span>

        <span class="c1"># Check the &#39;parallel_mode&#39; flag.</span>
        <span class="k">if</span> <span class="n">parallel_mode</span><span class="p">:</span>

            <span class="c1"># Evaluate the chromosomes in parallel mode.</span>
            <span class="n">fitness_i</span> <span class="o">=</span> <span class="n">Parallel</span><span class="p">(</span><span class="n">n_jobs</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">_n_cpus</span><span class="p">,</span> <span class="n">backend</span><span class="o">=</span><span class="n">backend</span><span class="p">)(</span>
                <span class="n">delayed</span><span class="p">(</span><span class="n">fit_func</span><span class="p">)(</span><span class="n">p</span><span class="p">)</span> <span class="k">for</span> <span class="n">p</span> <span class="ow">in</span> <span class="n">input_population</span>
            <span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>

            <span class="c1"># Evaluate the chromosomes in serial mode.</span>
            <span class="n">fitness_i</span> <span class="o">=</span> <span class="p">[</span><span class="n">fit_func</span><span class="p">(</span><span class="n">p</span><span class="p">)</span> <span class="k">for</span> <span class="n">p</span> <span class="ow">in</span> <span class="n">input_population</span><span class="p">]</span>
        <span class="c1"># _end_if_</span>

        <span class="c1"># Get the size of the population.</span>
        <span class="n">p_size</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">fitness_i</span><span class="p">)</span>

        <span class="c1"># Preallocate the fitness list.</span>
        <span class="n">fitness_values</span> <span class="o">=</span> <span class="n">p_size</span> <span class="o">*</span> <span class="p">[</span><span class="nb">float</span><span class="p">(</span><span class="s2">&quot;NaN&quot;</span><span class="p">)]</span>

        <span class="c1"># Flag to indicate if a solution has been found.</span>
        <span class="n">found_solution</span> <span class="o">=</span> <span class="kc">False</span>

        <span class="c1"># Update all chromosomes with their fitness and check if a solution</span>
        <span class="c1"># has been found.</span>
        <span class="k">for</span> <span class="n">n</span><span class="p">,</span> <span class="p">(</span><span class="n">p</span><span class="p">,</span> <span class="n">fit_result</span><span class="p">)</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="n">input_population</span><span class="p">,</span> <span class="n">fitness_i</span><span class="p">)):</span>
            <span class="c1"># Attach the fitness to each chromosome.</span>
            <span class="n">p</span><span class="o">.</span><span class="n">fitness</span> <span class="o">=</span> <span class="n">fit_result</span><span class="p">[</span><span class="s2">&quot;f_value&quot;</span><span class="p">]</span>

            <span class="c1"># Collect the fitness in a separate list.</span>
            <span class="n">fitness_values</span><span class="p">[</span><span class="n">n</span><span class="p">]</span> <span class="o">=</span> <span class="n">fit_result</span><span class="p">[</span><span class="s2">&quot;f_value&quot;</span><span class="p">]</span>

            <span class="c1"># Update the &quot;found solution&quot;.</span>
            <span class="n">found_solution</span> <span class="o">|=</span> <span class="n">fit_result</span><span class="p">[</span><span class="s2">&quot;solution_is_found&quot;</span><span class="p">]</span>
        <span class="c1"># _end_for_</span>

        <span class="c1"># Update the counter of function evaluations.</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_f_eval</span> <span class="o">+=</span> <span class="n">p_size</span>

        <span class="c1"># Return the fitness values.</span>
        <span class="k">return</span> <span class="n">fitness_values</span><span class="p">,</span> <span class="n">found_solution</span></div>

    <span class="c1"># _end_def_</span>

<div class="viewcode-block" id="GenericGA.run">
<a class="viewcode-back" href="../../../pygenalgo.engines.html#pygenalgo.engines.generic_ga.GenericGA.run">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">run</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="o">*</span><span class="n">args</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Main method of the Generic GA class, that implements the evolutionary routine.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">raise</span> <span class="ne">NotImplementedError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="vm">__class__</span><span class="o">.</span><span class="vm">__name__</span><span class="si">}</span><span class="s2">: &quot;</span>
                                  <span class="sa">f</span><span class="s2">&quot;You should implement this method!&quot;</span><span class="p">)</span></div>

    <span class="c1"># _end_def_</span>

    <span class="k">def</span><span class="w"> </span><span class="fm">__call__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="o">*</span><span class="n">args</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        This method is only a wrapper of the &quot;run&quot; method.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">run</span><span class="p">(</span><span class="o">*</span><span class="n">args</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span></div>

    <span class="c1"># _end_def_</span>

<span class="c1"># _end_class_</span>
</pre></div>

           </div>
          </div>
          <footer>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2024, Michalis Vrettas, PhD.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>