

<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="../../../">
<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>pygenalgo.engines.island_model_ga &mdash; PyGenAlgo 1.9.0 documentation</title>
      <link rel="stylesheet" type="text/css" href="../../../_static/pygments.css?v=b86133f3" />
      <link rel="stylesheet" type="text/css" href="../../../_static/css/theme.css?v=e59714d7" />

  
      <script src="../../../_static/jquery.js?v=5d32c60e"></script>
      <script src="../../../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
      <script src="../../../_static/documentation_options.js?v=c4cb232b"></script>
      <script src="../../../_static/doctools.js?v=9a2dae69"></script>
      <script src="../../../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../../../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../../../index.html" class="icon icon-home">
            PyGenAlgo
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../modules.html">Modules</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../examples.html">Examples</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../../index.html">PyGenAlgo</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="../../index.html">Module code</a></li>
      <li class="breadcrumb-item active">pygenalgo.engines.island_model_ga</li>
      <li class="wy-breadcrumbs-aside">
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <h1>Source code for pygenalgo.engines.island_model_ga</h1><div class="highlight"><pre>
<span></span><span class="kn">import</span><span class="w"> </span><span class="nn">time</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">math</span><span class="w"> </span><span class="kn">import</span> <span class="n">isnan</span><span class="p">,</span> <span class="n">isclose</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">operator</span><span class="w"> </span><span class="kn">import</span> <span class="n">attrgetter</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">collections</span><span class="w"> </span><span class="kn">import</span> <span class="n">defaultdict</span>

<span class="kn">from</span><span class="w"> </span><span class="nn">joblib</span><span class="w"> </span><span class="kn">import</span> <span class="p">(</span><span class="n">Parallel</span><span class="p">,</span> <span class="n">delayed</span><span class="p">)</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">numpy</span><span class="w"> </span><span class="kn">import</span> <span class="p">(</span><span class="n">nanmean</span><span class="p">,</span> <span class="n">nanstd</span><span class="p">,</span> <span class="n">isfinite</span><span class="p">)</span>

<span class="kn">from</span><span class="w"> </span><span class="nn">pygenalgo.engines</span><span class="w"> </span><span class="kn">import</span> <span class="n">logger</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">pygenalgo.utils.auxiliary</span><span class="w"> </span><span class="kn">import</span> <span class="p">(</span><span class="n">SubPopulation</span><span class="p">,</span>
                                       <span class="n">apply_corrections</span><span class="p">,</span>
                                       <span class="n">average_hamming_distance</span><span class="p">)</span>

<span class="kn">from</span><span class="w"> </span><span class="nn">pygenalgo.engines.generic_ga</span><span class="w"> </span><span class="kn">import</span> <span class="n">GenericGA</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">pygenalgo.operators.migration.meta_migration</span><span class="w"> </span><span class="kn">import</span> <span class="n">MetaMigration</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">pygenalgo.operators.migration.migration_operator</span><span class="w"> </span><span class="kn">import</span> <span class="n">MigrationOperator</span>

<span class="c1"># Public interface.</span>
<span class="n">__all__</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;IslandModelGA&quot;</span><span class="p">]</span>


<div class="viewcode-block" id="IslandModelGA">
<a class="viewcode-back" href="../../../pygenalgo.engines.html#pygenalgo.engines.island_model_ga.IslandModelGA">[docs]</a>
<span class="k">class</span><span class="w"> </span><span class="nc">IslandModelGA</span><span class="p">(</span><span class="n">GenericGA</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Description:</span>

<span class="sd">        In Island Model GA we run in parallel a number of &quot;islands&quot;, each one evolving its own</span>
<span class="sd">        (sub)-population. Optionally we can allow &quot;migration&quot;, among the best individuals from</span>
<span class="sd">        each island.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="c1"># Object variables (specific for the IslandModel).</span>
    <span class="vm">__slots__</span> <span class="o">=</span> <span class="p">(</span><span class="s2">&quot;_num_islands&quot;</span><span class="p">,</span> <span class="s2">&quot;_migrate_op&quot;</span><span class="p">)</span>

    <span class="k">def</span><span class="w"> </span><span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">num_islands</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span> <span class="n">migrate_op</span><span class="p">:</span> <span class="n">MigrationOperator</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Default constructor of IslandModelGA object.</span>

<span class="sd">        :param num_islands: (int) number of parallel evolving islands.</span>

<span class="sd">        :param migrate_op: migration operator (must inherit from class MigrationOperator).</span>

<span class="sd">        :return: a new GA object.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># Call the super constructor with all the input parameters.</span>
        <span class="nb">super</span><span class="p">()</span><span class="o">.</span><span class="fm">__init__</span><span class="p">(</span><span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>

        <span class="c1"># Sanity check.</span>
        <span class="k">if</span> <span class="n">num_islands</span> <span class="o">&gt;</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">population</span><span class="p">):</span>
            <span class="c1"># Raise an error if number of islands is too high.</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="vm">__class__</span><span class="o">.</span><span class="vm">__name__</span><span class="si">}</span><span class="s2">: &quot;</span>
                             <span class="sa">f</span><span class="s2">&quot;Number of requested islands (</span><span class="si">{</span><span class="n">num_islands</span><span class="si">}</span><span class="s2">) &quot;</span>
                             <span class="sa">f</span><span class="s2">&quot;exceeds the size of the population.&quot;</span><span class="p">)</span>
        <span class="c1"># _end_if_</span>

        <span class="c1"># Sanity check.</span>
        <span class="k">if</span> <span class="n">migrate_op</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="vm">__class__</span><span class="o">.</span><span class="vm">__name__</span><span class="si">}</span><span class="s2">: Migration operator is missing.&quot;</span><span class="p">)</span>
        <span class="c1"># _end_if_</span>

        <span class="c1"># Assign the number of islands.</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_num_islands</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="nb">int</span><span class="p">(</span><span class="n">num_islands</span><span class="p">))</span>

        <span class="c1"># Get Migration Operator.</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_migrate_op</span> <span class="o">=</span> <span class="n">migrate_op</span>
    <span class="c1"># _end_def_</span>

    <span class="nd">@property</span>
    <span class="k">def</span><span class="w"> </span><span class="nf">num_islands</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">int</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Accessor method that returns the</span>
<span class="sd">        number of islands in the model.</span>

<span class="sd">        :return: the number of islands.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_num_islands</span>
    <span class="c1"># _end_def_</span>

    <span class="nd">@property</span>
    <span class="k">def</span><span class="w"> </span><span class="nf">migrate_op</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">MigrationOperator</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Accessor method that returns the migration operator reference.</span>

<span class="sd">        :return: the MigrationOperator.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_migrate_op</span>
    <span class="c1"># _end_def_</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">_evolve_population</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">island</span><span class="p">:</span> <span class="n">SubPopulation</span><span class="p">,</span> <span class="n">epochs</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span> <span class="n">shuffle</span><span class="p">:</span> <span class="nb">bool</span><span class="p">,</span>
                           <span class="n">correction</span><span class="p">:</span> <span class="nb">bool</span><span class="p">,</span> <span class="n">elitism</span><span class="p">:</span> <span class="nb">bool</span><span class="p">,</span> <span class="n">f_tol</span><span class="p">:</span> <span class="nb">float</span><span class="p">,</span> <span class="n">adapt_probs</span><span class="p">:</span> <span class="nb">bool</span><span class="p">,</span>
                           <span class="n">prob_crossx</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">prob_mutate</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="kc">None</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">tuple</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        This is a helper method to be used inside the Parallel delayed method.</span>
<span class="sd">        It is responsible for running the evolution of a single population (island).</span>

<span class="sd">        :return: a tuple (island, has_converged, local_stats, elapsed_time)</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># Get the BitGenerator used by default_rng.</span>
        <span class="n">bit_gen</span> <span class="o">=</span> <span class="nb">type</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">rng_GA</span><span class="o">.</span><span class="n">bit_generator</span><span class="p">)</span>

        <span class="c1"># Use the state from a fresh bit generator to re-seed rng_GA.</span>
        <span class="c1"># This uses the current system time (in nanoseconds) to avoid</span>
        <span class="c1"># using the same seed value among different Parallel workers.</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">rng_GA</span><span class="o">.</span><span class="n">bit_generator</span><span class="o">.</span><span class="n">state</span> <span class="o">=</span> <span class="n">bit_gen</span><span class="p">(</span><span class="n">seed</span><span class="o">=</span><span class="n">time</span><span class="o">.</span><span class="n">time_ns</span><span class="p">())</span><span class="o">.</span><span class="n">state</span>

        <span class="c1"># Keeps track of the convergence /termination of the</span>
        <span class="c1"># population, along with the iteration that happened.</span>
        <span class="n">has_converged</span> <span class="o">=</span> <span class="p">(</span><span class="kc">False</span><span class="p">,</span> <span class="n">epochs</span><span class="p">)</span>

        <span class="c1"># Get the size of the population.</span>
        <span class="n">pop_size</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">island</span><span class="o">.</span><span class="n">population</span><span class="p">)</span>

        <span class="c1"># Define local dictionary to hold the statistics.</span>
        <span class="n">local_stats</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;avg&quot;</span><span class="p">:</span> <span class="p">[],</span> <span class="s2">&quot;std&quot;</span><span class="p">:</span> <span class="p">[],</span> <span class="s2">&quot;prob_crossx&quot;</span><span class="p">:</span> <span class="p">[],</span> <span class="s2">&quot;prob_mutate&quot;</span><span class="p">:</span> <span class="p">[]}</span>

        <span class="c1"># Initialize this auxiliary parameter to a large number.</span>
        <span class="n">avg_fitness_0</span> <span class="o">=</span> <span class="mf">1.0e+100</span>

        <span class="c1"># Check if initial probabilities have been given.</span>
        <span class="k">if</span> <span class="n">prob_crossx</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">prob_mutate</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_crossx_op</span><span class="o">.</span><span class="n">probability</span> <span class="o">=</span> <span class="n">prob_crossx</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_mutate_op</span><span class="o">.</span><span class="n">probability</span> <span class="o">=</span> <span class="n">prob_mutate</span>
        <span class="c1"># _end_if_</span>

        <span class="c1"># Start timing the loop.</span>
        <span class="n">time_t0</span> <span class="o">=</span> <span class="n">time</span><span class="o">.</span><span class="n">perf_counter</span><span class="p">()</span>

        <span class="c1"># Repeat &#39;epoch&#39; times.</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">epochs</span><span class="p">):</span>

            <span class="c1"># Update current iteration.</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">iteration</span> <span class="o">=</span> <span class="n">i</span>

            <span class="c1"># SELECT the parents.</span>
            <span class="n">population_i</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">select_op</span><span class="p">(</span><span class="n">island</span><span class="o">.</span><span class="n">population</span><span class="p">)</span>

            <span class="c1"># Shuffle the selected parents.</span>
            <span class="k">if</span> <span class="n">shuffle</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">rng_GA</span><span class="o">.</span><span class="n">shuffle</span><span class="p">(</span><span class="n">population_i</span><span class="p">)</span>
            <span class="c1"># _end_if_</span>

            <span class="c1"># CROSSOVER/MUTATE to produce offsprings.</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">crossover_mutate</span><span class="p">(</span><span class="n">population_i</span><span class="p">)</span>

            <span class="c1"># EVALUATE the i-th population.</span>
            <span class="n">fit_list_i</span><span class="p">,</span> <span class="n">found_solution</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">evaluate_fitness</span><span class="p">(</span><span class="n">population_i</span><span class="p">)</span>

            <span class="c1"># Check if &#39;corrections&#39; are enabled.</span>
            <span class="k">if</span> <span class="n">correction</span><span class="p">:</span>
                <span class="c1"># Apply the function.</span>
                <span class="n">total_corrections</span><span class="p">,</span> <span class="n">_</span> <span class="o">=</span> <span class="n">apply_corrections</span><span class="p">(</span><span class="n">population_i</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">fitness_func</span><span class="p">)</span>

                <span class="c1"># If corrections were made we will</span>
                <span class="c1"># need to update the fitness list.</span>
                <span class="k">if</span> <span class="n">total_corrections</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>

                    <span class="c1"># Update the fitness list to ensure consistency.</span>
                    <span class="n">fit_list_i</span> <span class="o">=</span> <span class="p">[</span><span class="n">p</span><span class="o">.</span><span class="n">fitness</span> <span class="k">for</span> <span class="n">p</span> <span class="ow">in</span> <span class="n">population_i</span><span class="p">]</span>
            <span class="c1"># _end_if_</span>

            <span class="c1"># Check if &#39;elitism&#39; is enabled.</span>
            <span class="k">if</span> <span class="n">elitism</span><span class="p">:</span>
                <span class="c1"># Find the individual chromosome with the highest fitness</span>
                <span class="c1"># value (from the old subpopulation of the current island).</span>
                <span class="n">best_chromosome</span> <span class="o">=</span> <span class="nb">max</span><span class="p">([</span><span class="n">p</span> <span class="k">for</span> <span class="n">p</span> <span class="ow">in</span> <span class="n">island</span><span class="o">.</span><span class="n">population</span> <span class="k">if</span> <span class="ow">not</span> <span class="n">isnan</span><span class="p">(</span><span class="n">p</span><span class="o">.</span><span class="n">fitness</span><span class="p">)],</span>
                                      <span class="n">key</span><span class="o">=</span><span class="n">attrgetter</span><span class="p">(</span><span class="s2">&quot;fitness&quot;</span><span class="p">),</span> <span class="n">default</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span>

                <span class="c1"># Check if the chromosome already exists.</span>
                <span class="k">if</span> <span class="n">best_chromosome</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">population_i</span><span class="p">:</span>
                    <span class="c1"># Select a position at random.</span>
                    <span class="n">locus</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">rng_GA</span><span class="o">.</span><span class="n">integers</span><span class="p">(</span><span class="n">pop_size</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="nb">int</span><span class="p">)</span>

                    <span class="c1"># Replace it with the previous best.</span>
                    <span class="n">population_i</span><span class="p">[</span><span class="n">locus</span><span class="p">]</span> <span class="o">=</span> <span class="n">best_chromosome</span>

                    <span class="c1"># Update the list of fitness values to reflect the update.</span>
                    <span class="n">fit_list_i</span><span class="p">[</span><span class="n">locus</span><span class="p">]</span> <span class="o">=</span> <span class="n">population_i</span><span class="p">[</span><span class="n">locus</span><span class="p">]</span><span class="o">.</span><span class="n">fitness</span>
                <span class="c1"># _end_if_</span>
            <span class="c1"># _end_if_</span>

            <span class="c1"># Compute the mean value.</span>
            <span class="n">avg_fitness_i</span> <span class="o">=</span> <span class="n">nanmean</span><span class="p">(</span><span class="n">fit_list_i</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="nb">float</span><span class="p">)</span>

            <span class="c1"># Compute the standard deviation value.</span>
            <span class="n">std_fitness_i</span> <span class="o">=</span> <span class="n">nanstd</span><span class="p">(</span><span class="n">fit_list_i</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="nb">float</span><span class="p">)</span>

            <span class="c1"># Update the i-th population mean/std.</span>
            <span class="k">if</span> <span class="nb">all</span><span class="p">(</span><span class="n">isfinite</span><span class="p">([</span><span class="n">avg_fitness_i</span><span class="p">,</span> <span class="n">std_fitness_i</span><span class="p">])):</span>

                <span class="c1"># Store them in the dictionary.</span>
                <span class="n">local_stats</span><span class="p">[</span><span class="s2">&quot;avg&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">avg_fitness_i</span><span class="p">)</span>
                <span class="n">local_stats</span><span class="p">[</span><span class="s2">&quot;std&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">std_fitness_i</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">RuntimeError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">i</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="mi">1</span><span class="si">}</span><span class="s2">: Mean=</span><span class="si">{</span><span class="n">avg_fitness_i</span><span class="si">:</span><span class="s2">.5f</span><span class="si">}</span><span class="s2">, Std=</span><span class="si">{</span><span class="n">std_fitness_i</span><span class="si">:</span><span class="s2">.5f</span><span class="si">}</span><span class="s2">.&quot;</span><span class="p">)</span>
            <span class="c1"># _end_if_</span>

            <span class="c1"># Update the old population with the new chromosomes.</span>
            <span class="n">island</span><span class="o">.</span><span class="n">population</span> <span class="o">=</span> <span class="n">population_i</span>

            <span class="c1"># Check for termination.</span>
            <span class="k">if</span> <span class="n">found_solution</span><span class="p">:</span>
                <span class="c1"># Switch the convergence flag and track the current iteration.</span>
                <span class="n">has_converged</span> <span class="o">=</span> <span class="p">(</span><span class="kc">True</span><span class="p">,</span> <span class="n">i</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span>

                <span class="c1"># Exit from the loop.</span>
                <span class="k">break</span>
            <span class="c1"># _end_if_</span>

            <span class="c1"># Check for convergence.</span>
            <span class="k">if</span> <span class="n">f_tol</span> <span class="ow">and</span> <span class="n">isclose</span><span class="p">(</span><span class="n">avg_fitness_i</span><span class="p">,</span> <span class="n">avg_fitness_0</span><span class="p">,</span> <span class="n">abs_tol</span><span class="o">=</span><span class="n">f_tol</span><span class="p">):</span>
                <span class="c1"># Switch the convergence flag and track the current iteration.</span>
                <span class="n">has_converged</span> <span class="o">=</span> <span class="p">(</span><span class="kc">True</span><span class="p">,</span> <span class="n">i</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span>

                <span class="c1"># Exit from the loop.</span>
                <span class="k">break</span>
            <span class="c1"># _end_if_</span>

            <span class="c1"># Check the adaptive flag.</span>
            <span class="k">if</span> <span class="n">adapt_probs</span><span class="p">:</span>
                <span class="c1"># Compute the current average Hamming distance.</span>
                <span class="n">avg_distance</span> <span class="o">=</span> <span class="n">average_hamming_distance</span><span class="p">(</span><span class="n">population_i</span><span class="p">)</span>

                <span class="c1"># Update the genetic probabilities.</span>
                <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">adapt_probabilities</span><span class="p">(</span><span class="n">threshold</span><span class="o">=</span><span class="n">avg_distance</span><span class="p">):</span>
                    <span class="c1"># Store the updated crossover and mutation probabilities.</span>
                    <span class="n">local_stats</span><span class="p">[</span><span class="s2">&quot;prob_crossx&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_crossx_op</span><span class="o">.</span><span class="n">probability</span><span class="p">)</span>
                    <span class="n">local_stats</span><span class="p">[</span><span class="s2">&quot;prob_mutate&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_mutate_op</span><span class="o">.</span><span class="n">probability</span><span class="p">)</span>
            <span class="c1"># _end_if_</span>

            <span class="c1"># Update the average value for the next iteration.</span>
            <span class="n">avg_fitness_0</span> <span class="o">=</span> <span class="n">avg_fitness_i</span>
        <span class="c1"># _end_for_</span>

        <span class="c1"># Compute the elapsed time (in seconds).</span>
        <span class="n">elapsed_time</span> <span class="o">=</span> <span class="n">time</span><span class="o">.</span><span class="n">perf_counter</span><span class="p">()</span> <span class="o">-</span> <span class="n">time_t0</span>

        <span class="k">return</span> <span class="n">island</span><span class="p">,</span> <span class="n">has_converged</span><span class="p">,</span> <span class="n">local_stats</span><span class="p">,</span> <span class="n">elapsed_time</span>
    <span class="c1"># _end_def_</span>

<div class="viewcode-block" id="IslandModelGA.run">
<a class="viewcode-back" href="../../../pygenalgo.engines.html#pygenalgo.engines.island_model_ga.IslandModelGA.run">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">run</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">epochs</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">500</span><span class="p">,</span> <span class="n">correction</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span> <span class="n">elitism</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
            <span class="n">f_tol</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">allow_migration</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span> <span class="n">n_periods</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">10</span><span class="p">,</span>
            <span class="n">adapt_probs</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span> <span class="n">shuffle</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span> <span class="n">f_max_eval</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
            <span class="n">verbose</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Main method of the IslandModelGA class, that implements the evolutionary routine.</span>

<span class="sd">        :param epochs: (int) maximum number of iterations in the evolution process.</span>

<span class="sd">        :param correction: (bool) flag that if set to &#39;True&#39; will check the validity of</span>
<span class="sd">                           the population (at the gene level) and attempt to correct the</span>
<span class="sd">                           genome by calling the random() method of the flawed gene.</span>

<span class="sd">        :param elitism: (bool) flag that enables elitism. If set to True then the chromosome</span>
<span class="sd">                        with the highest fitness will always be copied to the next generation</span>
<span class="sd">                        (unaltered).</span>

<span class="sd">        :param f_tol: (float) tolerance in the difference between the average values of two</span>
<span class="sd">                      consecutive populations. It is used to determine the convergence of the</span>
<span class="sd">                      population. If this value is None (default) the algorithm will terminate</span>
<span class="sd">                      using the epochs value.</span>

<span class="sd">        :param allow_migration: (bool) flag that if set to &#39;True&#39; will allow the migration</span>
<span class="sd">                                of the best individuals among the different islands.</span>

<span class="sd">        :param n_periods: (int) the number of times that we will break the main evolution</span>
<span class="sd">                          to allow for chromosomes to migrate. NB: This setting is active</span>
<span class="sd">                          only when the option allow_migration == True. Otherwise, is ignored.</span>

<span class="sd">        :param adapt_probs: (bool) If enabled (set to True), it will allow the crossover and</span>
<span class="sd">                            mutation probabilities to adapt according to the convergence of</span>
<span class="sd">                            the population to a single solution. Default is set to False.</span>

<span class="sd">        :param shuffle: (bool) If enabled (set to True), it will shuffle the population before</span>
<span class="sd">                        the application of the crossover and mutation operations. The default</span>
<span class="sd">                        is set to True.</span>

<span class="sd">        :param f_max_eval: (int) it sets an upper limit of function evaluations. If the number</span>
<span class="sd">                           is exceeded the genetic algorithm stops. If this value is set, the</span>
<span class="sd">                           epochs will be ignored and re-adjusted to meet the new requirement.</span>

<span class="sd">        :param verbose: (bool) if &#39;True&#39; it will display periodically information about the</span>
<span class="sd">                        current stats of the subpopulations. NB: This setting is active only</span>
<span class="sd">                        when the option allow_migration == True. Otherwise, is ignored.</span>
<span class="sd">        :return: None.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># Reset stats dictionary.</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_stats</span><span class="o">.</span><span class="n">clear</span><span class="p">()</span>

        <span class="c1"># Randomly shuffle (in place) the original population.</span>
        <span class="k">if</span> <span class="n">shuffle</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">rng_GA</span><span class="o">.</span><span class="n">shuffle</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">population</span><span class="p">)</span>
        <span class="c1"># _end_if_</span>

        <span class="c1"># Initial random split of the total population in (active) subpopulations.</span>
        <span class="c1"># Active here means &#39;still evolving&#39;.</span>
        <span class="n">active_population</span> <span class="o">=</span> <span class="p">[</span><span class="n">SubPopulation</span><span class="p">(</span><span class="n">i</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">population</span><span class="p">[</span><span class="n">i</span><span class="p">::</span><span class="bp">self</span><span class="o">.</span><span class="n">_num_islands</span><span class="p">])</span>
                             <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_num_islands</span><span class="p">)]</span>

        <span class="c1"># Initial evaluation of the subpopulations.</span>
        <span class="k">for</span> <span class="n">pop_n</span> <span class="ow">in</span> <span class="n">active_population</span><span class="p">:</span>

            <span class="c1"># Initialize the statistics dictionary.</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_stats</span><span class="p">[</span><span class="n">pop_n</span><span class="o">.</span><span class="n">id</span><span class="p">]</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;avg&quot;</span><span class="p">:</span> <span class="p">[],</span> <span class="s2">&quot;std&quot;</span><span class="p">:</span> <span class="p">[],</span> <span class="s2">&quot;prob_crossx&quot;</span><span class="p">:</span> <span class="p">[],</span> <span class="s2">&quot;prob_mutate&quot;</span><span class="p">:</span> <span class="p">[]}</span>

            <span class="c1"># Initial evaluation of the population. This can run also in parallel.</span>
            <span class="n">fit_list_0</span><span class="p">,</span> <span class="n">_</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">evaluate_fitness</span><span class="p">(</span><span class="n">pop_n</span><span class="o">.</span><span class="n">population</span><span class="p">,</span> <span class="n">parallel_mode</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                                                  <span class="n">backend</span><span class="o">=</span><span class="s2">&quot;loky&quot;</span><span class="p">)</span>
            <span class="c1"># Compute the mean value.</span>
            <span class="n">avg_fitness_0</span> <span class="o">=</span> <span class="n">nanmean</span><span class="p">(</span><span class="n">fit_list_0</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="nb">float</span><span class="p">)</span>

            <span class="c1"># Compute the standard deviation value.</span>
            <span class="n">std_fitness_0</span> <span class="o">=</span> <span class="n">nanstd</span><span class="p">(</span><span class="n">fit_list_0</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="nb">float</span><span class="p">)</span>

            <span class="c1"># Check for initial errors.</span>
            <span class="k">if</span> <span class="nb">all</span><span class="p">(</span><span class="n">isfinite</span><span class="p">([</span><span class="n">avg_fitness_0</span><span class="p">,</span> <span class="n">std_fitness_0</span><span class="p">])):</span>

                <span class="c1"># Store them in the dictionary.</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">_stats</span><span class="p">[</span><span class="n">pop_n</span><span class="o">.</span><span class="n">id</span><span class="p">][</span><span class="s2">&quot;avg&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">avg_fitness_0</span><span class="p">)</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">_stats</span><span class="p">[</span><span class="n">pop_n</span><span class="o">.</span><span class="n">id</span><span class="p">][</span><span class="s2">&quot;std&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">std_fitness_0</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>

                <span class="k">raise</span> <span class="ne">RuntimeError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">pop_n</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2">: Mean=</span><span class="si">{</span><span class="n">avg_fitness_0</span><span class="si">:</span><span class="s2">.5f</span><span class="si">}</span><span class="s2">, &quot;</span>
                                   <span class="sa">f</span><span class="s2">&quot;Std=</span><span class="si">{</span><span class="n">std_fitness_0</span><span class="si">:</span><span class="s2">.5f</span><span class="si">}</span><span class="s2">.&quot;</span><span class="p">)</span>
            <span class="c1"># _end_if_</span>

            <span class="c1"># Store the initial crossover and mutation probabilities.</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_stats</span><span class="p">[</span><span class="n">pop_n</span><span class="o">.</span><span class="n">id</span><span class="p">][</span><span class="s2">&quot;prob_crossx&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_crossx_op</span><span class="o">.</span><span class="n">probability</span><span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_stats</span><span class="p">[</span><span class="n">pop_n</span><span class="o">.</span><span class="n">id</span><span class="p">][</span><span class="s2">&quot;prob_mutate&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_mutate_op</span><span class="o">.</span><span class="n">probability</span><span class="p">)</span>
        <span class="c1"># _end_for_</span>

        <span class="c1"># Check if we have set a maximum number on function</span>
        <span class="c1"># evaluations and re-adjust the epochs.</span>
        <span class="k">if</span> <span class="n">f_max_eval</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>

            <span class="c1"># First remove the counts from the initial</span>
            <span class="c1"># evaluation of the population.</span>
            <span class="n">total_f_counts</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">f_max_eval</span><span class="p">)</span> <span class="o">-</span> <span class="bp">self</span><span class="o">.</span><span class="n">f_evals</span>

            <span class="c1"># Assuming each epoch performs N function evaluations.</span>
            <span class="n">epochs</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">total_f_counts</span> <span class="o">/</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">population</span><span class="p">))</span>

            <span class="c1"># Display a warning message.</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;The &#39;f_max_eval&#39; parameter has been set to: </span><span class="si">{</span><span class="n">f_max_eval</span><span class="si">}</span><span class="s2">. &quot;</span>
                           <span class="sa">f</span><span class="s2">&quot;The &#39;epochs&#39; value has been re-adjusted to: </span><span class="si">{</span><span class="n">epochs</span><span class="si">}</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="c1"># _end_if_</span>

        <span class="c1"># Display an information message.</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Parallel evolution in progress with </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">_num_islands</span><span class="si">}</span><span class="s2"> islands ...&quot;</span><span class="p">)</span>

        <span class="c1"># Final population.</span>
        <span class="n">final_population</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="c1"># Initial time instant.</span>
        <span class="n">time_t0</span> <span class="o">=</span> <span class="n">time</span><span class="o">.</span><span class="n">perf_counter</span><span class="p">()</span>

        <span class="c1"># Check if we allow migration among the populations.</span>
        <span class="k">if</span> <span class="n">allow_migration</span><span class="p">:</span>

            <span class="c1"># Initial values for the crossover and mutation operators will be used</span>
            <span class="c1"># to ensure continuity in the case of adaptable probabilities.</span>
            <span class="n">genetic_probs</span> <span class="o">=</span> <span class="n">defaultdict</span><span class="p">(</span><span class="nb">dict</span><span class="p">)</span>

            <span class="c1"># Initial assignment of the genetic probabilities.</span>
            <span class="k">for</span> <span class="n">pop_n</span> <span class="ow">in</span> <span class="n">active_population</span><span class="p">:</span>

                <span class="c1"># Use the values of the object operators itself.</span>
                <span class="n">genetic_probs</span><span class="p">[</span><span class="n">pop_n</span><span class="o">.</span><span class="n">id</span><span class="p">][</span><span class="s2">&quot;crossx&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_crossx_op</span><span class="o">.</span><span class="n">probability</span>
                <span class="n">genetic_probs</span><span class="p">[</span><span class="n">pop_n</span><span class="o">.</span><span class="n">id</span><span class="p">][</span><span class="s2">&quot;mutate&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_mutate_op</span><span class="o">.</span><span class="n">probability</span>
            <span class="c1"># _end_for_</span>

            <span class="c1"># Make sure &#39;n_periods&#39; is integer.</span>
            <span class="n">n_periods</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">n_periods</span><span class="p">)</span>

            <span class="c1"># Compute the in-between evolving epochs.</span>
            <span class="n">n_epochs</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="nb">float</span><span class="p">(</span><span class="n">epochs</span><span class="p">)</span><span class="o">/</span><span class="n">n_periods</span><span class="p">)</span>

            <span class="c1"># Compute the remainder epochs (if any).</span>
            <span class="n">rem_epochs</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">epochs</span> <span class="o">%</span> <span class="n">n_periods</span><span class="p">)</span>

            <span class="c1"># Reuse the pool of workers.</span>
            <span class="k">with</span> <span class="n">Parallel</span><span class="p">(</span><span class="n">n_jobs</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">n_cpus</span><span class="p">,</span> <span class="n">backend</span><span class="o">=</span><span class="s2">&quot;loky&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">parallel</span><span class="p">:</span>

                <span class="c1"># Break the total &#39;epochs&#39; in n_periods.</span>
                <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">n_periods</span><span class="p">):</span>

                    <span class="c1"># Check if we want information on to be logged.</span>
                    <span class="k">if</span> <span class="n">verbose</span><span class="p">:</span>
                        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Current period </span><span class="si">{</span><span class="n">i</span><span class="o">+</span><span class="mi">1</span><span class="si">}</span><span class="s2"> / </span><span class="si">{</span><span class="n">n_periods</span><span class="si">}</span><span class="s2">:&quot;</span><span class="p">)</span>
                    <span class="c1"># _end_if_</span>

                    <span class="c1"># If the remainder epochs is not zero, add them in the</span>
                    <span class="c1"># last iteration to complete the total number of epochs.</span>
                    <span class="k">if</span> <span class="n">rem_epochs</span> <span class="ow">and</span> <span class="n">i</span> <span class="o">==</span> <span class="n">n_periods</span><span class="o">-</span><span class="mi">1</span><span class="p">:</span>

                        <span class="c1"># Update the n_epochs ONLY in the last period.</span>
                        <span class="n">n_epochs</span> <span class="o">+=</span> <span class="n">rem_epochs</span>
                    <span class="c1"># _end_if_</span>

                    <span class="c1"># Evolve the subpopulations in parallel for &#39;n_epochs&#39;.</span>
                    <span class="n">results_i</span> <span class="o">=</span> <span class="n">parallel</span><span class="p">(</span>
                        <span class="n">delayed</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_evolve_population</span><span class="p">)(</span><span class="n">island</span><span class="o">=</span><span class="n">pop_i</span><span class="p">,</span>
                                                         <span class="n">epochs</span><span class="o">=</span><span class="n">n_epochs</span><span class="p">,</span>
                                                         <span class="n">shuffle</span><span class="o">=</span><span class="n">shuffle</span><span class="p">,</span>
                                                         <span class="n">elitism</span><span class="o">=</span><span class="n">elitism</span><span class="p">,</span>
                                                         <span class="n">correction</span><span class="o">=</span><span class="n">correction</span><span class="p">,</span>
                                                         <span class="n">f_tol</span><span class="o">=</span><span class="n">f_tol</span><span class="p">,</span>
                                                         <span class="n">adapt_probs</span><span class="o">=</span><span class="n">adapt_probs</span><span class="p">,</span>
                                                         <span class="n">prob_crossx</span><span class="o">=</span><span class="n">genetic_probs</span><span class="p">[</span><span class="n">pop_i</span><span class="o">.</span><span class="n">id</span><span class="p">][</span><span class="s2">&quot;crossx&quot;</span><span class="p">],</span>
                                                         <span class="n">prob_mutate</span><span class="o">=</span><span class="n">genetic_probs</span><span class="p">[</span><span class="n">pop_i</span><span class="o">.</span><span class="n">id</span><span class="p">][</span><span class="s2">&quot;mutate&quot;</span><span class="p">])</span>
                        <span class="k">for</span> <span class="n">pop_i</span> <span class="ow">in</span> <span class="n">active_population</span>
                    <span class="p">)</span>

                    <span class="c1"># Empty the list of active populations.</span>
                    <span class="n">active_population</span> <span class="o">=</span> <span class="p">[]</span>

                    <span class="c1"># Process the results if the i-th period.</span>
                    <span class="k">for</span> <span class="n">res</span> <span class="ow">in</span> <span class="n">results_i</span><span class="p">:</span>

                        <span class="c1"># Extract the results.</span>
                        <span class="n">island</span><span class="p">,</span> <span class="n">has_converged</span><span class="p">,</span> <span class="n">local_stats</span><span class="p">,</span> <span class="n">_</span> <span class="o">=</span> <span class="n">res</span>

                        <span class="c1"># Check if we want information on the screen.</span>
                        <span class="k">if</span> <span class="n">verbose</span><span class="p">:</span>

                            <span class="c1"># Find the current highest fitness.</span>
                            <span class="n">best_fitness</span> <span class="o">=</span> <span class="nb">max</span><span class="p">((</span><span class="n">p</span><span class="o">.</span><span class="n">fitness</span> <span class="k">for</span> <span class="n">p</span> <span class="ow">in</span> <span class="n">island</span><span class="o">.</span><span class="n">population</span>
                                                <span class="k">if</span> <span class="ow">not</span> <span class="n">isnan</span><span class="p">(</span><span class="n">p</span><span class="o">.</span><span class="n">fitness</span><span class="p">)))</span>

                            <span class="c1"># Log an update of the progress.</span>
                            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Best Fitness in island </span><span class="si">{</span><span class="n">island</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2"> &quot;</span>
                                        <span class="sa">f</span><span class="s2">&quot;is:= </span><span class="si">{</span><span class="n">best_fitness</span><span class="si">:</span><span class="s2">.5f</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
                        <span class="c1"># _end_if_</span>

                        <span class="c1"># First check if the island has converged.</span>
                        <span class="k">if</span> <span class="n">has_converged</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span>
                            <span class="c1"># Copy the population in the final list.</span>
                            <span class="n">final_population</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span><span class="n">island</span><span class="o">.</span><span class="n">population</span><span class="p">)</span>

                            <span class="c1"># Check for verbosity.</span>
                            <span class="k">if</span> <span class="n">verbose</span><span class="p">:</span>
                                <span class="c1"># Compute the total number of iterations.</span>
                                <span class="n">itr</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">i</span><span class="o">*</span><span class="n">n_epochs</span> <span class="o">+</span> <span class="n">has_converged</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>

                                <span class="c1"># Log a warning message to the screen.</span>
                                <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Island population </span><span class="si">{</span><span class="n">island</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2"> &quot;</span>
                                               <span class="sa">f</span><span class="s2">&quot;finished in </span><span class="si">{</span><span class="n">itr</span><span class="si">}</span><span class="s2"> iterations. &quot;</span><span class="p">)</span>
                            <span class="c1"># _end_if_</span>
                        <span class="k">else</span><span class="p">:</span>
                            <span class="c1"># Add the island population to the new active list.</span>
                            <span class="n">active_population</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">island</span><span class="p">)</span>
                        <span class="c1"># _end_if_</span>

                        <span class="c1"># Update statistics.</span>
                        <span class="bp">self</span><span class="o">.</span><span class="n">_stats</span><span class="p">[</span><span class="n">island</span><span class="o">.</span><span class="n">id</span><span class="p">][</span><span class="s2">&quot;avg&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span><span class="n">local_stats</span><span class="p">[</span><span class="s2">&quot;avg&quot;</span><span class="p">])</span>
                        <span class="bp">self</span><span class="o">.</span><span class="n">_stats</span><span class="p">[</span><span class="n">island</span><span class="o">.</span><span class="n">id</span><span class="p">][</span><span class="s2">&quot;std&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span><span class="n">local_stats</span><span class="p">[</span><span class="s2">&quot;std&quot;</span><span class="p">])</span>

                        <span class="c1"># Check if we were adapting the probabilities.</span>
                        <span class="k">if</span> <span class="n">adapt_probs</span><span class="p">:</span>

                            <span class="c1"># Make sure there is at least one entry</span>
                            <span class="c1"># to avoid &quot;index out of bound&quot; errors.</span>
                            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">local_stats</span><span class="p">[</span><span class="s2">&quot;prob_crossx&quot;</span><span class="p">])</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>

                                <span class="c1"># Update the values for the next interval.</span>
                                <span class="n">genetic_probs</span><span class="p">[</span><span class="n">island</span><span class="o">.</span><span class="n">id</span><span class="p">][</span><span class="s2">&quot;crossx&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">local_stats</span><span class="p">[</span><span class="s2">&quot;prob_crossx&quot;</span><span class="p">][</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
                                <span class="n">genetic_probs</span><span class="p">[</span><span class="n">island</span><span class="o">.</span><span class="n">id</span><span class="p">][</span><span class="s2">&quot;mutate&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">local_stats</span><span class="p">[</span><span class="s2">&quot;prob_mutate&quot;</span><span class="p">][</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
                            <span class="c1"># _end_if_</span>

                            <span class="c1"># Store the updated crossover and mutation values.</span>
                            <span class="bp">self</span><span class="o">.</span><span class="n">_stats</span><span class="p">[</span><span class="n">island</span><span class="o">.</span><span class="n">id</span><span class="p">][</span><span class="s2">&quot;prob_crossx&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span><span class="n">local_stats</span><span class="p">[</span><span class="s2">&quot;prob_crossx&quot;</span><span class="p">])</span>
                            <span class="bp">self</span><span class="o">.</span><span class="n">_stats</span><span class="p">[</span><span class="n">island</span><span class="o">.</span><span class="n">id</span><span class="p">][</span><span class="s2">&quot;prob_mutate&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span><span class="n">local_stats</span><span class="p">[</span><span class="s2">&quot;prob_mutate&quot;</span><span class="p">])</span>
                        <span class="c1"># _end_if_</span>

                    <span class="c1"># _end_for_</span>

                    <span class="c1"># Check for early termination.</span>
                    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">active_population</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
                        <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="s2">&quot;No active islands found.&quot;</span><span class="p">)</span>
                        <span class="k">break</span>
                    <span class="c1"># _end_if_</span>

                    <span class="c1"># Here we call the migration policy.</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">_migrate_op</span><span class="p">(</span><span class="n">active_population</span><span class="p">)</span>
                <span class="c1"># _end_for_</span>

            <span class="c1"># _end_parallel_with_</span>

            <span class="c1"># Get the rest of the populations that have not yet converged.</span>
            <span class="k">for</span> <span class="n">pop_n</span> <span class="ow">in</span> <span class="n">active_population</span><span class="p">:</span>
                <span class="n">final_population</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span><span class="n">pop_n</span><span class="o">.</span><span class="n">population</span><span class="p">)</span>
            <span class="c1"># _end_for_</span>

        <span class="k">else</span><span class="p">:</span>

            <span class="c1"># Evolve the subpopulations in parallel for &#39;epoch&#39; iterations.</span>
            <span class="n">results</span> <span class="o">=</span> <span class="n">Parallel</span><span class="p">(</span><span class="n">n_jobs</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">n_cpus</span><span class="p">,</span> <span class="n">backend</span><span class="o">=</span><span class="s2">&quot;loky&quot;</span><span class="p">)(</span>
                <span class="n">delayed</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_evolve_population</span><span class="p">)(</span><span class="n">island</span><span class="o">=</span><span class="n">pop_n</span><span class="p">,</span>
                                                 <span class="n">epochs</span><span class="o">=</span><span class="n">epochs</span><span class="p">,</span>
                                                 <span class="n">shuffle</span><span class="o">=</span><span class="n">shuffle</span><span class="p">,</span>
                                                 <span class="n">elitism</span><span class="o">=</span><span class="n">elitism</span><span class="p">,</span>
                                                 <span class="n">correction</span><span class="o">=</span><span class="n">correction</span><span class="p">,</span>
                                                 <span class="n">f_tol</span><span class="o">=</span><span class="n">f_tol</span><span class="p">,</span>
                                                 <span class="n">adapt_probs</span><span class="o">=</span><span class="n">adapt_probs</span><span class="p">)</span>
                <span class="k">for</span> <span class="n">pop_n</span> <span class="ow">in</span> <span class="n">active_population</span>
            <span class="p">)</span>

            <span class="c1"># Process the final results.</span>
            <span class="k">for</span> <span class="n">res_n</span> <span class="ow">in</span> <span class="n">results</span><span class="p">:</span>

                <span class="c1"># Extract the results.</span>
                <span class="n">island</span><span class="p">,</span> <span class="n">has_converged</span><span class="p">,</span> <span class="n">local_stats</span><span class="p">,</span> <span class="n">_</span> <span class="o">=</span> <span class="n">res_n</span>

                <span class="c1"># Check if we want to log output.</span>
                <span class="k">if</span> <span class="n">has_converged</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span>
                    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Island population </span><span class="si">{</span><span class="n">island</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2">, &quot;</span>
                                <span class="sa">f</span><span class="s2">&quot;finished in </span><span class="si">{</span><span class="n">has_converged</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="si">}</span><span class="s2"> iterations.&quot;</span><span class="p">)</span>

                <span class="c1"># Copy only the population.</span>
                <span class="n">final_population</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span><span class="n">island</span><span class="o">.</span><span class="n">population</span><span class="p">)</span>

                <span class="c1"># Update the statistics.</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">_stats</span><span class="p">[</span><span class="n">island</span><span class="o">.</span><span class="n">id</span><span class="p">][</span><span class="s2">&quot;avg&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span><span class="n">local_stats</span><span class="p">[</span><span class="s2">&quot;avg&quot;</span><span class="p">])</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">_stats</span><span class="p">[</span><span class="n">island</span><span class="o">.</span><span class="n">id</span><span class="p">][</span><span class="s2">&quot;std&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span><span class="n">local_stats</span><span class="p">[</span><span class="s2">&quot;std&quot;</span><span class="p">])</span>

                <span class="c1"># Check if we were adapting the probabilities.</span>
                <span class="k">if</span> <span class="n">adapt_probs</span><span class="p">:</span>
                    <span class="c1"># Store the updated crossover and mutation values.</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">_stats</span><span class="p">[</span><span class="n">island</span><span class="o">.</span><span class="n">id</span><span class="p">][</span><span class="s2">&quot;prob_crossx&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span><span class="n">local_stats</span><span class="p">[</span><span class="s2">&quot;prob_crossx&quot;</span><span class="p">])</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">_stats</span><span class="p">[</span><span class="n">island</span><span class="o">.</span><span class="n">id</span><span class="p">][</span><span class="s2">&quot;prob_mutate&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span><span class="n">local_stats</span><span class="p">[</span><span class="s2">&quot;prob_mutate&quot;</span><span class="p">])</span>
            <span class="c1"># _end_for_</span>

        <span class="c1"># _end_if_</span>

        <span class="c1"># Update the population in the class.</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">population</span> <span class="o">=</span> <span class="n">final_population</span>

        <span class="c1"># Make a final fitness evaluation (to ensure consistency).</span>
        <span class="n">fit_list_final</span><span class="p">,</span> <span class="n">_</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">evaluate_fitness</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">population</span><span class="p">,</span>
                                                  <span class="n">parallel_mode</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">backend</span><span class="o">=</span><span class="s2">&quot;loky&quot;</span><span class="p">)</span>
        <span class="c1"># Compute the mean value.</span>
        <span class="n">avg_fitness_final</span> <span class="o">=</span> <span class="n">nanmean</span><span class="p">(</span><span class="n">fit_list_final</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="nb">float</span><span class="p">)</span>

        <span class="c1"># Final time instant.</span>
        <span class="n">time_tf</span> <span class="o">=</span> <span class="n">time</span><span class="o">.</span><span class="n">perf_counter</span><span class="p">()</span>

        <span class="c1"># Print message.</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Final Avg. Fitness = </span><span class="si">{</span><span class="n">avg_fitness_final</span><span class="si">:</span><span class="s2">.4f</span><span class="si">}</span><span class="s2">.&quot;</span><span class="p">)</span>

        <span class="c1"># Print final duration in seconds.</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Elapsed time: </span><span class="si">{</span><span class="p">(</span><span class="n">time_tf</span><span class="w"> </span><span class="o">-</span><span class="w"> </span><span class="n">time_t0</span><span class="p">)</span><span class="si">:</span><span class="s2">.3f</span><span class="si">}</span><span class="s2"> seconds.&quot;</span><span class="p">)</span></div>

    <span class="c1"># _end_def_</span>

<div class="viewcode-block" id="IslandModelGA.print_migration_stats">
<a class="viewcode-back" href="../../../pygenalgo.engines.html#pygenalgo.engines.island_model_ga.IslandModelGA.print_migration_stats">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">print_migration_stats</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Print the migration operators stats.</span>

<span class="sd">        :return: None.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># First print the migration operator.</span>
        <span class="nb">print</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">migrate_op</span><span class="p">)</span>

        <span class="c1"># Check if we used the MetaMigration.</span>
        <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">migrate_op</span><span class="p">,</span> <span class="n">MetaMigration</span><span class="p">):</span>
            <span class="c1"># Call internally all operators.</span>
            <span class="k">for</span> <span class="n">op</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">migrate_op</span><span class="o">.</span><span class="n">items</span><span class="p">:</span>
                <span class="nb">print</span><span class="p">(</span><span class="n">op</span><span class="p">)</span></div>
</div>

    <span class="c1"># _end_def_</span>

<span class="c1"># _end_class_</span>
</pre></div>

           </div>
          </div>
          <footer>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2024, Michalis Vrettas, PhD.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>