<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="../../../">
<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>pygenalgo.engines.standard_ga &mdash; PyGenAlgo 0.1.0 documentation</title>
      <link rel="stylesheet" type="text/css" href="../../../_static/pygments.css?v=92fd9be5" />
      <link rel="stylesheet" type="text/css" href="../../../_static/css/theme.css?v=19f00094" />

  
  <!--[if lt IE 9]>
    <script src="../../../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script src="../../../_static/jquery.js?v=5d32c60e"></script>
        <script src="../../../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
        <script src="../../../_static/documentation_options.js?v=01f34227"></script>
        <script src="../../../_static/doctools.js?v=9a2dae69"></script>
        <script src="../../../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../../../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../../../index.html" class="icon icon-home">
            PyGenAlgo
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../modules.html">code</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../../index.html">PyGenAlgo</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="../../index.html">Module code</a></li>
      <li class="breadcrumb-item active">pygenalgo.engines.standard_ga</li>
      <li class="wy-breadcrumbs-aside">
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <h1>Source code for pygenalgo.engines.standard_ga</h1><div class="highlight"><pre>
<span></span><span class="kn">import</span> <span class="nn">time</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">from</span> <span class="nn">typing</span> <span class="kn">import</span> <span class="n">Callable</span>
<span class="kn">from</span> <span class="nn">collections</span> <span class="kn">import</span> <span class="n">defaultdict</span>

<span class="kn">from</span> <span class="nn">pygenalgo.genome.chromosome</span> <span class="kn">import</span> <span class="n">Chromosome</span>
<span class="kn">from</span> <span class="nn">joblib</span> <span class="kn">import</span> <span class="p">(</span><span class="n">Parallel</span><span class="p">,</span> <span class="n">delayed</span><span class="p">,</span> <span class="n">cpu_count</span><span class="p">)</span>
<span class="kn">from</span> <span class="nn">pygenalgo.engines.auxiliary</span> <span class="kn">import</span> <span class="n">apply_corrections</span>
<span class="kn">from</span> <span class="nn">pygenalgo.operators.mutation.mutate_operator</span> <span class="kn">import</span> <span class="n">MutationOperator</span>
<span class="kn">from</span> <span class="nn">pygenalgo.operators.selection.select_operator</span> <span class="kn">import</span> <span class="n">SelectionOperator</span>
<span class="kn">from</span> <span class="nn">pygenalgo.operators.crossover.crossover_operator</span> <span class="kn">import</span> <span class="n">CrossoverOperator</span>

<span class="kn">from</span> <span class="nn">pygenalgo.operators.mutation.super_mutator</span> <span class="kn">import</span> <span class="n">SuperMutator</span>
<span class="kn">from</span> <span class="nn">pygenalgo.operators.crossover.super_crossover</span> <span class="kn">import</span> <span class="n">SuperCrossover</span>

<span class="c1"># Public interface.</span>
<span class="n">__all__</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;StandardGA&quot;</span><span class="p">]</span>


<div class="viewcode-block" id="StandardGA">
<a class="viewcode-back" href="../../../pygenalgo.engines.html#pygenalgo.engines.standard_ga.StandardGA">[docs]</a>
<span class="k">class</span> <span class="nc">StandardGA</span><span class="p">(</span><span class="nb">object</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Description:</span>

<span class="sd">        Standard GA model that at each iteration replaces the whole population using</span>
<span class="sd">        the genetic operators (crossover and mutation).</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="c1"># Make a random number generator.</span>
    <span class="n">rng_GA</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">default_rng</span><span class="p">()</span>

    <span class="c1"># Get the maximum number of CPUs.</span>
    <span class="n">MAX_CPUs</span> <span class="o">=</span> <span class="n">cpu_count</span><span class="p">()</span>

    <span class="c1"># Object variables.</span>
    <span class="vm">__slots__</span> <span class="o">=</span> <span class="p">(</span><span class="s2">&quot;population&quot;</span><span class="p">,</span> <span class="s2">&quot;fitness_func&quot;</span><span class="p">,</span> <span class="s2">&quot;_select_op&quot;</span><span class="p">,</span> <span class="s2">&quot;_cross_op&quot;</span><span class="p">,</span> <span class="s2">&quot;_mutate_op&quot;</span><span class="p">,</span>
                 <span class="s2">&quot;_stats&quot;</span><span class="p">)</span>

    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">initial_pop</span><span class="p">:</span> <span class="nb">list</span><span class="p">[</span><span class="n">Chromosome</span><span class="p">],</span> <span class="n">fit_func</span><span class="p">:</span> <span class="n">Callable</span><span class="p">,</span> <span class="n">select_op</span><span class="p">:</span> <span class="n">SelectionOperator</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
                 <span class="n">mutate_op</span><span class="p">:</span> <span class="n">MutationOperator</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">cross_op</span><span class="p">:</span> <span class="n">CrossoverOperator</span> <span class="o">=</span> <span class="kc">None</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">object</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Default constructor of StandardGA object.</span>

<span class="sd">        :param initial_pop: list of the initial population of (randomized) chromosomes.</span>

<span class="sd">        :param fit_func: callable fitness function.</span>

<span class="sd">        :param select_op: selection operator (must inherit from class SelectionOperator).</span>

<span class="sd">        :param mutate_op: mutation operator (must inherit from class MutationOperator).</span>

<span class="sd">        :param cross_op: crossover operator (must inherit from class CrossoverOperator).</span>

<span class="sd">        :return: a new GA object.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="c1"># Copy the reference of the population.</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">population</span> <span class="o">=</span> <span class="n">initial_pop</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>

        <span class="c1"># Make sure the fitness function is indeed callable.</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">callable</span><span class="p">(</span><span class="n">fit_func</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="vm">__class__</span><span class="o">.</span><span class="vm">__name__</span><span class="si">}</span><span class="s2">: Fitness function is not callable.&quot;</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="c1"># Get the fitness function.</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">fitness_func</span> <span class="o">=</span> <span class="n">fit_func</span>
        <span class="c1"># _end_if_</span>

        <span class="c1"># Get Selection Operator.</span>
        <span class="k">if</span> <span class="n">select_op</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="vm">__class__</span><span class="o">.</span><span class="vm">__name__</span><span class="si">}</span><span class="s2">: Selection operator is missing.&quot;</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_select_op</span> <span class="o">=</span> <span class="n">select_op</span>
        <span class="c1"># _end_if_</span>

        <span class="c1"># Get Mutation Operator.</span>
        <span class="k">if</span> <span class="n">mutate_op</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="vm">__class__</span><span class="o">.</span><span class="vm">__name__</span><span class="si">}</span><span class="s2">: Mutation operator is missing.&quot;</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_mutate_op</span> <span class="o">=</span> <span class="n">mutate_op</span>
        <span class="c1"># _end_if_</span>

        <span class="c1"># Get Crossover Operator.</span>
        <span class="k">if</span> <span class="n">cross_op</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="vm">__class__</span><span class="o">.</span><span class="vm">__name__</span><span class="si">}</span><span class="s2">: Crossover operator is missing.&quot;</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_cross_op</span> <span class="o">=</span> <span class="n">cross_op</span>
        <span class="c1"># _end_if_</span>

        <span class="c1"># Dictionary with stats.</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_stats</span> <span class="o">=</span> <span class="n">defaultdict</span><span class="p">(</span><span class="nb">list</span><span class="p">)</span>

    <span class="c1"># _end_def_</span>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">stats</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">dict</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Accessor method that returns the &#39;stats&#39; dictionary.</span>

<span class="sd">        :return: the dictionary with the statistics from the run.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_stats</span>

    <span class="c1"># _end_def_</span>

<div class="viewcode-block" id="StandardGA.update_stats">
<a class="viewcode-back" href="../../../pygenalgo.engines.html#pygenalgo.engines.standard_ga.StandardGA.update_stats">[docs]</a>
    <span class="k">def</span> <span class="nf">update_stats</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">avg_fitness</span><span class="p">:</span> <span class="nb">float</span><span class="p">,</span> <span class="n">std_fitness</span><span class="p">:</span> <span class="nb">float</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Update the stats dictionary with the input mean/std values</span>
<span class="sd">        of the fitness.</span>

<span class="sd">        :param avg_fitness: (float) mean fitness value of the population.</span>

<span class="sd">        :param std_fitness: (float) std fitness value of the population.</span>

<span class="sd">        :return: None.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="c1"># Make sure the input values are finite.</span>
        <span class="k">if</span> <span class="nb">all</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">isfinite</span><span class="p">([</span><span class="n">avg_fitness</span><span class="p">,</span> <span class="n">std_fitness</span><span class="p">])):</span>

            <span class="c1"># Store them in the dictionary.</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_stats</span><span class="p">[</span><span class="s2">&quot;avg&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">avg_fitness</span><span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_stats</span><span class="p">[</span><span class="s2">&quot;std&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">std_fitness</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">RuntimeError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="vm">__class__</span><span class="o">.</span><span class="vm">__name__</span><span class="si">}</span><span class="s2">: Something went wrong with current &quot;</span>
                               <span class="sa">f</span><span class="s2">&quot;population. Mean=</span><span class="si">{</span><span class="n">avg_fitness</span><span class="si">:</span><span class="s2">.5f</span><span class="si">}</span><span class="s2">, Std=</span><span class="si">{</span><span class="n">std_fitness</span><span class="si">:</span><span class="s2">.5f</span><span class="si">}</span><span class="s2">.&quot;</span><span class="p">)</span></div>

        <span class="c1"># _end_if_</span>

    <span class="c1"># _end_def_</span>

<div class="viewcode-block" id="StandardGA.individual_fitness">
<a class="viewcode-back" href="../../../pygenalgo.engines.html#pygenalgo.engines.standard_ga.StandardGA.individual_fitness">[docs]</a>
    <span class="k">def</span> <span class="nf">individual_fitness</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">index</span><span class="p">:</span> <span class="nb">int</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">float</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Get the fitness value of an individual member of the population.</span>

<span class="sd">        :param index: Position of the individual in the population.</span>

<span class="sd">        :return: The fitness value (float).</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">population</span><span class="p">[</span><span class="n">index</span><span class="p">]</span><span class="o">.</span><span class="n">fitness</span></div>


    <span class="c1"># _end_def_</span>

<div class="viewcode-block" id="StandardGA.population_fitness">
<a class="viewcode-back" href="../../../pygenalgo.engines.html#pygenalgo.engines.standard_ga.StandardGA.population_fitness">[docs]</a>
    <span class="k">def</span> <span class="nf">population_fitness</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">list</span><span class="p">[</span><span class="nb">float</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Get the fitness of all the population.</span>

<span class="sd">        :return: A list with all the fitness values.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="p">[</span><span class="n">p</span><span class="o">.</span><span class="n">fitness</span> <span class="k">for</span> <span class="n">p</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">population</span><span class="p">]</span></div>

    <span class="c1"># _end_def_</span>

<div class="viewcode-block" id="StandardGA.evaluate_fitness">
<a class="viewcode-back" href="../../../pygenalgo.engines.html#pygenalgo.engines.standard_ga.StandardGA.evaluate_fitness">[docs]</a>
    <span class="k">def</span> <span class="nf">evaluate_fitness</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">input_population</span><span class="p">:</span> <span class="nb">list</span><span class="p">[</span><span class="n">Chromosome</span><span class="p">],</span> <span class="n">parallel</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Evaluate all the chromosomes of the input list with the custom fitness function.</span>

<span class="sd">        :param input_population: (list) The population of Chromosomes that we want to evaluate</span>
<span class="sd">        their fitness.</span>

<span class="sd">        :param parallel: (bool) Flag that enables parallel computation of the fitness function.</span>

<span class="sd">        :return: the mean and std of the fitness values.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="c1"># Get a local copy of the fitness function.</span>
        <span class="n">fit_func</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">fitness_func</span>

        <span class="c1"># Check the &#39;parallel&#39; flag.</span>
        <span class="k">if</span> <span class="n">parallel</span><span class="p">:</span>

            <span class="c1"># Evaluate the chromosomes in parallel mode.</span>
            <span class="n">fitness_i</span> <span class="o">=</span> <span class="n">Parallel</span><span class="p">(</span><span class="n">n_jobs</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">MAX_CPUs</span><span class="p">,</span>
                                 <span class="n">backend</span><span class="o">=</span><span class="s2">&quot;threading&quot;</span><span class="p">)(</span>
                <span class="n">delayed</span><span class="p">(</span><span class="n">fit_func</span><span class="p">)(</span><span class="n">p</span><span class="p">)</span> <span class="k">for</span> <span class="n">p</span> <span class="ow">in</span> <span class="n">input_population</span>
            <span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>

            <span class="c1"># Evaluate the chromosomes in serial mode.</span>
            <span class="n">fitness_i</span> <span class="o">=</span> <span class="p">[</span><span class="n">fit_func</span><span class="p">(</span><span class="n">p</span><span class="p">)</span> <span class="k">for</span> <span class="n">p</span> <span class="ow">in</span> <span class="n">input_population</span><span class="p">]</span>

        <span class="c1"># _end_if_</span>

        <span class="c1"># Attach the fitness to each chromosome.</span>
        <span class="k">for</span> <span class="n">p</span><span class="p">,</span> <span class="n">fit_eval</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">input_population</span><span class="p">,</span> <span class="n">fitness_i</span><span class="p">):</span>
            <span class="n">p</span><span class="o">.</span><span class="n">fitness</span> <span class="o">=</span> <span class="n">fit_eval</span>
        <span class="c1"># _end_for_</span>

        <span class="c1"># Convert list to numpy array.</span>
        <span class="n">arr</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">fitness_i</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="nb">float</span><span class="p">)</span>

        <span class="c1"># Return the mean and std of the fitness values.</span>
        <span class="k">return</span> <span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">arr</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="nb">float</span><span class="p">),</span> <span class="n">np</span><span class="o">.</span><span class="n">std</span><span class="p">(</span><span class="n">arr</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="nb">float</span><span class="p">)</span></div>

    <span class="c1"># _end_def_</span>

<div class="viewcode-block" id="StandardGA.best_chromosome">
<a class="viewcode-back" href="../../../pygenalgo.engines.html#pygenalgo.engines.standard_ga.StandardGA.best_chromosome">[docs]</a>
    <span class="k">def</span> <span class="nf">best_chromosome</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Chromosome</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Auxiliary method.</span>

<span class="sd">        :return: Return the chromosome with the highest fitness.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># Return the chromosome with the highest fitness.</span>
        <span class="k">return</span> <span class="nb">max</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">population</span><span class="p">,</span> <span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">c</span><span class="p">:</span> <span class="n">c</span><span class="o">.</span><span class="n">fitness</span><span class="p">)</span></div>


    <span class="c1"># _end_def_</span>

<div class="viewcode-block" id="StandardGA.run">
<a class="viewcode-back" href="../../../pygenalgo.engines.html#pygenalgo.engines.standard_ga.StandardGA.run">[docs]</a>
    <span class="k">def</span> <span class="nf">run</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">epochs</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">100</span><span class="p">,</span> <span class="n">elitism</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span> <span class="n">correction</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
            <span class="n">f_tol</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">1.0e-8</span><span class="p">,</span> <span class="n">parallel</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span> <span class="n">verbose</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Main method of the StandardGA class, that implements the evolutionary routine.</span>

<span class="sd">        :param epochs: (int) maximum number of iterations in the evolution process.</span>

<span class="sd">        :param elitism: (bool) flag that defines elitism. If &#39;True&#39; then the chromosome</span>
<span class="sd">        with the higher fitness will always be copied to the next generation (unaltered).</span>

<span class="sd">        :param correction: (bool) flag that if set to &#39;True&#39; will check the validity of</span>
<span class="sd">        the population (at the gene level) and attempt to correct the genome by calling</span>
<span class="sd">        the random() method of the flawed gene.</span>

<span class="sd">        :param f_tol: (float) tolerance in the difference between the average values of</span>
<span class="sd">        two consecutive populations. It is used to determine the convergence of the population.</span>

<span class="sd">        :param parallel: (bool) Flag that enables parallel computation of the fitness function.</span>

<span class="sd">        :param verbose: (bool) if &#39;True&#39; it will display periodically information about</span>
<span class="sd">        the current average fitness and spread of the population.</span>

<span class="sd">        :return: None.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="c1"># Make sure the genetic operator counters are reset before each run().</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_cross_op</span><span class="o">.</span><span class="n">reset_counter</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_mutate_op</span><span class="o">.</span><span class="n">reset_counter</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_select_op</span><span class="o">.</span><span class="n">reset_counter</span><span class="p">()</span>

        <span class="c1"># Reset stats dictionary.</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_stats</span><span class="p">[</span><span class="s2">&quot;avg&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">clear</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_stats</span><span class="p">[</span><span class="s2">&quot;std&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">clear</span><span class="p">()</span>

        <span class="c1"># Get the size of the population.</span>
        <span class="n">N</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">population</span><span class="p">)</span>

        <span class="c1"># Get the average/std fitness before optimisation.</span>
        <span class="n">avg_fitness_0</span><span class="p">,</span> <span class="n">std_fitness_0</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">evaluate_fitness</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">population</span><span class="p">,</span> <span class="n">parallel</span><span class="p">)</span>

        <span class="c1"># Update the mean/std in the dictionary.</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">update_stats</span><span class="p">(</span><span class="n">avg_fitness_0</span><span class="p">,</span> <span class="n">std_fitness_0</span><span class="p">)</span>

        <span class="c1"># Display an information message.</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Initial Avg. Fitness = </span><span class="si">{</span><span class="n">avg_fitness_0</span><span class="si">:</span><span class="s2">.4f</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="c1"># Initial time instant.</span>
        <span class="n">time_t0</span> <span class="o">=</span> <span class="n">time</span><span class="o">.</span><span class="n">perf_counter</span><span class="p">()</span>

        <span class="c1"># Repeat &#39;epoch&#39; times.</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">epochs</span><span class="p">):</span>

            <span class="c1"># SELECT the parents.</span>
            <span class="c1"># This will create a NEW copy of the population.</span>
            <span class="n">population_i</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_select_op</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">population</span><span class="p">)</span>

            <span class="c1"># CROSSOVER/MUTATE to produce offsprings.</span>
            <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">N</span> <span class="o">-</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">):</span>
                <span class="c1"># Replace directly the OLD parents with the NEW offsprings.</span>
                <span class="n">population_i</span><span class="p">[</span><span class="n">j</span><span class="p">],</span> <span class="n">population_i</span><span class="p">[</span><span class="n">j</span> <span class="o">+</span> <span class="mi">1</span><span class="p">]</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_cross_op</span><span class="p">(</span><span class="n">population_i</span><span class="p">[</span><span class="n">j</span><span class="p">],</span>
                                                                      <span class="n">population_i</span><span class="p">[</span><span class="n">j</span> <span class="o">+</span> <span class="mi">1</span><span class="p">])</span>
                <span class="c1"># MUTATE in place the 1st offspring.</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">_mutate_op</span><span class="p">(</span><span class="n">population_i</span><span class="p">[</span><span class="n">j</span><span class="p">])</span>

                <span class="c1"># MUTATE in place the 2nd offspring.</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">_mutate_op</span><span class="p">(</span><span class="n">population_i</span><span class="p">[</span><span class="n">j</span> <span class="o">+</span> <span class="mi">1</span><span class="p">])</span>
            <span class="c1"># _end_for_</span>

            <span class="c1"># Check if &#39;corrections&#39; are enabled.</span>
            <span class="k">if</span> <span class="n">correction</span><span class="p">:</span>
                <span class="c1"># Apply the function.</span>
                <span class="n">total_corrections</span> <span class="o">=</span> <span class="n">apply_corrections</span><span class="p">(</span><span class="n">population_i</span><span class="p">)</span>

                <span class="c1"># Print only if there were corrections,</span>
                <span class="c1"># to avoid cluttering the screen.</span>
                <span class="k">if</span> <span class="n">total_corrections</span><span class="p">:</span>
                    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;&gt; </span><span class="si">{</span><span class="n">total_corrections</span><span class="si">}</span><span class="s2"> correction(s) took place at epoch: </span><span class="si">{</span><span class="n">i</span><span class="si">}</span><span class="s2">.&quot;</span><span class="p">)</span>
                <span class="c1"># _end_if_</span>

            <span class="c1"># _end_if_</span>

            <span class="c1"># Check if &#39;elitism&#39; is enabled.</span>
            <span class="k">if</span> <span class="n">elitism</span><span class="p">:</span>
                <span class="c1"># Find the individual chromosome with the highest fitness value</span>
                <span class="c1"># (from the old population).</span>
                <span class="n">best_chromosome</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">population</span><span class="p">,</span> <span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">c</span><span class="p">:</span> <span class="n">c</span><span class="o">.</span><span class="n">fitness</span><span class="p">)</span>

                <span class="c1"># Select randomly a position.</span>
                <span class="n">locus</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">rng_GA</span><span class="o">.</span><span class="n">integers</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">N</span><span class="p">)</span>

                <span class="c1"># Replace the chromosome with the previous best.</span>
                <span class="n">population_i</span><span class="p">[</span><span class="n">locus</span><span class="p">]</span> <span class="o">=</span> <span class="n">best_chromosome</span><span class="o">.</span><span class="n">clone</span><span class="p">()</span>
            <span class="c1"># _end_if_</span>

            <span class="c1"># Calculate the (new) average/std of the fitness.</span>
            <span class="n">avg_fitness_i</span><span class="p">,</span> <span class="n">std_fitness_i</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">evaluate_fitness</span><span class="p">(</span><span class="n">population_i</span><span class="p">,</span> <span class="n">parallel</span><span class="p">)</span>

            <span class="c1"># Update the mean/std in the dictionary.</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">update_stats</span><span class="p">(</span><span class="n">avg_fitness_i</span><span class="p">,</span> <span class="n">std_fitness_i</span><span class="p">)</span>

            <span class="c1"># Check if we want to print output.</span>
            <span class="k">if</span> <span class="n">verbose</span> <span class="ow">and</span> <span class="n">np</span><span class="o">.</span><span class="n">mod</span><span class="p">(</span><span class="n">i</span><span class="p">,</span> <span class="mi">10</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
                <span class="c1"># Display an information message.</span>
                <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Epoch: </span><span class="si">{</span><span class="n">i</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="mi">1</span><span class="si">:</span><span class="s2">&gt;5</span><span class="si">}</span><span class="s2"> -&gt; Avg. Fitness = </span><span class="si">{</span><span class="n">avg_fitness_i</span><span class="si">:</span><span class="s2">.4f</span><span class="si">}</span><span class="s2">, &quot;</span>
                      <span class="sa">f</span><span class="s2">&quot;Spread = </span><span class="si">{</span><span class="n">std_fitness_i</span><span class="si">:</span><span class="s2">.4f</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="c1"># _end_if_</span>

            <span class="c1"># Check for convergence.</span>
            <span class="c1"># Here we don&#39;t only check the average performance, but we also check the</span>
            <span class="c1"># spread of the population. If all the chromosomes are similar the spread</span>
            <span class="c1"># should be small (very close to zero).</span>
            <span class="k">if</span> <span class="n">np</span><span class="o">.</span><span class="n">fabs</span><span class="p">(</span><span class="n">avg_fitness_i</span> <span class="o">-</span> <span class="n">avg_fitness_0</span><span class="p">)</span> <span class="o">&lt;</span> <span class="n">f_tol</span> <span class="ow">and</span> <span class="n">std_fitness_i</span> <span class="o">&lt;</span> <span class="mf">1.0E-1</span><span class="p">:</span>
                <span class="c1"># Display a warning message.</span>
                <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="vm">__class__</span><span class="o">.</span><span class="vm">__name__</span><span class="si">}</span><span class="s2"> finished in </span><span class="si">{</span><span class="n">i</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="mi">1</span><span class="si">}</span><span class="s2"> iterations.&quot;</span><span class="p">)</span>

                <span class="c1"># Exit from the loop.</span>
                <span class="k">break</span>
            <span class="c1"># _end_if_</span>

            <span class="c1"># Update the average value for the next iteration.</span>
            <span class="n">avg_fitness_0</span> <span class="o">=</span> <span class="n">avg_fitness_i</span>

            <span class="c1"># Update the old population with the new chromosomes.</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">population</span> <span class="o">=</span> <span class="n">population_i</span>
        <span class="c1"># _end_for_</span>

        <span class="c1"># Final time instant.</span>
        <span class="n">time_tf</span> <span class="o">=</span> <span class="n">time</span><span class="o">.</span><span class="n">perf_counter</span><span class="p">()</span>

        <span class="c1"># Display the final average fitness value.</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Final   Avg. Fitness = </span><span class="si">{</span><span class="n">avg_fitness_0</span><span class="si">:</span><span class="s2">.4f</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="c1"># Print final duration in seconds.</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Elapsed time: </span><span class="si">{</span><span class="p">(</span><span class="n">time_tf</span><span class="w"> </span><span class="o">-</span><span class="w"> </span><span class="n">time_t0</span><span class="p">)</span><span class="si">:</span><span class="s2">.3f</span><span class="si">}</span><span class="s2"> seconds.&quot;</span><span class="p">,</span> <span class="n">end</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span></div>

    <span class="c1"># _end_def_</span>

    <span class="c1"># Auxiliary.</span>
    <span class="k">def</span> <span class="fm">__call__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="o">*</span><span class="n">args</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        This is only a wrapper of the &quot;run&quot; method.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">run</span><span class="p">(</span><span class="o">*</span><span class="n">args</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>

    <span class="c1"># _end_def_</span>

<div class="viewcode-block" id="StandardGA.print_operator_stats">
<a class="viewcode-back" href="../../../pygenalgo.engines.html#pygenalgo.engines.standard_ga.StandardGA.print_operator_stats">[docs]</a>
    <span class="k">def</span> <span class="nf">print_operator_stats</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Print the genetic operators stats.</span>

<span class="sd">        :return: None.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="c1"># First print the selection operator.</span>
        <span class="nb">print</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_select_op</span><span class="p">)</span>

        <span class="c1"># Second print the crossover operator.</span>
        <span class="nb">print</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_cross_op</span><span class="p">)</span>

        <span class="c1"># Check if we used the SuperCrossover.</span>
        <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_cross_op</span><span class="p">,</span> <span class="n">SuperCrossover</span><span class="p">):</span>
            <span class="c1"># Call internally all operators.</span>
            <span class="k">for</span> <span class="n">op</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_cross_op</span><span class="o">.</span><span class="n">list_all</span><span class="p">:</span>
                <span class="nb">print</span><span class="p">(</span><span class="n">op</span><span class="p">)</span>
            <span class="c1"># _end_for_</span>
        <span class="c1"># _end_if_</span>

        <span class="c1"># Lastly print the mutation operator.</span>
        <span class="nb">print</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_mutate_op</span><span class="p">)</span>

        <span class="c1"># Check if we used the SuperMutator.</span>
        <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_mutate_op</span><span class="p">,</span> <span class="n">SuperMutator</span><span class="p">):</span>
            <span class="c1"># Call internally all operators.</span>
            <span class="k">for</span> <span class="n">op</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_mutate_op</span><span class="o">.</span><span class="n">list_all</span><span class="p">:</span>
                <span class="nb">print</span><span class="p">(</span><span class="n">op</span><span class="p">)</span></div>
</div>

            <span class="c1"># _end_for_</span>
        <span class="c1"># _end_if_</span>

    <span class="c1"># _end_def_</span>

<span class="c1"># _end_class_</span>
</pre></div>

           </div>
          </div>
          <footer>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2024, Michalis Vrettas, PhD.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>