

<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="../../../">
<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>pygenalgo.engines.standard_ga &mdash; PyGenAlgo 01.70.50 documentation</title>
      <link rel="stylesheet" type="text/css" href="../../../_static/pygments.css?v=b86133f3" />
      <link rel="stylesheet" type="text/css" href="../../../_static/css/theme.css?v=e59714d7" />

  
      <script src="../../../_static/jquery.js?v=5d32c60e"></script>
      <script src="../../../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
      <script src="../../../_static/documentation_options.js?v=d0c4109e"></script>
      <script src="../../../_static/doctools.js?v=9a2dae69"></script>
      <script src="../../../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../../../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../../../index.html" class="icon icon-home">
            PyGenAlgo
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../modules.html">pygenalgo</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../../index.html">PyGenAlgo</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="../../index.html">Module code</a></li>
      <li class="breadcrumb-item active">pygenalgo.engines.standard_ga</li>
      <li class="wy-breadcrumbs-aside">
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <h1>Source code for pygenalgo.engines.standard_ga</h1><div class="highlight"><pre>
<span></span><span class="kn">import</span><span class="w"> </span><span class="nn">time</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">math</span><span class="w"> </span><span class="kn">import</span> <span class="n">isclose</span>

<span class="kn">from</span><span class="w"> </span><span class="nn">numpy</span><span class="w"> </span><span class="kn">import</span> <span class="p">(</span><span class="n">array</span><span class="p">,</span> <span class="n">nanmean</span><span class="p">,</span> <span class="n">nanstd</span><span class="p">,</span> <span class="n">isfinite</span><span class="p">)</span>

<span class="kn">from</span><span class="w"> </span><span class="nn">pygenalgo.engines.generic_ga</span><span class="w"> </span><span class="kn">import</span> <span class="n">GenericGA</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">pygenalgo.utils.auxiliary</span><span class="w"> </span><span class="kn">import</span> <span class="p">(</span><span class="n">apply_corrections</span><span class="p">,</span>
                                       <span class="n">average_hamming_distance</span><span class="p">)</span>

<span class="kn">from</span><span class="w"> </span><span class="nn">pygenalgo.operators.mutation.meta_mutator</span><span class="w"> </span><span class="kn">import</span> <span class="n">MetaMutator</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">pygenalgo.operators.crossover.meta_crossover</span><span class="w"> </span><span class="kn">import</span> <span class="n">MetaCrossover</span>


<span class="c1"># Public interface.</span>
<span class="n">__all__</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;StandardGA&quot;</span><span class="p">]</span>


<div class="viewcode-block" id="StandardGA">
<a class="viewcode-back" href="../../../pygenalgo.engines.html#pygenalgo.engines.standard_ga.StandardGA">[docs]</a>
<span class="k">class</span><span class="w"> </span><span class="nc">StandardGA</span><span class="p">(</span><span class="n">GenericGA</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Description:</span>

<span class="sd">        StandardGA model provides a basic implementation of the &quot;GenericGA&quot;,</span>
<span class="sd">        which at each iteration (epoch) replaces the whole population using</span>
<span class="sd">        the genetic operators (crossover and mutation).</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">def</span><span class="w"> </span><span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Default constructor of StandardGA object.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># Call the super constructor with all the input parameters.</span>
        <span class="nb">super</span><span class="p">()</span><span class="o">.</span><span class="fm">__init__</span><span class="p">(</span><span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>
    <span class="c1"># _end_def_</span>

<div class="viewcode-block" id="StandardGA.update_stats">
<a class="viewcode-back" href="../../../pygenalgo.engines.html#pygenalgo.engines.standard_ga.StandardGA.update_stats">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">update_stats</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">fit_list</span><span class="p">:</span> <span class="nb">list</span><span class="p">[</span><span class="nb">float</span><span class="p">])</span> <span class="o">-&gt;</span> <span class="nb">tuple</span><span class="p">[</span><span class="nb">float</span><span class="p">,</span> <span class="nb">float</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Update the stats dictionary with the mean/std values of the</span>
<span class="sd">        population fitness values.</span>

<span class="sd">        :param fit_list: (list) fitness values of the population.</span>

<span class="sd">        :return: the mean and std of the fitness values.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="c1"># Convert the fitness list in a numpy array.</span>
        <span class="n">arr</span> <span class="o">=</span> <span class="n">array</span><span class="p">(</span><span class="n">fit_list</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="nb">float</span><span class="p">)</span>

        <span class="c1"># Get the mean and std values.</span>
        <span class="n">avg_fitness</span> <span class="o">=</span> <span class="n">nanmean</span><span class="p">(</span><span class="n">arr</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="nb">float</span><span class="p">)</span>
        <span class="n">std_fitness</span> <span class="o">=</span> <span class="n">nanstd</span><span class="p">(</span><span class="n">arr</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="nb">float</span><span class="p">)</span>

        <span class="c1"># Make sure the stat values are finite.</span>
        <span class="k">if</span> <span class="nb">all</span><span class="p">(</span><span class="n">isfinite</span><span class="p">([</span><span class="n">avg_fitness</span><span class="p">,</span> <span class="n">std_fitness</span><span class="p">])):</span>

            <span class="c1"># Store them in the dictionary.</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_stats</span><span class="p">[</span><span class="s2">&quot;avg&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">avg_fitness</span><span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_stats</span><span class="p">[</span><span class="s2">&quot;std&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">std_fitness</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">RuntimeError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="vm">__class__</span><span class="o">.</span><span class="vm">__name__</span><span class="si">}</span><span class="s2">: Something went wrong with current &quot;</span>
                               <span class="sa">f</span><span class="s2">&quot;population. Mean=</span><span class="si">{</span><span class="n">avg_fitness</span><span class="si">:</span><span class="s2">.5f</span><span class="si">}</span><span class="s2">, Std=</span><span class="si">{</span><span class="n">std_fitness</span><span class="si">:</span><span class="s2">.5f</span><span class="si">}</span><span class="s2">.&quot;</span><span class="p">)</span>
        <span class="c1"># _end_if_</span>

        <span class="c1"># Return the average statistics.</span>
        <span class="k">return</span> <span class="n">avg_fitness</span><span class="p">,</span> <span class="n">std_fitness</span></div>

    <span class="c1"># _end_def_</span>

<div class="viewcode-block" id="StandardGA.run">
<a class="viewcode-back" href="../../../pygenalgo.engines.html#pygenalgo.engines.standard_ga.StandardGA.run">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">run</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">epochs</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">100</span><span class="p">,</span> <span class="n">elitism</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span> <span class="n">correction</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
            <span class="n">f_tol</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">parallel</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span> <span class="n">adapt_probs</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
            <span class="n">shuffle</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span> <span class="n">f_max_eval</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">verbose</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Main method of the StandardGA class, that implements the evolutionary routine.</span>

<span class="sd">        :param epochs: (int) maximum number of iterations in the evolution process.</span>

<span class="sd">        :param elitism: (bool) flag that defines elitism. If &#39;True&#39; then the chromosome</span>
<span class="sd">        with the higher fitness will always be copied to the next generation (unaltered).</span>

<span class="sd">        :param correction: (bool) flag that if set to &#39;True&#39; will check the validity of</span>
<span class="sd">        the population (at the gene level) and attempt to correct the genome by calling</span>
<span class="sd">        the random() method of the flawed gene.</span>

<span class="sd">        :param f_tol: (float) tolerance in the difference between the average values of two</span>
<span class="sd">        consecutive populations. It is used to determine the convergence of the population.</span>
<span class="sd">        If this value is None (default) the algorithm will terminate using the epochs value.</span>

<span class="sd">        :param parallel: (bool) Flag that enables parallel computation of the fitness function.</span>

<span class="sd">        :param adapt_probs: (bool) If enabled (set to True), it will allow the crossover and</span>
<span class="sd">        mutation probabilities to adapt according to the convergence of the population to a</span>
<span class="sd">        single solution. Default is set to False.</span>

<span class="sd">        :param shuffle: (bool) If enabled (set to True), it will shuffle the population before</span>
<span class="sd">        the application of the crossover and mutation operations. Default is set to True.</span>

<span class="sd">        :param f_max_eval: (int) it sets an upper limit of function evaluations. If the number</span>
<span class="sd">        is exceeded the genetic algorithm stops.</span>

<span class="sd">        :param verbose: (bool) if &#39;True&#39; it will display periodically information about the</span>
<span class="sd">        current average fitness and spread of the population.</span>

<span class="sd">        :return: None.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="c1"># Make sure everything is cleared.</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">clear_all</span><span class="p">()</span>

        <span class="c1"># Get the size of the population.</span>
        <span class="n">pop_size</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">population</span><span class="p">)</span>

        <span class="c1"># Get the fitness values before optimisation.</span>
        <span class="n">fit_list_0</span><span class="p">,</span> <span class="n">_</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">evaluate_fitness</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">population</span><span class="p">,</span> <span class="n">parallel</span><span class="p">)</span>

        <span class="c1"># Update the average statistics in the dictionary.</span>
        <span class="n">avg_fitness_0</span><span class="p">,</span> <span class="n">_</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">update_stats</span><span class="p">(</span><span class="n">fit_list_0</span><span class="p">)</span>

        <span class="c1"># Store the initial crossover and mutation probabilities.</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_stats</span><span class="p">[</span><span class="s2">&quot;prob_crossx&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_crossx_op</span><span class="o">.</span><span class="n">probability</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_stats</span><span class="p">[</span><span class="s2">&quot;prob_mutate&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_mutate_op</span><span class="o">.</span><span class="n">probability</span><span class="p">)</span>

        <span class="c1"># Local variable to display information on the screen.</span>
        <span class="c1"># To avoid cluttering the screen we print info only 10</span>
        <span class="c1"># times regardless of the total number of epochs.</span>
        <span class="n">its_time_to_print</span> <span class="o">=</span> <span class="p">(</span><span class="n">epochs</span><span class="o">//</span><span class="mi">10</span><span class="p">)</span>

        <span class="c1"># Display an information message.</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Initial Avg. Fitness = </span><span class="si">{</span><span class="n">avg_fitness_0</span><span class="si">:</span><span class="s2">.4f</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="c1"># Initial time instant.</span>
        <span class="n">time_t0</span> <span class="o">=</span> <span class="n">time</span><span class="o">.</span><span class="n">perf_counter</span><span class="p">()</span>

        <span class="c1"># Repeat &#39;epoch&#39; times.</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">epochs</span><span class="p">):</span>

            <span class="c1"># Update current iteration in the selection operator.</span>
            <span class="c1"># Currently, this is used only from Boltzmann Selector.</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">select_op</span><span class="o">.</span><span class="n">iter</span> <span class="o">=</span> <span class="n">i</span>

            <span class="c1"># SELECT the parents.</span>
            <span class="n">population_i</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">select_op</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">population</span><span class="p">)</span>

            <span class="c1"># Shuffle the selected parents.</span>
            <span class="k">if</span> <span class="n">shuffle</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">rng_GA</span><span class="o">.</span><span class="n">shuffle</span><span class="p">(</span><span class="n">population_i</span><span class="p">)</span>
            <span class="c1"># _end_def_</span>

            <span class="c1"># CROSSOVER/MUTATE to produce offsprings.</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">crossover_mutate</span><span class="p">(</span><span class="n">population_i</span><span class="p">)</span>

            <span class="c1"># Calculate the new fitness values.</span>
            <span class="n">fit_list_i</span><span class="p">,</span> <span class="n">found_solution</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">evaluate_fitness</span><span class="p">(</span><span class="n">population_i</span><span class="p">,</span> <span class="n">parallel</span><span class="p">)</span>

            <span class="c1"># Check if &#39;corrections&#39; are enabled.</span>
            <span class="k">if</span> <span class="n">correction</span><span class="p">:</span>
                <span class="c1"># Apply the function.</span>
                <span class="n">total_corrections</span><span class="p">,</span> <span class="n">f_counts</span> <span class="o">=</span> <span class="n">apply_corrections</span><span class="p">(</span><span class="n">population_i</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">fitness_func</span><span class="p">)</span>

                <span class="c1"># Update the function evaluation counter.</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">f_eval_increase_by</span><span class="p">(</span><span class="n">f_counts</span><span class="p">)</span>

                <span class="c1"># Print only if there were corrections,</span>
                <span class="c1"># to avoid cluttering the screen.</span>
                <span class="k">if</span> <span class="n">verbose</span> <span class="ow">and</span> <span class="n">total_corrections</span><span class="p">:</span>
                    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;&gt; </span><span class="si">{</span><span class="n">total_corrections</span><span class="si">}</span><span class="s2"> correction(s) took place at epoch: </span><span class="si">{</span><span class="n">i</span><span class="si">}</span><span class="s2">.&quot;</span><span class="p">)</span>
                <span class="c1"># _end_if_</span>
            <span class="c1"># _end_if_</span>

            <span class="c1"># Check if &#39;elitism&#39; is enabled.</span>
            <span class="k">if</span> <span class="n">elitism</span><span class="p">:</span>
                <span class="c1"># Get the reference of the best chromosome</span>
                <span class="c1"># from the previous generation.</span>
                <span class="n">previous_best</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">best_chromosome</span><span class="p">()</span>

                <span class="c1"># Check if the chromosome already exists in</span>
                <span class="c1"># the current generation to avoid flooding</span>
                <span class="c1"># the new pool with the same chromosome.</span>
                <span class="k">if</span> <span class="n">previous_best</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">population_i</span><span class="p">:</span>

                    <span class="c1"># Select a position at random.</span>
                    <span class="n">locus</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">rng_GA</span><span class="o">.</span><span class="n">integers</span><span class="p">(</span><span class="n">pop_size</span><span class="p">)</span>

                    <span class="c1"># Replace it with the previous best.</span>
                    <span class="n">population_i</span><span class="p">[</span><span class="n">locus</span><span class="p">]</span> <span class="o">=</span> <span class="n">previous_best</span>

                    <span class="c1"># Update the list of fitness values to reflect the update.</span>
                    <span class="n">fit_list_i</span><span class="p">[</span><span class="n">locus</span><span class="p">]</span> <span class="o">=</span> <span class="n">population_i</span><span class="p">[</span><span class="n">locus</span><span class="p">]</span><span class="o">.</span><span class="n">fitness</span>
                <span class="c1"># _end_if_</span>
            <span class="c1"># _end_if_</span>

            <span class="c1"># Update the mean/std in the dictionary.</span>
            <span class="n">avg_fitness_i</span><span class="p">,</span> <span class="n">std_fitness_i</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">update_stats</span><span class="p">(</span><span class="n">fit_list_i</span><span class="p">)</span>

            <span class="c1"># Check if we want to print output.</span>
            <span class="k">if</span> <span class="n">verbose</span> <span class="ow">and</span> <span class="p">(</span><span class="n">i</span> <span class="o">%</span> <span class="n">its_time_to_print</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
                <span class="c1"># Display an information message.</span>
                <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Epoch: </span><span class="si">{</span><span class="n">i</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="mi">1</span><span class="si">:</span><span class="s2">&gt;5</span><span class="si">}</span><span class="s2"> -&gt; &quot;</span>
                      <span class="sa">f</span><span class="s2">&quot;Avg. Fitness = </span><span class="si">{</span><span class="n">avg_fitness_i</span><span class="si">:</span><span class="s2">.4f</span><span class="si">}</span><span class="s2">, &quot;</span>
                      <span class="sa">f</span><span class="s2">&quot;Spread = </span><span class="si">{</span><span class="n">std_fitness_i</span><span class="si">:</span><span class="s2">.4f</span><span class="si">}</span><span class="s2">.&quot;</span><span class="p">)</span>
            <span class="c1"># _end_if_</span>

            <span class="c1"># Update the old population with the new chromosomes.</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">population</span> <span class="o">=</span> <span class="n">population_i</span>

            <span class="c1"># Check for termination.</span>
            <span class="k">if</span> <span class="n">found_solution</span><span class="p">:</span>
                <span class="c1"># Display a warning message.</span>
                <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="vm">__class__</span><span class="o">.</span><span class="vm">__name__</span><span class="si">}</span><span class="s2"> finished in </span><span class="si">{</span><span class="n">i</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="mi">1</span><span class="si">}</span><span class="s2"> iterations.&quot;</span><span class="p">)</span>

                <span class="c1"># Exit from the loop.</span>
                <span class="k">break</span>
            <span class="c1"># _end_if_</span>

            <span class="c1"># Check for the maximum function evaluations.</span>
            <span class="k">if</span> <span class="n">f_max_eval</span> <span class="ow">and</span> <span class="bp">self</span><span class="o">.</span><span class="n">f_eval</span> <span class="o">&gt;=</span> <span class="n">f_max_eval</span><span class="p">:</span>

                <span class="c1"># Display an information message.</span>
                <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="vm">__class__</span><span class="o">.</span><span class="vm">__name__</span><span class="si">}</span><span class="s2"> &quot;</span>
                      <span class="s2">&quot;Reached the maximum number of function evaluations.&quot;</span><span class="p">)</span>

                <span class="c1"># Exit from the loop.</span>
                <span class="k">break</span>
            <span class="c1"># _end_if_</span>

            <span class="c1"># Compute the current average Hamming distance.</span>
            <span class="n">avg_distance</span> <span class="o">=</span> <span class="n">average_hamming_distance</span><span class="p">(</span><span class="n">population_i</span><span class="p">)</span>

            <span class="c1"># Check for convergence.</span>
            <span class="k">if</span> <span class="n">f_tol</span> <span class="ow">and</span> <span class="n">isclose</span><span class="p">(</span><span class="n">avg_fitness_i</span><span class="p">,</span> <span class="n">avg_fitness_0</span><span class="p">,</span>
                                 <span class="n">rel_tol</span><span class="o">=</span><span class="mf">1.0E-5</span><span class="p">,</span> <span class="n">abs_tol</span><span class="o">=</span><span class="n">f_tol</span><span class="p">)</span> <span class="ow">and</span> <span class="n">avg_distance</span> <span class="o">&lt;</span> <span class="mf">0.025</span><span class="p">:</span>
                <span class="c1"># Display a warning message.</span>
                <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="vm">__class__</span><span class="o">.</span><span class="vm">__name__</span><span class="si">}</span><span class="s2"> converged in </span><span class="si">{</span><span class="n">i</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="mi">1</span><span class="si">}</span><span class="s2"> iterations.&quot;</span><span class="p">)</span>

                <span class="c1"># Exit from the loop.</span>
                <span class="k">break</span>
            <span class="c1"># _end_if_</span>

            <span class="c1"># Update the average value for the next iteration.</span>
            <span class="n">avg_fitness_0</span> <span class="o">=</span> <span class="n">avg_fitness_i</span>

            <span class="c1"># Check the adaptive flag.</span>
            <span class="k">if</span> <span class="n">adapt_probs</span><span class="p">:</span>

                <span class="c1"># For threshold, we use the average Hamming</span>
                <span class="c1"># distance of the &#39;current&#39; population.</span>
                <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">adapt_probabilities</span><span class="p">(</span><span class="n">threshold</span><span class="o">=</span><span class="n">avg_distance</span><span class="p">):</span>
                    <span class="c1"># Store the updated crossover and mutation probabilities.</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">_stats</span><span class="p">[</span><span class="s2">&quot;prob_crossx&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_crossx_op</span><span class="o">.</span><span class="n">probability</span><span class="p">)</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">_stats</span><span class="p">[</span><span class="s2">&quot;prob_mutate&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_mutate_op</span><span class="o">.</span><span class="n">probability</span><span class="p">)</span>
            <span class="c1"># _end_if_</span>

        <span class="c1"># _end_for_</span>

        <span class="c1"># Final time instant.</span>
        <span class="n">time_tf</span> <span class="o">=</span> <span class="n">time</span><span class="o">.</span><span class="n">perf_counter</span><span class="p">()</span>

        <span class="c1"># Display the final average fitness value.</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Final   Avg. Fitness = </span><span class="si">{</span><span class="n">avg_fitness_0</span><span class="si">:</span><span class="s2">.4f</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="c1"># Print final duration in seconds.</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Elapsed time: </span><span class="si">{</span><span class="p">(</span><span class="n">time_tf</span><span class="w"> </span><span class="o">-</span><span class="w"> </span><span class="n">time_t0</span><span class="p">)</span><span class="si">:</span><span class="s2">.3f</span><span class="si">}</span><span class="s2"> seconds.&quot;</span><span class="p">)</span></div>

    <span class="c1"># _end_def_</span>

<div class="viewcode-block" id="StandardGA.print_operator_stats">
<a class="viewcode-back" href="../../../pygenalgo.engines.html#pygenalgo.engines.standard_ga.StandardGA.print_operator_stats">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">print_operator_stats</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Print the genetic operators stats.</span>

<span class="sd">        :return: None.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="c1"># First print the selection operator.</span>
        <span class="nb">print</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">select_op</span><span class="p">)</span>

        <span class="c1"># Second print the crossover operator.</span>
        <span class="nb">print</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">crossover_op</span><span class="p">)</span>

        <span class="c1"># Check if we used the MetaCrossover.</span>
        <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">crossover_op</span><span class="p">,</span> <span class="n">MetaCrossover</span><span class="p">):</span>
            <span class="c1"># Call internally all operators.</span>
            <span class="k">for</span> <span class="n">op</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">crossover_op</span><span class="o">.</span><span class="n">items</span><span class="p">:</span>
                <span class="nb">print</span><span class="p">(</span><span class="n">op</span><span class="p">)</span>
            <span class="c1"># _end_for_</span>
        <span class="c1"># _end_if_</span>

        <span class="c1"># Lastly print the mutation operator.</span>
        <span class="nb">print</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">mutate_op</span><span class="p">)</span>

        <span class="c1"># Check if we used the MetaMutator.</span>
        <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">mutate_op</span><span class="p">,</span> <span class="n">MetaMutator</span><span class="p">):</span>
            <span class="c1"># Call internally all operators.</span>
            <span class="k">for</span> <span class="n">op</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">mutate_op</span><span class="o">.</span><span class="n">items</span><span class="p">:</span>
                <span class="nb">print</span><span class="p">(</span><span class="n">op</span><span class="p">)</span></div>
</div>

            <span class="c1"># _end_for_</span>
        <span class="c1"># _end_if_</span>
    <span class="c1"># _end_def_</span>

<span class="c1"># _end_class_</span>
</pre></div>

           </div>
          </div>
          <footer>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2024, Michalis Vrettas, PhD.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>