

<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="./">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Gaussian Mixture - 2D &mdash; PyGenAlgo 1.8.2 documentation</title>
      <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=b86133f3" />
      <link rel="stylesheet" type="text/css" href="_static/css/theme.css?v=e59714d7" />

  
      <script src="_static/jquery.js?v=5d32c60e"></script>
      <script src="_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
      <script src="_static/documentation_options.js?v=269e6701"></script>
      <script src="_static/doctools.js?v=9a2dae69"></script>
      <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
      <script async="async" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="prev" title="Travelling Salesman Problem" href="tsp_path.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="index.html" class="icon icon-home">
            PyGenAlgo
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="modules.html">Modules</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="examples.html">Examples</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="examples.html#simple-optimization">Simple optimization</a></li>
<li class="toctree-l2"><a class="reference internal" href="examples.html#combinatorial-problem-with-constraints">Combinatorial problem (with constraints)</a></li>
<li class="toctree-l2 current"><a class="reference internal" href="examples.html#multimodal-function">Multimodal function</a><ul class="current">
<li class="toctree-l3 current"><a class="current reference internal" href="#">Gaussian Mixture - 2D</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#step-1-import-python-libraries-and-pygenalgo-classes">Step 1: Import python libraries and PyGenAlgo classes</a></li>
<li class="toctree-l4"><a class="reference internal" href="#step-2-define-the-objective-function">Step 2: Define the objective function</a></li>
<li class="toctree-l4"><a class="reference internal" href="#step-3-set-the-ga-parameters">Step 3: Set the GA parameters</a></li>
<li class="toctree-l4"><a class="reference internal" href="#step-4-run-the-optimization">Step 4: Run the optimization</a></li>
<li class="toctree-l4"><a class="reference internal" href="#step-5-extract-the-data-for-analysis-and-plotting">Step 5: Extract the data for analysis and plotting</a></li>
<li class="toctree-l4"><a class="reference internal" href="#step-6-visualize-the-solutions">Step 6: Visualize the solutions</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="examples.html#more-examples">More examples</a></li>
</ul>
</li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">PyGenAlgo</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="examples.html">Examples</a></li>
      <li class="breadcrumb-item active">Gaussian Mixture - 2D</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/gaussian_2d.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="gaussian-mixture-2d">
<h1>Gaussian Mixture - 2D<a class="headerlink" href="#gaussian-mixture-2d" title="Link to this heading"></a></h1>
<p>Description:</p>
<blockquote>
<div><ul class="simple">
<li><p>Optimization (max)</p></li>
<li><p>Multimodal (yes)</p></li>
</ul>
</div></blockquote>
<p>This function provides a 2D Gaussian mixture model.</p>
<blockquote>
<div><p>The equations are given by the Multivariate Normal Distribution,
with four modes (2 global and 2 local):</p>
<div class="math notranslate nohighlight">
\[f(x) = \sum_{i=1}^{4} \mathcal{N}(\mu_i, \Sigma_i)\]</div>
<p>with mean vectors:</p>
<div class="math notranslate nohighlight">
\[ \begin{align}\begin{aligned}\mu_1 = [-0.0, -1.0]\\\mu_2 = [-4.0, -6.0]\\\mu_3 = [-5.0, +1.0]\\\mu_4 = [5.0, -10.0]\end{aligned}\end{align} \]</div>
<p>and covariances:</p>
<div class="math notranslate nohighlight">
\[ \begin{align}\begin{aligned}\Sigma_1 = [ [ 1.0, 0.1 ], [ 0.1, 1.0 ] ]\\\Sigma_2 = [ [ 1.0, 0.1 ], [ 0.1, 1.0 ] ]\\\Sigma_3 = [ [ 1.2, 0.3 ], [ 0.3, 1.2 ] ]\\\Sigma_4 = [ [ 1.2, 0.3 ], [ 0.3, 1.2 ] ]\end{aligned}\end{align} \]</div>
</div></blockquote>
<section id="step-1-import-python-libraries-and-pygenalgo-classes">
<h2>Step 1: Import python libraries and PyGenAlgo classes<a class="headerlink" href="#step-1-import-python-libraries-and-pygenalgo-classes" title="Link to this heading"></a></h2>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span><span class="w"> </span><span class="nn">numpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">np</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">matplotlib</span><span class="w"> </span><span class="kn">import</span> <span class="n">pyplot</span> <span class="k">as</span> <span class="n">plt</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">scipy.stats</span><span class="w"> </span><span class="kn">import</span> <span class="n">multivariate_normal</span>

<span class="c1"># Enable LaTex in plotting.</span>
<span class="n">plt</span><span class="o">.</span><span class="n">rcParams</span><span class="p">[</span><span class="s2">&quot;text.usetex&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="kc">True</span>

<span class="c1"># Import main classes.</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">pygenalgo.genome.gene</span><span class="w"> </span><span class="kn">import</span> <span class="n">Gene</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">pygenalgo.genome.chromosome</span><span class="w"> </span><span class="kn">import</span> <span class="n">Chromosome</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">pygenalgo.utils.utilities</span><span class="w"> </span><span class="kn">import</span> <span class="n">cost_function</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">pygenalgo.engines.standard_ga</span><span class="w"> </span><span class="kn">import</span> <span class="n">StandardGA</span>

<span class="c1"># Import Selection Operator(s).</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">pygenalgo.operators.selection.neighborhood_selector</span><span class="w"> </span><span class="kn">import</span> <span class="n">NeighborhoodSelector</span>

<span class="c1"># Import Crossover Operator(s).</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">pygenalgo.operators.crossover.blend_crossover</span><span class="w"> </span><span class="kn">import</span> <span class="n">BlendCrossover</span>

<span class="c1"># Import Mutation Operator(s).</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">pygenalgo.operators.mutation.gaussian_mutator</span><span class="w"> </span><span class="kn">import</span> <span class="n">GaussianMutator</span>
</pre></div>
</div>
</section>
<section id="step-2-define-the-objective-function">
<h2>Step 2: Define the objective function<a class="headerlink" href="#step-2-define-the-objective-function" title="Link to this heading"></a></h2>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># Setup the 2D Gaussian functions.</span>
<span class="c1"># NOTE: Because of their covariance setup, the first two (mvn_1 and mvn_2)</span>
<span class="c1"># will have higher modes then the other two. So the maximum will be in one</span>
<span class="c1"># of these two modes.</span>

<span class="c1"># Each one with different mean and covariance matrix.</span>
<span class="n">mvn_1</span> <span class="o">=</span> <span class="n">multivariate_normal</span><span class="p">([</span><span class="o">-</span><span class="mf">0.0</span><span class="p">,</span> <span class="o">-</span><span class="mf">1.0</span><span class="p">],</span> <span class="p">[[</span><span class="mf">1.0</span><span class="p">,</span> <span class="mf">0.1</span><span class="p">],</span> <span class="p">[</span><span class="mf">0.1</span><span class="p">,</span> <span class="mf">1.0</span><span class="p">]])</span>

<span class="n">mvn_2</span> <span class="o">=</span> <span class="n">multivariate_normal</span><span class="p">([</span><span class="o">-</span><span class="mf">4.0</span><span class="p">,</span> <span class="o">-</span><span class="mf">6.0</span><span class="p">],</span> <span class="p">[[</span><span class="mf">1.0</span><span class="p">,</span> <span class="mf">0.1</span><span class="p">],</span> <span class="p">[</span><span class="mf">0.1</span><span class="p">,</span> <span class="mf">1.0</span><span class="p">]])</span>

<span class="n">mvn_3</span> <span class="o">=</span> <span class="n">multivariate_normal</span><span class="p">([</span><span class="o">-</span><span class="mf">10.0</span><span class="p">,</span> <span class="mf">5.0</span><span class="p">],</span> <span class="p">[[</span><span class="mf">1.2</span><span class="p">,</span> <span class="mf">0.3</span><span class="p">],</span> <span class="p">[</span><span class="mf">0.3</span><span class="p">,</span> <span class="mf">1.2</span><span class="p">]])</span>

<span class="n">mvn_4</span> <span class="o">=</span> <span class="n">multivariate_normal</span><span class="p">([</span><span class="mf">5.0</span><span class="p">,</span> <span class="o">-</span><span class="mf">10.0</span><span class="p">],</span> <span class="p">[[</span><span class="mf">1.2</span><span class="p">,</span> <span class="mf">0.3</span><span class="p">],</span> <span class="p">[</span><span class="mf">0.3</span><span class="p">,</span> <span class="mf">1.2</span><span class="p">]])</span>

<span class="c1"># Define the negative log of the pdfs.</span>
<span class="k">def</span><span class="w"> </span><span class="nf">negative_log_pdfx</span><span class="p">(</span><span class="n">x</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">float</span><span class="p">:</span>
    <span class="k">return</span> <span class="o">-</span><span class="n">np</span><span class="o">.</span><span class="n">log</span><span class="p">(</span><span class="n">mvn_1</span><span class="o">.</span><span class="n">pdf</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="o">+</span> <span class="n">mvn_2</span><span class="o">.</span><span class="n">pdf</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="o">+</span> <span class="n">mvn_3</span><span class="o">.</span><span class="n">pdf</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="o">+</span> <span class="n">mvn_4</span><span class="o">.</span><span class="n">pdf</span><span class="p">(</span><span class="n">x</span><span class="p">))</span>
<span class="c1"># _end_def_</span>

<span class="c1"># Multimodal cost function.</span>
<span class="nd">@cost_function</span><span class="p">(</span><span class="n">minimize</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="k">def</span><span class="w"> </span><span class="nf">fun_test2D</span><span class="p">(</span><span class="n">individual</span><span class="p">:</span> <span class="n">Chromosome</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">float</span><span class="p">:</span>

    <span class="c1"># Extract the x, y values.</span>
    <span class="n">x</span><span class="p">,</span> <span class="n">y</span> <span class="o">=</span> <span class="n">individual</span><span class="o">.</span><span class="n">values</span><span class="p">()</span>

    <span class="c1"># Compute the final value.</span>
    <span class="n">f_value</span> <span class="o">=</span> <span class="n">negative_log_pdfx</span><span class="p">([</span><span class="n">x</span><span class="p">,</span> <span class="n">y</span><span class="p">])</span>

    <span class="c1"># Return the solution.</span>
    <span class="k">return</span> <span class="n">f_value</span><span class="o">.</span><span class="n">item</span><span class="p">()</span>
</pre></div>
</div>
</section>
<section id="step-3-set-the-ga-parameters">
<h2>Step 3: Set the GA parameters<a class="headerlink" href="#step-3-set-the-ga-parameters" title="Link to this heading"></a></h2>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># Set a seed for reproducible initial population.</span>
<span class="n">SEED</span> <span class="o">=</span> <span class="mi">1821</span>

<span class="c1"># Random number generator.</span>
<span class="n">rng</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">default_rng</span><span class="p">(</span><span class="n">SEED</span><span class="p">)</span>

<span class="c1"># Random function that enforce the boundaries in x/y.</span>
<span class="n">boundary_xy</span> <span class="o">=</span> <span class="k">lambda</span><span class="p">:</span> <span class="n">rng</span><span class="o">.</span><span class="n">uniform</span><span class="p">(</span><span class="o">-</span><span class="mf">15.0</span><span class="p">,</span> <span class="mf">15.0</span><span class="p">)</span>

<span class="c1"># Define the number of chromosomes.</span>
<span class="n">n_pop</span><span class="p">,</span> <span class="n">n_dim</span> <span class="o">=</span> <span class="mi">50</span><span class="p">,</span> <span class="mi">2</span>

<span class="c1"># Draw random samples for the initial points.</span>
<span class="n">xy_init</span> <span class="o">=</span> <span class="n">rng</span><span class="o">.</span><span class="n">uniform</span><span class="p">(</span><span class="o">-</span><span class="mf">15.0</span><span class="p">,</span> <span class="mf">15.0</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="p">(</span><span class="n">n_pop</span><span class="p">,</span> <span class="n">n_dim</span><span class="p">))</span>

<span class="c1"># Initial population.</span>
<span class="n">population</span> <span class="o">=</span> <span class="p">[</span><span class="n">Chromosome</span><span class="p">([</span><span class="n">Gene</span><span class="p">(</span><span class="n">xy_init</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="mi">0</span><span class="p">],</span> <span class="n">boundary_xy</span><span class="p">),</span>
                          <span class="n">Gene</span><span class="p">(</span><span class="n">xy_init</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="mi">1</span><span class="p">],</span> <span class="n">boundary_xy</span><span class="p">)],</span> <span class="n">np</span><span class="o">.</span><span class="n">nan</span><span class="p">,</span> <span class="kc">True</span><span class="p">)</span>
              <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">n_pop</span><span class="p">)]</span>

<span class="c1"># Create the StandardGA object that will carry on the optimization.</span>
<span class="n">test_GA</span> <span class="o">=</span> <span class="n">StandardGA</span><span class="p">(</span><span class="n">initial_pop</span><span class="o">=</span><span class="n">population</span><span class="p">,</span>
                     <span class="n">fit_func</span><span class="o">=</span><span class="n">fun_test2D</span><span class="p">,</span>
                     <span class="n">select_op</span><span class="o">=</span><span class="n">NeighborhoodSelector</span><span class="p">(),</span>
                     <span class="n">mutate_op</span><span class="o">=</span><span class="n">GaussianMutator</span><span class="p">(</span><span class="n">lower_val</span><span class="o">=-</span><span class="mf">15.0</span><span class="p">,</span>
                                               <span class="n">upper_val</span><span class="o">=+</span><span class="mf">15.0</span><span class="p">),</span>
                     <span class="n">crossx_op</span><span class="o">=</span><span class="n">BlendCrossover</span><span class="p">(</span><span class="n">lower_val</span><span class="o">=-</span><span class="mf">15.0</span><span class="p">,</span>
                                              <span class="n">upper_val</span><span class="o">=+</span><span class="mf">15.0</span><span class="p">))</span>
</pre></div>
</div>
</section>
<section id="step-4-run-the-optimization">
<h2>Step 4: Run the optimization<a class="headerlink" href="#step-4-run-the-optimization" title="Link to this heading"></a></h2>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">test_GA</span><span class="p">(</span><span class="n">epochs</span><span class="o">=</span><span class="mi">500</span><span class="p">,</span> <span class="n">elitism</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">shuffle</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="step-5-extract-the-data-for-analysis-and-plotting">
<h2>Step 5: Extract the data for analysis and plotting<a class="headerlink" href="#step-5-extract-the-data-for-analysis-and-plotting" title="Link to this heading"></a></h2>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># Extract the data values as &#39;x&#39; and &#39;y&#39;, for parsimony.</span>
<span class="n">x_opt</span><span class="p">,</span> <span class="n">y_opt</span> <span class="o">=</span> <span class="n">test_GA</span><span class="o">.</span><span class="n">best_chromosome</span><span class="p">()</span><span class="o">.</span><span class="n">values</span><span class="p">()</span>

<span class="c1"># Compute the final objective functions.</span>
<span class="n">f_opt</span> <span class="o">=</span> <span class="n">negative_log_pdfx</span><span class="p">([</span><span class="n">x_opt</span><span class="p">,</span> <span class="n">y_opt</span><span class="p">])</span>

<span class="c1"># Print the results.</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;x=</span><span class="si">{</span><span class="n">x_opt</span><span class="si">:</span><span class="s2">.5f</span><span class="si">}</span><span class="s2">, y=</span><span class="si">{</span><span class="n">y_opt</span><span class="si">:</span><span class="s2">.5f</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">end</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\n\n</span><span class="s1">&#39;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;f_opt(x, y) = </span><span class="si">{</span><span class="n">f_opt</span><span class="si">:</span><span class="s2">.5f</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="step-6-visualize-the-solutions">
<h2>Step 6: Visualize the solutions<a class="headerlink" href="#step-6-visualize-the-solutions" title="Link to this heading"></a></h2>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># Store the additional solutions.</span>
<span class="n">best_n</span> <span class="o">=</span> <span class="p">[]</span>

<span class="k">for</span> <span class="n">p</span> <span class="ow">in</span> <span class="n">test_GA</span><span class="o">.</span><span class="n">best_n</span><span class="p">(</span><span class="n">n</span><span class="o">=</span><span class="n">n_pop</span><span class="o">//</span><span class="mi">2</span><span class="p">):</span>
    <span class="n">best_n</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">p</span><span class="o">.</span><span class="n">values</span><span class="p">())</span>

<span class="n">best_n</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">best_n</span><span class="p">)</span>

<span class="c1"># Prepare the plot of the real density.</span>
<span class="n">x</span><span class="p">,</span> <span class="n">y</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">mgrid</span><span class="p">[</span><span class="o">-</span><span class="mi">15</span><span class="p">:</span><span class="mi">15</span><span class="p">:</span><span class="mf">0.01</span><span class="p">,</span> <span class="o">-</span><span class="mi">15</span><span class="p">:</span><span class="mi">15</span><span class="p">:</span><span class="mf">0.01</span><span class="p">]</span>

<span class="c1"># Stack the position of the grid together.</span>
<span class="n">pos</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">dstack</span><span class="p">((</span><span class="n">x</span><span class="p">,</span> <span class="n">y</span><span class="p">))</span>

<span class="c1"># First plot the contour of the &quot;true&quot; function.</span>
<span class="n">plt</span><span class="o">.</span><span class="n">contour</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">y</span><span class="p">,</span> <span class="n">negative_log_pdfx</span><span class="p">(</span><span class="n">pos</span><span class="p">),</span> <span class="n">levels</span><span class="o">=</span><span class="mi">30</span><span class="p">)</span>

<span class="c1"># Add the three modes.</span>
<span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="o">+</span><span class="mf">0.0</span><span class="p">,</span> <span class="o">-</span><span class="mf">1.0</span><span class="p">,</span> <span class="s2">&quot;r+&quot;</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s2">&quot;$\mu_1$&quot;</span><span class="p">,</span> <span class="n">markersize</span><span class="o">=</span><span class="mi">12</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="o">-</span><span class="mf">4.0</span><span class="p">,</span> <span class="o">-</span><span class="mf">6.0</span><span class="p">,</span> <span class="s2">&quot;r+&quot;</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s2">&quot;$\mu_2$&quot;</span><span class="p">,</span> <span class="n">markersize</span><span class="o">=</span><span class="mi">12</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="o">-</span><span class="mf">10.0</span><span class="p">,</span> <span class="mf">5.0</span><span class="p">,</span> <span class="s2">&quot;r+&quot;</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s2">&quot;$\mu_3$&quot;</span><span class="p">,</span> <span class="n">markersize</span><span class="o">=</span><span class="mi">12</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="mf">5.0</span><span class="p">,</span> <span class="o">-</span><span class="mf">10.0</span><span class="p">,</span> <span class="s2">&quot;r+&quot;</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s2">&quot;$\mu_4$&quot;</span><span class="p">,</span> <span class="n">markersize</span><span class="o">=</span><span class="mi">12</span><span class="p">)</span>

<span class="c1"># Plot the optimal GA.</span>
<span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">x_opt</span><span class="p">,</span> <span class="n">y_opt</span><span class="p">,</span> <span class="s2">&quot;kx&quot;</span><span class="p">,</span> <span class="n">markersize</span><span class="o">=</span><span class="mi">12</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s2">&quot;PyGenAlgo optimal&quot;</span><span class="p">)</span>

<span class="c1"># Plot the best_n.</span>
<span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">best_n</span><span class="p">[:,</span> <span class="mi">0</span><span class="p">],</span> <span class="n">best_n</span><span class="p">[:,</span> <span class="mi">1</span><span class="p">],</span> <span class="s2">&quot;ko&quot;</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.5</span><span class="p">,</span> <span class="n">markersize</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s2">&quot;PyGenAlgo solutions&quot;</span><span class="p">)</span>

<span class="c1"># Add labels.</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s2">&quot;$x$&quot;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">14</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s2">&quot;$y$&quot;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">14</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s2">&quot;$-log[\sum_{i=1}^4{\cal N}(\mu_i, \Sigma_i)]$&quot;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">14</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">legend</span><span class="p">()</span>

<span class="c1"># Final setup.</span>
<span class="n">plt</span><span class="o">.</span><span class="n">colorbar</span><span class="p">()</span>
<span class="n">plt</span><span class="o">.</span><span class="n">grid</span><span class="p">()</span>
</pre></div>
</div>
<figure class="align-center">
<img alt="_images/gaussian_mixture_2d.png" src="_images/gaussian_mixture_2d.png" />
</figure>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="tsp_path.html" class="btn btn-neutral float-left" title="Travelling Salesman Problem" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2024, Michalis Vrettas, PhD.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>