

<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="./">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>pygenalgo.genome package &mdash; PyGenAlgo 01.70.50 documentation</title>
      <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=b86133f3" />
      <link rel="stylesheet" type="text/css" href="_static/css/theme.css?v=e59714d7" />

  
      <script src="_static/jquery.js?v=5d32c60e"></script>
      <script src="_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
      <script src="_static/documentation_options.js?v=d0c4109e"></script>
      <script src="_static/doctools.js?v=9a2dae69"></script>
      <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="pygenalgo.operators package" href="pygenalgo.operators.html" />
    <link rel="prev" title="pygenalgo.engines package" href="pygenalgo.engines.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="index.html" class="icon icon-home">
            PyGenAlgo
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="reference internal" href="modules.html">pygenalgo</a><ul class="current">
<li class="toctree-l2 current"><a class="reference internal" href="pygenalgo.html">pygenalgo package</a><ul class="current">
<li class="toctree-l3 current"><a class="reference internal" href="pygenalgo.html#subpackages">Subpackages</a><ul class="current">
<li class="toctree-l4"><a class="reference internal" href="pygenalgo.engines.html">pygenalgo.engines package</a></li>
<li class="toctree-l4 current"><a class="current reference internal" href="#">pygenalgo.genome package</a></li>
<li class="toctree-l4"><a class="reference internal" href="pygenalgo.operators.html">pygenalgo.operators package</a></li>
<li class="toctree-l4"><a class="reference internal" href="pygenalgo.utils.html">pygenalgo.utils package</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="pygenalgo.html#module-pygenalgo">Module contents</a></li>
</ul>
</li>
</ul>
</li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">PyGenAlgo</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="modules.html">pygenalgo</a></li>
          <li class="breadcrumb-item"><a href="pygenalgo.html">pygenalgo package</a></li>
      <li class="breadcrumb-item active">pygenalgo.genome package</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/pygenalgo.genome.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="pygenalgo-genome-package">
<h1>pygenalgo.genome package<a class="headerlink" href="#pygenalgo-genome-package" title="Link to this heading"></a></h1>
<section id="submodules">
<h2>Submodules<a class="headerlink" href="#submodules" title="Link to this heading"></a></h2>
</section>
<section id="module-pygenalgo.genome.chromosome">
<span id="pygenalgo-genome-chromosome-module"></span><h2>pygenalgo.genome.chromosome module<a class="headerlink" href="#module-pygenalgo.genome.chromosome" title="Link to this heading"></a></h2>
<dl class="py class">
<dt class="sig sig-object py" id="pygenalgo.genome.chromosome.Chromosome">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">pygenalgo.genome.chromosome.</span></span><span class="sig-name descname"><span class="pre">Chromosome</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">genome</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span><span class="p"><span class="pre">[</span></span><a class="reference internal" href="#pygenalgo.genome.gene.Gene" title="pygenalgo.genome.gene.Gene"><span class="pre">Gene</span></a><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fitness</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">nan</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">valid</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pygenalgo/genome/chromosome.html#Chromosome"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pygenalgo.genome.chromosome.Chromosome" title="Link to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Description:</p>
<blockquote>
<div><p>Implements a dataclass for the Chromosome entity. This class is responsible
for holding the individual solution(s), of the optimization problem, during
the evolution process.</p>
</div></blockquote>
<dl class="py method">
<dt class="sig sig-object py" id="pygenalgo.genome.chromosome.Chromosome.clone">
<span class="sig-name descname"><span class="pre">clone</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference internal" href="#pygenalgo.genome.chromosome.Chromosome" title="pygenalgo.genome.chromosome.Chromosome"><span class="pre">Chromosome</span></a></span></span><a class="reference internal" href="_modules/pygenalgo/genome/chromosome.html#Chromosome.clone"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pygenalgo.genome.chromosome.Chromosome.clone" title="Link to this definition"></a></dt>
<dd><p>Makes a duplicate of the self object
by deep-coping only the genome field.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>a “deep-copy” of the object.</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="pygenalgo.genome.chromosome.Chromosome.fitness">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">fitness</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">float</span></em><a class="headerlink" href="#pygenalgo.genome.chromosome.Chromosome.fitness" title="Link to this definition"></a></dt>
<dd><p>Accessor of the fitness value of the chromosome.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>the fitness (float) of the genome.</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="pygenalgo.genome.chromosome.Chromosome.genome">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">genome</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">list</span><span class="p"><span class="pre">[</span></span><a class="reference internal" href="#pygenalgo.genome.gene.Gene" title="pygenalgo.genome.gene.Gene"><span class="pre">Gene</span></a><span class="p"><span class="pre">]</span></span></em><a class="headerlink" href="#pygenalgo.genome.chromosome.Chromosome.genome" title="Link to this definition"></a></dt>
<dd><p>Accessor of the genome list of the chromosome.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>the list (of Genes) of the chromosome.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pygenalgo.genome.chromosome.Chromosome.hamming_distance">
<span class="sig-name descname"><span class="pre">hamming_distance</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">int</span></span></span><a class="reference internal" href="_modules/pygenalgo/genome/chromosome.html#Chromosome.hamming_distance"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pygenalgo.genome.chromosome.Chromosome.hamming_distance" title="Link to this definition"></a></dt>
<dd><p>Compute the Hamming distance of the “self” object, with the
“other” chromosome. In practice, it’s the number of positions
at which the corresponding genes are different.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>other</strong> – (Chromosome) to compare the Hamming distance.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>(int) the number of dissimilarities between the two</p>
</dd>
</dl>
<p>input chromosomes.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pygenalgo.genome.chromosome.Chromosome.invalidate_fitness">
<span class="sig-name descname"><span class="pre">invalidate_fitness</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="reference internal" href="_modules/pygenalgo/genome/chromosome.html#Chromosome.invalidate_fitness"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pygenalgo.genome.chromosome.Chromosome.invalidate_fitness" title="Link to this definition"></a></dt>
<dd><p>Invalidates the fitness of the chromosome
by setting the value to NaN. This is used
during the evolution process (mutation).</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>None.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pygenalgo.genome.chromosome.Chromosome.is_genome_valid">
<span class="sig-name descname"><span class="pre">is_genome_valid</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">bool</span></span></span><a class="reference internal" href="_modules/pygenalgo/genome/chromosome.html#Chromosome.is_genome_valid"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pygenalgo.genome.chromosome.Chromosome.is_genome_valid" title="Link to this definition"></a></dt>
<dd><p>Checks the validity of the whole chromosome, by
calling individually all genes is_valid method.</p>
<p>In addition, it “double-checks” that all entries
in the genome are of type ‘Gene’.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>True if ALL genes are valid, else False.</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="pygenalgo.genome.chromosome.Chromosome.valid">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">valid</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">bool</span></em><a class="headerlink" href="#pygenalgo.genome.chromosome.Chromosome.valid" title="Link to this definition"></a></dt>
<dd><p>Accessor (getter) of the validity parameter.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>the valid value.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pygenalgo.genome.chromosome.Chromosome.values">
<span class="sig-name descname"><span class="pre">values</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">list</span></span></span><a class="reference internal" href="_modules/pygenalgo/genome/chromosome.html#Chromosome.values"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pygenalgo.genome.chromosome.Chromosome.values" title="Link to this definition"></a></dt>
<dd><p>Returns the gene values of the chromosome
as list.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>the list values of the genome.</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

</section>
<section id="module-pygenalgo.genome.gene">
<span id="pygenalgo-genome-gene-module"></span><h2>pygenalgo.genome.gene module<a class="headerlink" href="#module-pygenalgo.genome.gene" title="Link to this heading"></a></h2>
<dl class="py class">
<dt class="sig sig-object py" id="pygenalgo.genome.gene.Gene">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">pygenalgo.genome.gene.</span></span><span class="sig-name descname"><span class="pre">Gene</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">datum</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Any</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">func</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Callable</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">valid</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pygenalgo/genome/gene.html#Gene"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pygenalgo.genome.gene.Gene" title="Link to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Description:</p>
<blockquote>
<div><p>This is the main class that encodes the data of a single Gene in the chromosome.
The class encapsulates not only the data, but also the way that this gene can be
mutated using a random function. This Gene can be from a single ‘bit’ to a whole
image. This way provides us with flexibility to parameterize the chromosome with
different “kinds of genes” each one responsible for a specific function.</p>
</div></blockquote>
<dl class="py method">
<dt class="sig sig-object py" id="pygenalgo.genome.gene.Gene.clone">
<span class="sig-name descname"><span class="pre">clone</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference internal" href="#pygenalgo.genome.gene.Gene" title="pygenalgo.genome.gene.Gene"><span class="pre">Gene</span></a></span></span><a class="reference internal" href="_modules/pygenalgo/genome/gene.html#Gene.clone"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pygenalgo.genome.gene.Gene.clone" title="Link to this definition"></a></dt>
<dd><p>Makes a duplicate of the self object
by deep-coping only the datum field.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>a “deep-copy” of the object.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pygenalgo.genome.gene.Gene.flip">
<span class="sig-name descname"><span class="pre">flip</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="reference internal" href="_modules/pygenalgo/genome/gene.html#Gene.flip"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pygenalgo.genome.gene.Gene.flip" title="Link to this definition"></a></dt>
<dd><p>This method flips the value of the gene data. It is used only by the
FlipMutator operator for problems where the chromosome is represented
by a list of bits.</p>
<blockquote>
<div><ol class="lowerroman simple">
<li><p>1 -&gt; 0</p></li>
</ol>
</div></blockquote>
<ol class="lowerroman simple" start="2">
<li><p>0 -&gt; 1</p></li>
</ol>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>None.</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="pygenalgo.genome.gene.Gene.func">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">func</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">Callable</span></em><a class="headerlink" href="#pygenalgo.genome.gene.Gene.func" title="Link to this definition"></a></dt>
<dd><p>Accessor (getter) of the gene function reference.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>the _func reference.</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="pygenalgo.genome.gene.Gene.is_valid">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">is_valid</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">bool</span></em><a class="headerlink" href="#pygenalgo.genome.gene.Gene.is_valid" title="Link to this definition"></a></dt>
<dd><p>Accessor (getter) of the validity parameter.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>the valid value.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pygenalgo.genome.gene.Gene.random">
<span class="sig-name descname"><span class="pre">random</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="reference internal" href="_modules/pygenalgo/genome/gene.html#Gene.random"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pygenalgo.genome.gene.Gene.random" title="Link to this definition"></a></dt>
<dd><p>This method should be different for each type of Gene. It describes
how a specific type of Gene creates a random version of itself. The
main  idea is that inside the Chromosome, each Gene can represent a
very different concept of the  problem solution, so its Gene should
have its own way to perform random mutation.</p>
<p>This way by calling on the random() method, each Gene will know how
to mutate itself without breaking any rules/constraints.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>None.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pygenalgo.genome.gene.Gene.set_seed">
<em class="property"><span class="pre">classmethod</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">set_seed</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">new_seed</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="reference internal" href="_modules/pygenalgo/genome/gene.html#Gene.set_seed"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pygenalgo.genome.gene.Gene.set_seed" title="Link to this definition"></a></dt>
<dd><p>Sets a new seed for the random number generator.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>new_seed</strong> – New seed value (default=None).</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>None.</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="pygenalgo.genome.gene.Gene.value">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">value</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">Any</span></em><a class="headerlink" href="#pygenalgo.genome.gene.Gene.value" title="Link to this definition"></a></dt>
<dd><p>Accessor (getter) of the data reference.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>the datum value.</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

</section>
<section id="module-pygenalgo.genome">
<span id="module-contents"></span><h2>Module contents<a class="headerlink" href="#module-pygenalgo.genome" title="Link to this heading"></a></h2>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="pygenalgo.engines.html" class="btn btn-neutral float-left" title="pygenalgo.engines package" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="pygenalgo.operators.html" class="btn btn-neutral float-right" title="pygenalgo.operators package" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2024, Michalis Vrettas, PhD.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>