

<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="./">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Engines &mdash; PyGenAlgo 01.70.50 documentation</title>
      <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=b86133f3" />
      <link rel="stylesheet" type="text/css" href="_static/css/theme.css?v=e59714d7" />

  
      <script src="_static/jquery.js?v=5d32c60e"></script>
      <script src="_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
      <script src="_static/documentation_options.js?v=d0c4109e"></script>
      <script src="_static/doctools.js?v=9a2dae69"></script>
      <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Genome" href="pygenalgo.genome.html" />
    <link rel="prev" title="PyGenAlgo" href="pygenalgo.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="index.html" class="icon icon-home">
            PyGenAlgo
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="reference internal" href="modules.html">Modules</a><ul class="current">
<li class="toctree-l2 current"><a class="reference internal" href="pygenalgo.html">PyGenAlgo</a><ul class="current">
<li class="toctree-l3 current"><a class="reference internal" href="pygenalgo.html#subpackages">Subpackages</a><ul class="current">
<li class="toctree-l4 current"><a class="current reference internal" href="#">Engines</a></li>
<li class="toctree-l4"><a class="reference internal" href="pygenalgo.genome.html">Genome</a></li>
<li class="toctree-l4"><a class="reference internal" href="pygenalgo.operators.html">Operators</a></li>
<li class="toctree-l4"><a class="reference internal" href="pygenalgo.utils.html">Utils</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="pygenalgo.html#module-pygenalgo">Module contents</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="examples.html">Examples</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">PyGenAlgo</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="modules.html">Modules</a></li>
          <li class="breadcrumb-item"><a href="pygenalgo.html">PyGenAlgo</a></li>
      <li class="breadcrumb-item active">Engines</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/pygenalgo.engines.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="engines">
<h1>Engines<a class="headerlink" href="#engines" title="Link to this heading"></a></h1>
<section id="submodules">
<h2>Submodules<a class="headerlink" href="#submodules" title="Link to this heading"></a></h2>
</section>
<section id="module-pygenalgo.engines.generic_ga">
<span id="pygenalgo-engines-generic-ga-module"></span><h2>pygenalgo.engines.generic_ga module<a class="headerlink" href="#module-pygenalgo.engines.generic_ga" title="Link to this heading"></a></h2>
<dl class="py class">
<dt class="sig sig-object py" id="pygenalgo.engines.generic_ga.GenericGA">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">pygenalgo.engines.generic_ga.</span></span><span class="sig-name descname"><span class="pre">GenericGA</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">initial_pop</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span><span class="p"><span class="pre">[</span></span><a class="reference internal" href="pygenalgo.genome.html#pygenalgo.genome.chromosome.Chromosome" title="pygenalgo.genome.chromosome.Chromosome"><span class="pre">Chromosome</span></a><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fit_func</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Callable</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">select_op</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="pygenalgo.operators.selection.html#pygenalgo.operators.selection.select_operator.SelectionOperator" title="pygenalgo.operators.selection.select_operator.SelectionOperator"><span class="pre">SelectionOperator</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mutate_op</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="pygenalgo.operators.mutation.html#pygenalgo.operators.mutation.mutate_operator.MutationOperator" title="pygenalgo.operators.mutation.mutate_operator.MutationOperator"><span class="pre">MutationOperator</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">crossx_op</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="pygenalgo.operators.crossover.html#pygenalgo.operators.crossover.crossover_operator.CrossoverOperator" title="pygenalgo.operators.crossover.crossover_operator.CrossoverOperator"><span class="pre">CrossoverOperator</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_cpus</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pygenalgo/engines/generic_ga.html#GenericGA"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pygenalgo.engines.generic_ga.GenericGA" title="Link to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Description:</p>
<blockquote>
<div><p>Generic GA class models the interface of a specific genetic algorithm model
(or engine). It provides the common variables and functionality that all GA
models should share.</p>
</div></blockquote>
<dl class="py attribute">
<dt class="sig sig-object py" id="pygenalgo.engines.generic_ga.GenericGA.MAX_CPUs">
<span class="sig-name descname"><span class="pre">MAX_CPUs</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">int</span></em><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">4</span></em><a class="headerlink" href="#pygenalgo.engines.generic_ga.GenericGA.MAX_CPUs" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pygenalgo.engines.generic_ga.GenericGA.adapt_probabilities">
<span class="sig-name descname"><span class="pre">adapt_probabilities</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">threshold</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">bool</span></span></span><a class="reference internal" href="_modules/pygenalgo/engines/generic_ga.html#GenericGA.adapt_probabilities"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pygenalgo.engines.generic_ga.GenericGA.adapt_probabilities" title="Link to this definition"></a></dt>
<dd><p>This method is used (optionally) to adjust simultaneously the crossover
and mutation parameters of the GenericGA object.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>threshold</strong> – This float parameter is used to determine whether we are
going to increase or decrease the crossover and mutation
parameters.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>True if the parameters have changed, else False.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pygenalgo.engines.generic_ga.GenericGA.best_chromosome">
<span class="sig-name descname"><span class="pre">best_chromosome</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference internal" href="pygenalgo.genome.html#pygenalgo.genome.chromosome.Chromosome" title="pygenalgo.genome.chromosome.Chromosome"><span class="pre">Chromosome</span></a></span></span><a class="reference internal" href="_modules/pygenalgo/engines/generic_ga.html#GenericGA.best_chromosome"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pygenalgo.engines.generic_ga.GenericGA.best_chromosome" title="Link to this definition"></a></dt>
<dd><p>Auxiliary method that returns the chromosome with the
highest fitness value. Safeguarded with ignoring NaNs.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Return the chromosome with the highest fitness.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pygenalgo.engines.generic_ga.GenericGA.best_n">
<span class="sig-name descname"><span class="pre">best_n</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">n</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">1</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">list</span><span class="p"><span class="pre">[</span></span><a class="reference internal" href="pygenalgo.genome.html#pygenalgo.genome.chromosome.Chromosome" title="pygenalgo.genome.chromosome.Chromosome"><span class="pre">Chromosome</span></a><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="_modules/pygenalgo/engines/generic_ga.html#GenericGA.best_n"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pygenalgo.engines.generic_ga.GenericGA.best_n" title="Link to this definition"></a></dt>
<dd><p>Auxiliary method that returns the best ‘n’ chromosomes
with the highest fitness value.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>n</strong> – the number of the best chromosomes. Default = 1.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Return the ‘n’ chromosomes with the highest fitness.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pygenalgo.engines.generic_ga.GenericGA.clear_all">
<span class="sig-name descname"><span class="pre">clear_all</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="reference internal" href="_modules/pygenalgo/engines/generic_ga.html#GenericGA.clear_all"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pygenalgo.engines.generic_ga.GenericGA.clear_all" title="Link to this definition"></a></dt>
<dd><p>Make sure all the genetic operator counters and the stats
are cleared. This reset everything before each run().</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>None.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pygenalgo.engines.generic_ga.GenericGA.crossover_mutate">
<span class="sig-name descname"><span class="pre">crossover_mutate</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">input_population</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span><span class="p"><span class="pre">[</span></span><a class="reference internal" href="pygenalgo.genome.html#pygenalgo.genome.chromosome.Chromosome" title="pygenalgo.genome.chromosome.Chromosome"><span class="pre">Chromosome</span></a><span class="p"><span class="pre">]</span></span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="reference internal" href="_modules/pygenalgo/engines/generic_ga.html#GenericGA.crossover_mutate"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pygenalgo.engines.generic_ga.GenericGA.crossover_mutate" title="Link to this definition"></a></dt>
<dd><p>This is an auxiliary method that combines the crossover and mutation
operations in one call. Since these operations happen in place the
‘input_population’ will be modified directly.</p>
<p>This method should be called AFTER the selection of the parents that
have been selected for breeding.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>input_population</strong> – this is the population that we will apply
the two genetic operators.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>None.</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="pygenalgo.engines.generic_ga.GenericGA.crossx_op">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">crossx_op</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference internal" href="pygenalgo.operators.crossover.html#pygenalgo.operators.crossover.crossover_operator.CrossoverOperator" title="pygenalgo.operators.crossover.crossover_operator.CrossoverOperator"><span class="pre">CrossoverOperator</span></a></em><a class="headerlink" href="#pygenalgo.engines.generic_ga.GenericGA.crossx_op" title="Link to this definition"></a></dt>
<dd><p>Accessor method that returns the crossover operator reference.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>the CrossoverOperator.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pygenalgo.engines.generic_ga.GenericGA.evaluate_fitness">
<span class="sig-name descname"><span class="pre">evaluate_fitness</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">input_population</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span><span class="p"><span class="pre">[</span></span><a class="reference internal" href="pygenalgo.genome.html#pygenalgo.genome.chromosome.Chromosome" title="pygenalgo.genome.chromosome.Chromosome"><span class="pre">Chromosome</span></a><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="n"><span class="pre">parallel_mode</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">backend</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'threading'</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">tuple</span><span class="p"><span class="pre">[</span></span><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">float</span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">bool</span><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="_modules/pygenalgo/engines/generic_ga.html#GenericGA.evaluate_fitness"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pygenalgo.engines.generic_ga.GenericGA.evaluate_fitness" title="Link to this definition"></a></dt>
<dd><p>Evaluate all the chromosomes of the input list with the custom
fitness function. The parallel_mode is optional. Moreover, the
default backend is “threading”, but in the IslandModelGA it is
better to select “loky”.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>input_population</strong> – (list) The population of Chromosomes
that we want to evaluate their fitness.</p></li>
<li><p><strong>parallel_mode</strong> – (bool) Enables parallel computation of
the fitness function.</p></li>
<li><p><strong>backend</strong> – (str) Backend for the parallel Joblib framework.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>a list with the fitness values and the found solution flag.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pygenalgo.engines.generic_ga.GenericGA.f_eval_increase_by">
<span class="sig-name descname"><span class="pre">f_eval_increase_by</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">new_counts</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="reference internal" href="_modules/pygenalgo/engines/generic_ga.html#GenericGA.f_eval_increase_by"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pygenalgo.engines.generic_ga.GenericGA.f_eval_increase_by" title="Link to this definition"></a></dt>
<dd><p>Utility method to allow the ‘_f_eval’ to be updated with
new counts outside the main class. This can happen during
gene correction.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>new_counts</strong> – the new value of ‘counts’ that we want
to add on the current f_eval.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>None.</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="pygenalgo.engines.generic_ga.GenericGA.f_evals">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">f_evals</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">int</span></em><a class="headerlink" href="#pygenalgo.engines.generic_ga.GenericGA.f_evals" title="Link to this definition"></a></dt>
<dd><p>Accessor method that returns the value of the f_eval.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>(int) the counted number of function evaluations.</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="pygenalgo.engines.generic_ga.GenericGA.fitness_func">
<span class="sig-name descname"><span class="pre">fitness_func</span></span><a class="headerlink" href="#pygenalgo.engines.generic_ga.GenericGA.fitness_func" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pygenalgo.engines.generic_ga.GenericGA.individual_fitness">
<span class="sig-name descname"><span class="pre">individual_fitness</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">index</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">float</span></span></span><a class="reference internal" href="_modules/pygenalgo/engines/generic_ga.html#GenericGA.individual_fitness"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pygenalgo.engines.generic_ga.GenericGA.individual_fitness" title="Link to this definition"></a></dt>
<dd><p>Get the fitness value of an individual member of the population.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>index</strong> – Position of the individual in the population.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The fitness value (float).</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="pygenalgo.engines.generic_ga.GenericGA.iteration">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">iteration</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">int</span></em><a class="headerlink" href="#pygenalgo.engines.generic_ga.GenericGA.iteration" title="Link to this definition"></a></dt>
<dd><p>Accessor (getter) of the iteration parameter.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>the iteration value.</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="pygenalgo.engines.generic_ga.GenericGA.mutate_op">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">mutate_op</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference internal" href="pygenalgo.operators.mutation.html#pygenalgo.operators.mutation.mutate_operator.MutationOperator" title="pygenalgo.operators.mutation.mutate_operator.MutationOperator"><span class="pre">MutationOperator</span></a></em><a class="headerlink" href="#pygenalgo.engines.generic_ga.GenericGA.mutate_op" title="Link to this definition"></a></dt>
<dd><p>Accessor method that returns the mutation operator reference.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>the MutationOperator.</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="pygenalgo.engines.generic_ga.GenericGA.n_cpus">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">n_cpus</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">int</span></em><a class="headerlink" href="#pygenalgo.engines.generic_ga.GenericGA.n_cpus" title="Link to this definition"></a></dt>
<dd><p>Accessor method that returns the number of CPUs.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>the n_cpus.</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="pygenalgo.engines.generic_ga.GenericGA.population">
<span class="sig-name descname"><span class="pre">population</span></span><a class="headerlink" href="#pygenalgo.engines.generic_ga.GenericGA.population" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pygenalgo.engines.generic_ga.GenericGA.population_fitness">
<span class="sig-name descname"><span class="pre">population_fitness</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">float</span><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="_modules/pygenalgo/engines/generic_ga.html#GenericGA.population_fitness"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pygenalgo.engines.generic_ga.GenericGA.population_fitness" title="Link to this definition"></a></dt>
<dd><p>Get the fitness values of all the population.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>A list with all the fitness values.</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="pygenalgo.engines.generic_ga.GenericGA.rng_GA">
<span class="sig-name descname"><span class="pre">rng_GA</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">Generator</span></em><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">Generator(PCG64)</span> <span class="pre">at</span> <span class="pre">0x1404B09E0</span></em><a class="headerlink" href="#pygenalgo.engines.generic_ga.GenericGA.rng_GA" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pygenalgo.engines.generic_ga.GenericGA.run">
<span class="sig-name descname"><span class="pre">run</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="reference internal" href="_modules/pygenalgo/engines/generic_ga.html#GenericGA.run"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pygenalgo.engines.generic_ga.GenericGA.run" title="Link to this definition"></a></dt>
<dd><p>Main method of the Generic GA class, that implements the evolutionary routine.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="pygenalgo.engines.generic_ga.GenericGA.select_op">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">select_op</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference internal" href="pygenalgo.operators.selection.html#pygenalgo.operators.selection.select_operator.SelectionOperator" title="pygenalgo.operators.selection.select_operator.SelectionOperator"><span class="pre">SelectionOperator</span></a></em><a class="headerlink" href="#pygenalgo.engines.generic_ga.GenericGA.select_op" title="Link to this definition"></a></dt>
<dd><p>Accessor method that returns the selection operator reference.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>the SelectionOperator.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pygenalgo.engines.generic_ga.GenericGA.set_seed">
<em class="property"><span class="pre">classmethod</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">set_seed</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">new_seed</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="reference internal" href="_modules/pygenalgo/engines/generic_ga.html#GenericGA.set_seed"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pygenalgo.engines.generic_ga.GenericGA.set_seed" title="Link to this definition"></a></dt>
<dd><p>Sets a new seed for the random number generator.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>new_seed</strong> – New seed value (default=None).</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>None.</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="pygenalgo.engines.generic_ga.GenericGA.stats">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">stats</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">dict</span></em><a class="headerlink" href="#pygenalgo.engines.generic_ga.GenericGA.stats" title="Link to this definition"></a></dt>
<dd><p>Accessor method that returns the ‘stats’ dictionary.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>the dictionary with the statistics from the run.</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

</section>
<section id="module-pygenalgo.engines.island_model_ga">
<span id="pygenalgo-engines-island-model-ga-module"></span><h2>pygenalgo.engines.island_model_ga module<a class="headerlink" href="#module-pygenalgo.engines.island_model_ga" title="Link to this heading"></a></h2>
<dl class="py class">
<dt class="sig sig-object py" id="pygenalgo.engines.island_model_ga.IslandModelGA">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">pygenalgo.engines.island_model_ga.</span></span><span class="sig-name descname"><span class="pre">IslandModelGA</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">num_islands</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">migrate_op</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="pygenalgo.operators.migration.html#pygenalgo.operators.migration.migration_operator.MigrationOperator" title="pygenalgo.operators.migration.migration_operator.MigrationOperator"><span class="pre">MigrationOperator</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pygenalgo/engines/island_model_ga.html#IslandModelGA"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pygenalgo.engines.island_model_ga.IslandModelGA" title="Link to this definition"></a></dt>
<dd><p>Bases: <a class="reference internal" href="#pygenalgo.engines.generic_ga.GenericGA" title="pygenalgo.engines.generic_ga.GenericGA"><code class="xref py py-class docutils literal notranslate"><span class="pre">GenericGA</span></code></a></p>
<p>Description:</p>
<blockquote>
<div><p>In Island Model GA we run in parallel a number of “islands”, each one evolving its own
(sub)-population. Optionally we can allow “migration”, among the best individuals from
each island.</p>
</div></blockquote>
<dl class="py property">
<dt class="sig sig-object py" id="pygenalgo.engines.island_model_ga.IslandModelGA.migrate_op">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">migrate_op</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference internal" href="pygenalgo.operators.migration.html#pygenalgo.operators.migration.migration_operator.MigrationOperator" title="pygenalgo.operators.migration.migration_operator.MigrationOperator"><span class="pre">MigrationOperator</span></a></em><a class="headerlink" href="#pygenalgo.engines.island_model_ga.IslandModelGA.migrate_op" title="Link to this definition"></a></dt>
<dd><p>Accessor method that returns the migration operator reference.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>the MigrationOperator.</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="pygenalgo.engines.island_model_ga.IslandModelGA.num_islands">
<span class="sig-name descname"><span class="pre">num_islands</span></span><a class="headerlink" href="#pygenalgo.engines.island_model_ga.IslandModelGA.num_islands" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pygenalgo.engines.island_model_ga.IslandModelGA.print_migration_stats">
<span class="sig-name descname"><span class="pre">print_migration_stats</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="reference internal" href="_modules/pygenalgo/engines/island_model_ga.html#IslandModelGA.print_migration_stats"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pygenalgo.engines.island_model_ga.IslandModelGA.print_migration_stats" title="Link to this definition"></a></dt>
<dd><p>Print the migration operators stats.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>None.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pygenalgo.engines.island_model_ga.IslandModelGA.run">
<span class="sig-name descname"><span class="pre">run</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">epochs</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">500</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">correction</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">elitism</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">f_tol</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">allow_migration</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_periods</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">10</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">adapt_probs</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">shuffle</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">f_max_eval</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbose</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="reference internal" href="_modules/pygenalgo/engines/island_model_ga.html#IslandModelGA.run"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pygenalgo.engines.island_model_ga.IslandModelGA.run" title="Link to this definition"></a></dt>
<dd><p>Main method of the IslandModelGA class, that implements the evolutionary routine.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>epochs</strong> – (int) maximum number of iterations in the evolution process.</p></li>
<li><p><strong>correction</strong> – (bool) flag that if set to ‘True’ will check the validity of
the population (at the gene level) and attempt to correct the
genome by calling the random() method of the flawed gene.</p></li>
<li><p><strong>elitism</strong> – (bool) flag that enables elitism. If set to True then the chromosome
with the highest fitness will always be copied to the next generation
(unaltered).</p></li>
<li><p><strong>f_tol</strong> – (float) tolerance in the difference between the average values of two
consecutive populations. It is used to determine the convergence of the
population. If this value is None (default) the algorithm will terminate
using the epochs value.</p></li>
<li><p><strong>allow_migration</strong> – (bool) flag that if set to ‘True’ will allow the migration
of the best individuals among the different islands.</p></li>
<li><p><strong>n_periods</strong> – (int) the number of times that we will break the main evolution
to allow for chromosomes to migrate. NB: This setting is active
only when the option allow_migration == True. Otherwise, is ignored.</p></li>
<li><p><strong>adapt_probs</strong> – (bool) If enabled (set to True), it will allow the crossover and
mutation probabilities to adapt according to the convergence of
the population to a single solution. Default is set to False.</p></li>
<li><p><strong>shuffle</strong> – (bool) If enabled (set to True), it will shuffle the population before
the application of the crossover and mutation operations. The default
is set to True.</p></li>
<li><p><strong>f_max_eval</strong> – (int) it sets an upper limit of function evaluations. If the number
is exceeded the genetic algorithm stops. If this value is set, the
epochs will be ignored and re-adjusted to meet the new requirement.</p></li>
<li><p><strong>verbose</strong> – (bool) if ‘True’ it will display periodically information about the
current stats of the subpopulations. NB: This setting is active only
when the option allow_migration == True. Otherwise, is ignored.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>None.</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

</section>
<section id="module-pygenalgo.engines.standard_ga">
<span id="pygenalgo-engines-standard-ga-module"></span><h2>pygenalgo.engines.standard_ga module<a class="headerlink" href="#module-pygenalgo.engines.standard_ga" title="Link to this heading"></a></h2>
<dl class="py class">
<dt class="sig sig-object py" id="pygenalgo.engines.standard_ga.StandardGA">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">pygenalgo.engines.standard_ga.</span></span><span class="sig-name descname"><span class="pre">StandardGA</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pygenalgo/engines/standard_ga.html#StandardGA"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pygenalgo.engines.standard_ga.StandardGA" title="Link to this definition"></a></dt>
<dd><p>Bases: <a class="reference internal" href="#pygenalgo.engines.generic_ga.GenericGA" title="pygenalgo.engines.generic_ga.GenericGA"><code class="xref py py-class docutils literal notranslate"><span class="pre">GenericGA</span></code></a></p>
<p>Description:</p>
<blockquote>
<div><p>StandardGA model provides a basic implementation of the “GenericGA”,
which at each iteration (epoch) replaces the whole population using
the genetic operators (crossover and mutation).</p>
</div></blockquote>
<dl class="py attribute">
<dt class="sig sig-object py" id="pygenalgo.engines.standard_ga.StandardGA.fitness_func">
<span class="sig-name descname"><span class="pre">fitness_func</span></span><a class="headerlink" href="#pygenalgo.engines.standard_ga.StandardGA.fitness_func" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="pygenalgo.engines.standard_ga.StandardGA.population">
<span class="sig-name descname"><span class="pre">population</span></span><a class="headerlink" href="#pygenalgo.engines.standard_ga.StandardGA.population" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pygenalgo.engines.standard_ga.StandardGA.print_operator_stats">
<span class="sig-name descname"><span class="pre">print_operator_stats</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="reference internal" href="_modules/pygenalgo/engines/standard_ga.html#StandardGA.print_operator_stats"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pygenalgo.engines.standard_ga.StandardGA.print_operator_stats" title="Link to this definition"></a></dt>
<dd><p>Print the genetic operators stats.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>None.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pygenalgo.engines.standard_ga.StandardGA.run">
<span class="sig-name descname"><span class="pre">run</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">epochs</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">100</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">elitism</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">correction</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">f_tol</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">parallel</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">adapt_probs</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">shuffle</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">f_max_eval</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbose</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="reference internal" href="_modules/pygenalgo/engines/standard_ga.html#StandardGA.run"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pygenalgo.engines.standard_ga.StandardGA.run" title="Link to this definition"></a></dt>
<dd><p>Main method of the StandardGA class, that implements the evolutionary routine.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>epochs</strong> – (int) maximum number of iterations in the evolution process.</p></li>
<li><p><strong>elitism</strong> – (bool) flag that enables elitism. If True then the chromosome with
the highest fitness will always be copied to the next generation
(unaltered).</p></li>
<li><p><strong>correction</strong> – (bool) flag that if set to ‘True’ will check the validity of
the population (at the gene level) and attempt to correct the
genome by calling the random() method of the flawed gene.</p></li>
<li><p><strong>f_tol</strong> – (float) tolerance in the difference between the average values of two
consecutive populations. It is used to determine the convergence of the
population. If this value is None (default) the algorithm will terminate
using the epochs value.</p></li>
<li><p><strong>parallel</strong> – (bool) Flag that enables parallel computation of the fitness function.</p></li>
<li><p><strong>adapt_probs</strong> – (bool) If enabled (set to True), it will allow the crossover and
mutation probabilities to adapt according to the convergence of
the population to a single solution. Default is set to False.</p></li>
<li><p><strong>shuffle</strong> – (bool) If enabled (set to True), it will shuffle the population before
the application of the crossover and mutation operations. The default
is set to True.</p></li>
<li><p><strong>f_max_eval</strong> – (int) it sets an upper limit of function evaluations. If the number
is exceeded the genetic algorithm stops.</p></li>
<li><p><strong>verbose</strong> – (bool) if ‘True’ it will display periodically information about the
current average fitness and spread of the population.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>None.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pygenalgo.engines.standard_ga.StandardGA.update_stats">
<span class="sig-name descname"><span class="pre">update_stats</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fit_list</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">float</span><span class="p"><span class="pre">]</span></span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">tuple</span><span class="p"><span class="pre">[</span></span><span class="pre">float</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">float</span><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="_modules/pygenalgo/engines/standard_ga.html#StandardGA.update_stats"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pygenalgo.engines.standard_ga.StandardGA.update_stats" title="Link to this definition"></a></dt>
<dd><p>Update the stats dictionary with the mean/std values of the
population fitness values.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>fit_list</strong> – (list) fitness values of the population.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>the mean and std of the fitness values.</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

</section>
<section id="module-pygenalgo.engines">
<span id="module-contents"></span><h2>Module contents<a class="headerlink" href="#module-pygenalgo.engines" title="Link to this heading"></a></h2>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="pygenalgo.html" class="btn btn-neutral float-left" title="PyGenAlgo" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="pygenalgo.genome.html" class="btn btn-neutral float-right" title="Genome" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2024, Michalis Vrettas, PhD.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>