<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="./">
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.18.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>pygenalgo.operators.crossover package &mdash; PyGenAlgo 00.00.01 documentation</title>
      <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=92fd9be5" />
      <link rel="stylesheet" type="text/css" href="_static/css/theme.css?v=19f00094" />

  
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script src="_static/jquery.js?v=5d32c60e"></script>
        <script src="_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
        <script src="_static/documentation_options.js?v=d7bb8d4d"></script>
        <script src="_static/doctools.js?v=9a2dae69"></script>
        <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="pygenalgo.operators.migration package" href="pygenalgo.operators.migration.html" />
    <link rel="prev" title="pygenalgo.operators package" href="pygenalgo.operators.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="index.html" class="icon icon-home">
            PyGenAlgo
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="reference internal" href="modules.html">pygenalgo</a><ul class="current">
<li class="toctree-l2 current"><a class="reference internal" href="pygenalgo.html">pygenalgo package</a><ul class="current">
<li class="toctree-l3 current"><a class="reference internal" href="pygenalgo.html#subpackages">Subpackages</a><ul class="current">
<li class="toctree-l4"><a class="reference internal" href="pygenalgo.engines.html">pygenalgo.engines package</a></li>
<li class="toctree-l4"><a class="reference internal" href="pygenalgo.genome.html">pygenalgo.genome package</a></li>
<li class="toctree-l4 current"><a class="reference internal" href="pygenalgo.operators.html">pygenalgo.operators package</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="pygenalgo.html#module-pygenalgo">Module contents</a></li>
</ul>
</li>
</ul>
</li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">PyGenAlgo</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="modules.html">pygenalgo</a></li>
          <li class="breadcrumb-item"><a href="pygenalgo.html">pygenalgo package</a></li>
          <li class="breadcrumb-item"><a href="pygenalgo.operators.html">pygenalgo.operators package</a></li>
      <li class="breadcrumb-item active">pygenalgo.operators.crossover package</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/pygenalgo.operators.crossover.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="pygenalgo-operators-crossover-package">
<h1>pygenalgo.operators.crossover package<a class="headerlink" href="#pygenalgo-operators-crossover-package" title="Link to this heading"></a></h1>
<section id="submodules">
<h2>Submodules<a class="headerlink" href="#submodules" title="Link to this heading"></a></h2>
</section>
<section id="module-pygenalgo.operators.crossover.crossover_operator">
<span id="pygenalgo-operators-crossover-crossover-operator-module"></span><h2>pygenalgo.operators.crossover.crossover_operator module<a class="headerlink" href="#module-pygenalgo.operators.crossover.crossover_operator" title="Link to this heading"></a></h2>
<dl class="py class">
<dt class="sig sig-object py" id="pygenalgo.operators.crossover.crossover_operator.CrossoverOperator">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">pygenalgo.operators.crossover.crossover_operator.</span></span><span class="sig-name descname"><span class="pre">CrossoverOperator</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">crossover_probability</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pygenalgo/operators/crossover/crossover_operator.html#CrossoverOperator"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pygenalgo.operators.crossover.crossover_operator.CrossoverOperator" title="Link to this definition"></a></dt>
<dd><p>Bases: <a class="reference internal" href="pygenalgo.operators.html#pygenalgo.operators.genetic_operator.GeneticOperator" title="pygenalgo.operators.genetic_operator.GeneticOperator"><code class="xref py py-class docutils literal notranslate"><span class="pre">GeneticOperator</span></code></a></p>
<p>Description:</p>
<blockquote>
<div><p>Provides the base class (interface) for a Crossover Operator.</p>
</div></blockquote>
<dl class="py method">
<dt class="sig sig-object py" id="pygenalgo.operators.crossover.crossover_operator.CrossoverOperator.crossover">
<span class="sig-name descname"><span class="pre">crossover</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">parent1</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="pygenalgo.genome.html#pygenalgo.genome.chromosome.Chromosome" title="pygenalgo.genome.chromosome.Chromosome"><span class="pre">Chromosome</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">parent2</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="pygenalgo.genome.html#pygenalgo.genome.chromosome.Chromosome" title="pygenalgo.genome.chromosome.Chromosome"><span class="pre">Chromosome</span></a></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pygenalgo/operators/crossover/crossover_operator.html#CrossoverOperator.crossover"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pygenalgo.operators.crossover.crossover_operator.CrossoverOperator.crossover" title="Link to this definition"></a></dt>
<dd><p>Abstract method that “reminds” the user that if they want to
create a Crossover Class that inherits from here they should
implement a crossover method.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>parent1</strong> – (Chromosome).</p></li>
<li><p><strong>parent2</strong> – (Chromosome).</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Nothing but raising an error.</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

</section>
<section id="module-pygenalgo.operators.crossover.meta_crossover">
<span id="pygenalgo-operators-crossover-meta-crossover-module"></span><h2>pygenalgo.operators.crossover.meta_crossover module<a class="headerlink" href="#module-pygenalgo.operators.crossover.meta_crossover" title="Link to this heading"></a></h2>
<dl class="py class">
<dt class="sig sig-object py" id="pygenalgo.operators.crossover.meta_crossover.MetaCrossover">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">pygenalgo.operators.crossover.meta_crossover.</span></span><span class="sig-name descname"><span class="pre">MetaCrossover</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">crossover_probability</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">0.9</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pygenalgo/operators/crossover/meta_crossover.html#MetaCrossover"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pygenalgo.operators.crossover.meta_crossover.MetaCrossover" title="Link to this definition"></a></dt>
<dd><p>Bases: <a class="reference internal" href="#pygenalgo.operators.crossover.crossover_operator.CrossoverOperator" title="pygenalgo.operators.crossover.crossover_operator.CrossoverOperator"><code class="xref py py-class docutils literal notranslate"><span class="pre">CrossoverOperator</span></code></a></p>
<p>Description:</p>
<blockquote>
<div><p>Meta-crossover, crosses the chromosomes by applying randomly
all other crossovers (one at a time), with equal probability.</p>
</div></blockquote>
<dl class="py property">
<dt class="sig sig-object py" id="pygenalgo.operators.crossover.meta_crossover.MetaCrossover.all_counters">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">all_counters</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">dict</span></em><a class="headerlink" href="#pygenalgo.operators.crossover.meta_crossover.MetaCrossover.all_counters" title="Link to this definition"></a></dt>
<dd><p>Accessor (getter) of the application counter from all the internal crossovers.
This is mostly to verify that everything is working as expected.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>a dictionary with the counter calls for all crossover methods.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pygenalgo.operators.crossover.meta_crossover.MetaCrossover.crossover">
<span class="sig-name descname"><span class="pre">crossover</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">parent1</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="pygenalgo.genome.html#pygenalgo.genome.chromosome.Chromosome" title="pygenalgo.genome.chromosome.Chromosome"><span class="pre">Chromosome</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">parent2</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="pygenalgo.genome.html#pygenalgo.genome.chromosome.Chromosome" title="pygenalgo.genome.chromosome.Chromosome"><span class="pre">Chromosome</span></a></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pygenalgo/operators/crossover/meta_crossover.html#MetaCrossover.crossover"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pygenalgo.operators.crossover.meta_crossover.MetaCrossover.crossover" title="Link to this definition"></a></dt>
<dd><p>Perform the crossover operation on the two input parent
chromosomes, by selecting randomly a predefined method.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>parent1</strong> – (Chromosome).</p></li>
<li><p><strong>parent2</strong> – (Chromosome).</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>child1 and child2 (as Chromosomes).</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pygenalgo.operators.crossover.meta_crossover.MetaCrossover.reset_counter">
<span class="sig-name descname"><span class="pre">reset_counter</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="reference internal" href="_modules/pygenalgo/operators/crossover/meta_crossover.html#MetaCrossover.reset_counter"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pygenalgo.operators.crossover.meta_crossover.MetaCrossover.reset_counter" title="Link to this definition"></a></dt>
<dd><p>Sets ALL the counters to ‘zero’. We have to override the super().reset_counter()
method, because we have to call explicitly the reset_counter on all the internal
operators.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>None.</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

</section>
<section id="module-pygenalgo.operators.crossover.mutli_point_crossover">
<span id="pygenalgo-operators-crossover-mutli-point-crossover-module"></span><h2>pygenalgo.operators.crossover.mutli_point_crossover module<a class="headerlink" href="#module-pygenalgo.operators.crossover.mutli_point_crossover" title="Link to this heading"></a></h2>
<dl class="py class">
<dt class="sig sig-object py" id="pygenalgo.operators.crossover.mutli_point_crossover.MultiPointCrossover">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">pygenalgo.operators.crossover.mutli_point_crossover.</span></span><span class="sig-name descname"><span class="pre">MultiPointCrossover</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">crossover_probability</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">0.9</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pygenalgo/operators/crossover/mutli_point_crossover.html#MultiPointCrossover"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pygenalgo.operators.crossover.mutli_point_crossover.MultiPointCrossover" title="Link to this definition"></a></dt>
<dd><p>Bases: <a class="reference internal" href="#pygenalgo.operators.crossover.crossover_operator.CrossoverOperator" title="pygenalgo.operators.crossover.crossover_operator.CrossoverOperator"><code class="xref py py-class docutils literal notranslate"><span class="pre">CrossoverOperator</span></code></a></p>
<p>Description:</p>
<blockquote>
<div><p>Multipoint crossover creates two children chromosomes (offsprings),
by taking two parent chromosomes and cutting them at randomly chosen,
sites (loci).</p>
<p>It produces faster mixing, compared with single-point crossover.</p>
</div></blockquote>
<dl class="py method">
<dt class="sig sig-object py" id="pygenalgo.operators.crossover.mutli_point_crossover.MultiPointCrossover.crossover">
<span class="sig-name descname"><span class="pre">crossover</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">parent1</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="pygenalgo.genome.html#pygenalgo.genome.chromosome.Chromosome" title="pygenalgo.genome.chromosome.Chromosome"><span class="pre">Chromosome</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">parent2</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="pygenalgo.genome.html#pygenalgo.genome.chromosome.Chromosome" title="pygenalgo.genome.chromosome.Chromosome"><span class="pre">Chromosome</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_loci</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">2</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pygenalgo/operators/crossover/mutli_point_crossover.html#MultiPointCrossover.crossover"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pygenalgo.operators.crossover.mutli_point_crossover.MultiPointCrossover.crossover" title="Link to this definition"></a></dt>
<dd><p>Perform the crossover operation on the two input parent
chromosomes, using multiple cutting points (num_loci).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>parent1</strong> – (Chromosome).</p></li>
<li><p><strong>parent2</strong> – (Chromosome).</p></li>
<li><p><strong>num_loci</strong> – (int).</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>child1 and child2 (as Chromosomes).</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

</section>
<section id="module-pygenalgo.operators.crossover.order_crossover">
<span id="pygenalgo-operators-crossover-order-crossover-module"></span><h2>pygenalgo.operators.crossover.order_crossover module<a class="headerlink" href="#module-pygenalgo.operators.crossover.order_crossover" title="Link to this heading"></a></h2>
<dl class="py class">
<dt class="sig sig-object py" id="pygenalgo.operators.crossover.order_crossover.OrderCrossover">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">pygenalgo.operators.crossover.order_crossover.</span></span><span class="sig-name descname"><span class="pre">OrderCrossover</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">crossover_probability</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">0.9</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pygenalgo/operators/crossover/order_crossover.html#OrderCrossover"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pygenalgo.operators.crossover.order_crossover.OrderCrossover" title="Link to this definition"></a></dt>
<dd><p>Bases: <a class="reference internal" href="#pygenalgo.operators.crossover.crossover_operator.CrossoverOperator" title="pygenalgo.operators.crossover.crossover_operator.CrossoverOperator"><code class="xref py py-class docutils literal notranslate"><span class="pre">CrossoverOperator</span></code></a></p>
<p>Description:</p>
<blockquote>
<div><p>Order crossover (OX1) creates two children chromosomes, by ensuring that the original
genome (from both parents) isn’t repeated, thus creating invalid offsprings.</p>
<p>It is used predominantly in combinatorial problems.</p>
</div></blockquote>
<dl class="py method">
<dt class="sig sig-object py" id="pygenalgo.operators.crossover.order_crossover.OrderCrossover.crossover">
<span class="sig-name descname"><span class="pre">crossover</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">parent1</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="pygenalgo.genome.html#pygenalgo.genome.chromosome.Chromosome" title="pygenalgo.genome.chromosome.Chromosome"><span class="pre">Chromosome</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">parent2</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="pygenalgo.genome.html#pygenalgo.genome.chromosome.Chromosome" title="pygenalgo.genome.chromosome.Chromosome"><span class="pre">Chromosome</span></a></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pygenalgo/operators/crossover/order_crossover.html#OrderCrossover.crossover"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pygenalgo.operators.crossover.order_crossover.OrderCrossover.crossover" title="Link to this definition"></a></dt>
<dd><p>Perform the crossover operation on the two input parent chromosomes.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>parent1</strong> – (Chromosome).</p></li>
<li><p><strong>parent2</strong> – (Chromosome).</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>child1 and child2 (as Chromosomes).</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

</section>
<section id="module-pygenalgo.operators.crossover.partially_mapped_crossover">
<span id="pygenalgo-operators-crossover-partially-mapped-crossover-module"></span><h2>pygenalgo.operators.crossover.partially_mapped_crossover module<a class="headerlink" href="#module-pygenalgo.operators.crossover.partially_mapped_crossover" title="Link to this heading"></a></h2>
<dl class="py class">
<dt class="sig sig-object py" id="pygenalgo.operators.crossover.partially_mapped_crossover.PartiallyMappedCrossover">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">pygenalgo.operators.crossover.partially_mapped_crossover.</span></span><span class="sig-name descname"><span class="pre">PartiallyMappedCrossover</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">crossover_probability</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">0.9</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pygenalgo/operators/crossover/partially_mapped_crossover.html#PartiallyMappedCrossover"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pygenalgo.operators.crossover.partially_mapped_crossover.PartiallyMappedCrossover" title="Link to this definition"></a></dt>
<dd><p>Bases: <a class="reference internal" href="#pygenalgo.operators.crossover.crossover_operator.CrossoverOperator" title="pygenalgo.operators.crossover.crossover_operator.CrossoverOperator"><code class="xref py py-class docutils literal notranslate"><span class="pre">CrossoverOperator</span></code></a></p>
<p>Description:</p>
<blockquote>
<div><p>Partially Mapped Crossover (PMX) creates two children chromosomes, by ensuring that the
original genome (from both parents) isn’t repeated, thus creating invalid offsprings.</p>
<p>It is used predominantly in combinatorial problems.</p>
</div></blockquote>
<dl class="py method">
<dt class="sig sig-object py" id="pygenalgo.operators.crossover.partially_mapped_crossover.PartiallyMappedCrossover.crossover">
<span class="sig-name descname"><span class="pre">crossover</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">parent1</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="pygenalgo.genome.html#pygenalgo.genome.chromosome.Chromosome" title="pygenalgo.genome.chromosome.Chromosome"><span class="pre">Chromosome</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">parent2</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="pygenalgo.genome.html#pygenalgo.genome.chromosome.Chromosome" title="pygenalgo.genome.chromosome.Chromosome"><span class="pre">Chromosome</span></a></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pygenalgo/operators/crossover/partially_mapped_crossover.html#PartiallyMappedCrossover.crossover"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pygenalgo.operators.crossover.partially_mapped_crossover.PartiallyMappedCrossover.crossover" title="Link to this definition"></a></dt>
<dd><p>Perform the crossover operation on the two input parent chromosomes.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>parent1</strong> – (Chromosome).</p></li>
<li><p><strong>parent2</strong> – (Chromosome).</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>child1 and child2 (as Chromosomes).</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

</section>
<section id="module-pygenalgo.operators.crossover.position_based_crossover">
<span id="pygenalgo-operators-crossover-position-based-crossover-module"></span><h2>pygenalgo.operators.crossover.position_based_crossover module<a class="headerlink" href="#module-pygenalgo.operators.crossover.position_based_crossover" title="Link to this heading"></a></h2>
<dl class="py class">
<dt class="sig sig-object py" id="pygenalgo.operators.crossover.position_based_crossover.PositionBasedCrossover">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">pygenalgo.operators.crossover.position_based_crossover.</span></span><span class="sig-name descname"><span class="pre">PositionBasedCrossover</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">crossover_probability</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">0.9</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pygenalgo/operators/crossover/position_based_crossover.html#PositionBasedCrossover"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pygenalgo.operators.crossover.position_based_crossover.PositionBasedCrossover" title="Link to this definition"></a></dt>
<dd><p>Bases: <a class="reference internal" href="#pygenalgo.operators.crossover.crossover_operator.CrossoverOperator" title="pygenalgo.operators.crossover.crossover_operator.CrossoverOperator"><code class="xref py py-class docutils literal notranslate"><span class="pre">CrossoverOperator</span></code></a></p>
<p>Description:</p>
<blockquote>
<div><p>Position based crossover (POS) creates two children chromosomes, by ensuring that the
original genome (from both parents) isn’t repeated, thus creating invalid offsprings.</p>
<p>It is used predominantly in combinatorial problems.</p>
</div></blockquote>
<dl class="py method">
<dt class="sig sig-object py" id="pygenalgo.operators.crossover.position_based_crossover.PositionBasedCrossover.crossover">
<span class="sig-name descname"><span class="pre">crossover</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">parent1</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="pygenalgo.genome.html#pygenalgo.genome.chromosome.Chromosome" title="pygenalgo.genome.chromosome.Chromosome"><span class="pre">Chromosome</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">parent2</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="pygenalgo.genome.html#pygenalgo.genome.chromosome.Chromosome" title="pygenalgo.genome.chromosome.Chromosome"><span class="pre">Chromosome</span></a></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pygenalgo/operators/crossover/position_based_crossover.html#PositionBasedCrossover.crossover"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pygenalgo.operators.crossover.position_based_crossover.PositionBasedCrossover.crossover" title="Link to this definition"></a></dt>
<dd><p>Perform the crossover operation on the two input parent chromosomes.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>parent1</strong> – (Chromosome).</p></li>
<li><p><strong>parent2</strong> – (Chromosome).</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>child1 and child2 (as Chromosomes).</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

</section>
<section id="module-pygenalgo.operators.crossover.single_point_crossover">
<span id="pygenalgo-operators-crossover-single-point-crossover-module"></span><h2>pygenalgo.operators.crossover.single_point_crossover module<a class="headerlink" href="#module-pygenalgo.operators.crossover.single_point_crossover" title="Link to this heading"></a></h2>
<dl class="py class">
<dt class="sig sig-object py" id="pygenalgo.operators.crossover.single_point_crossover.SinglePointCrossover">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">pygenalgo.operators.crossover.single_point_crossover.</span></span><span class="sig-name descname"><span class="pre">SinglePointCrossover</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">crossover_probability</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">0.9</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pygenalgo/operators/crossover/single_point_crossover.html#SinglePointCrossover"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pygenalgo.operators.crossover.single_point_crossover.SinglePointCrossover" title="Link to this definition"></a></dt>
<dd><p>Bases: <a class="reference internal" href="#pygenalgo.operators.crossover.crossover_operator.CrossoverOperator" title="pygenalgo.operators.crossover.crossover_operator.CrossoverOperator"><code class="xref py py-class docutils literal notranslate"><span class="pre">CrossoverOperator</span></code></a></p>
<p>Description:</p>
<blockquote>
<div><p>Single-point crossover creates two children chromosomes (offsprings),
by taking two parent chromosomes and cutting them at some, randomly
chosen, site (locus).</p>
<p>It produces very slow mixing, compared with multipoint or uniform crossover.</p>
</div></blockquote>
<dl class="py method">
<dt class="sig sig-object py" id="pygenalgo.operators.crossover.single_point_crossover.SinglePointCrossover.crossover">
<span class="sig-name descname"><span class="pre">crossover</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">parent1</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="pygenalgo.genome.html#pygenalgo.genome.chromosome.Chromosome" title="pygenalgo.genome.chromosome.Chromosome"><span class="pre">Chromosome</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">parent2</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="pygenalgo.genome.html#pygenalgo.genome.chromosome.Chromosome" title="pygenalgo.genome.chromosome.Chromosome"><span class="pre">Chromosome</span></a></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pygenalgo/operators/crossover/single_point_crossover.html#SinglePointCrossover.crossover"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pygenalgo.operators.crossover.single_point_crossover.SinglePointCrossover.crossover" title="Link to this definition"></a></dt>
<dd><p>Perform the crossover operation on the two input parent chromosomes.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>parent1</strong> – (Chromosome).</p></li>
<li><p><strong>parent2</strong> – (Chromosome).</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>child1 and child2 (as Chromosomes).</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

</section>
<section id="module-pygenalgo.operators.crossover.uniform_crossover">
<span id="pygenalgo-operators-crossover-uniform-crossover-module"></span><h2>pygenalgo.operators.crossover.uniform_crossover module<a class="headerlink" href="#module-pygenalgo.operators.crossover.uniform_crossover" title="Link to this heading"></a></h2>
<dl class="py class">
<dt class="sig sig-object py" id="pygenalgo.operators.crossover.uniform_crossover.UniformCrossover">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">pygenalgo.operators.crossover.uniform_crossover.</span></span><span class="sig-name descname"><span class="pre">UniformCrossover</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">crossover_probability</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">0.9</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pygenalgo/operators/crossover/uniform_crossover.html#UniformCrossover"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pygenalgo.operators.crossover.uniform_crossover.UniformCrossover" title="Link to this definition"></a></dt>
<dd><p>Bases: <a class="reference internal" href="#pygenalgo.operators.crossover.crossover_operator.CrossoverOperator" title="pygenalgo.operators.crossover.crossover_operator.CrossoverOperator"><code class="xref py py-class docutils literal notranslate"><span class="pre">CrossoverOperator</span></code></a></p>
<p>Description:</p>
<blockquote>
<div><p>Uniform crossover creates two children chromosomes (offsprings),
by taking two parent chromosomes and swap their genes in every
other location.</p>
<p>It produces fast mixing, compared with single-point crossover.</p>
</div></blockquote>
<dl class="py method">
<dt class="sig sig-object py" id="pygenalgo.operators.crossover.uniform_crossover.UniformCrossover.crossover">
<span class="sig-name descname"><span class="pre">crossover</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">parent1</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="pygenalgo.genome.html#pygenalgo.genome.chromosome.Chromosome" title="pygenalgo.genome.chromosome.Chromosome"><span class="pre">Chromosome</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">parent2</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="pygenalgo.genome.html#pygenalgo.genome.chromosome.Chromosome" title="pygenalgo.genome.chromosome.Chromosome"><span class="pre">Chromosome</span></a></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pygenalgo/operators/crossover/uniform_crossover.html#UniformCrossover.crossover"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pygenalgo.operators.crossover.uniform_crossover.UniformCrossover.crossover" title="Link to this definition"></a></dt>
<dd><p>Perform the crossover operation on the two input parent chromosomes.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>parent1</strong> – (Chromosome).</p></li>
<li><p><strong>parent2</strong> – (Chromosome).</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>child1 and child2 (as Chromosomes).</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

</section>
<section id="module-pygenalgo.operators.crossover">
<span id="module-contents"></span><h2>Module contents<a class="headerlink" href="#module-pygenalgo.operators.crossover" title="Link to this heading"></a></h2>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="pygenalgo.operators.html" class="btn btn-neutral float-left" title="pygenalgo.operators package" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="pygenalgo.operators.migration.html" class="btn btn-neutral float-right" title="pygenalgo.operators.migration package" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2024, Michalis Vrettas, PhD.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>