<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="./">
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.18.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>src.engines package &mdash; PyGeneticAlgorithms 0.1.0 documentation</title>
      <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=92fd9be5" />
      <link rel="stylesheet" type="text/css" href="_static/css/theme.css?v=19f00094" />

  
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script src="_static/jquery.js?v=5d32c60e"></script>
        <script src="_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
        <script src="_static/documentation_options.js?v=01f34227"></script>
        <script src="_static/doctools.js?v=9a2dae69"></script>
        <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="src.genome package" href="src.genome.html" />
    <link rel="prev" title="src package" href="src.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="index.html" class="icon icon-home">
            PyGeneticAlgorithms
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="reference internal" href="modules.html">code</a><ul class="current">
<li class="toctree-l2 current"><a class="reference internal" href="src.html">src package</a><ul class="current">
<li class="toctree-l3 current"><a class="reference internal" href="src.html#subpackages">Subpackages</a><ul class="current">
<li class="toctree-l4 current"><a class="current reference internal" href="#">src.engines package</a></li>
<li class="toctree-l4"><a class="reference internal" href="src.genome.html">src.genome package</a></li>
<li class="toctree-l4"><a class="reference internal" href="src.operators.html">src.operators package</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="src.html#module-src">Module contents</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="tests.html">tests package</a></li>
</ul>
</li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">PyGeneticAlgorithms</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="modules.html">code</a></li>
          <li class="breadcrumb-item"><a href="src.html">src package</a></li>
      <li class="breadcrumb-item active">src.engines package</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/src.engines.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="src-engines-package">
<h1>src.engines package<a class="headerlink" href="#src-engines-package" title="Link to this heading"></a></h1>
<section id="submodules">
<h2>Submodules<a class="headerlink" href="#submodules" title="Link to this heading"></a></h2>
</section>
<section id="module-src.engines.auxiliary">
<span id="src-engines-auxiliary-module"></span><h2>src.engines.auxiliary module<a class="headerlink" href="#module-src.engines.auxiliary" title="Link to this heading"></a></h2>
<dl class="py class">
<dt class="sig sig-object py" id="src.engines.auxiliary.SubPopulation">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">src.engines.auxiliary.</span></span><span class="sig-name descname"><span class="pre">SubPopulation</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">pop_id:</span> <span class="pre">int</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">population:</span> <span class="pre">list</span> <span class="pre">=</span> <span class="pre">&lt;factory&gt;</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/src/engines/auxiliary.html#SubPopulation"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#src.engines.auxiliary.SubPopulation" title="Link to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Auxiliary class container used in the IslandModelGA
to hold all the subpopulations (one on each island).</p>
<dl class="py property">
<dt class="sig sig-object py" id="src.engines.auxiliary.SubPopulation.id">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">id</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">int</span></em><a class="headerlink" href="#src.engines.auxiliary.SubPopulation.id" title="Link to this definition"></a></dt>
<dd><p>Accessor (getter) of the id parameter.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>the id value.</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="src.engines.auxiliary.SubPopulation.pop_id">
<span class="sig-name descname"><span class="pre">pop_id</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">int</span></em><a class="headerlink" href="#src.engines.auxiliary.SubPopulation.pop_id" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="src.engines.auxiliary.SubPopulation.population">
<span class="sig-name descname"><span class="pre">population</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">list</span></em><a class="headerlink" href="#src.engines.auxiliary.SubPopulation.population" title="Link to this definition"></a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="src.engines.auxiliary.apply_corrections">
<span class="sig-prename descclassname"><span class="pre">src.engines.auxiliary.</span></span><span class="sig-name descname"><span class="pre">apply_corrections</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">input_population</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span><span class="p"><span class="pre">[</span></span><a class="reference internal" href="src.genome.html#src.genome.chromosome.Chromosome" title="src.genome.chromosome.Chromosome"><span class="pre">Chromosome</span></a><span class="p"><span class="pre">]</span></span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">int</span></span></span><a class="reference internal" href="_modules/src/engines/auxiliary.html#apply_corrections"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#src.engines.auxiliary.apply_corrections" title="Link to this definition"></a></dt>
<dd><p>Check the population for invalid genes and correct them by applying
directly the random method. It is assumed that the random method of
the Gene is always returning a ‘valid’ value for the Gene.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>input_population</strong> – List(Chromosome) the population we want to</p>
</dd>
</dl>
<p>apply corrections (if applicable).</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>the total number of corrected genes in the population.</p>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-src.engines.island_model_ga">
<span id="src-engines-island-model-ga-module"></span><h2>src.engines.island_model_ga module<a class="headerlink" href="#module-src.engines.island_model_ga" title="Link to this heading"></a></h2>
<dl class="py class">
<dt class="sig sig-object py" id="src.engines.island_model_ga.IslandModelGA">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">src.engines.island_model_ga.</span></span><span class="sig-name descname"><span class="pre">IslandModelGA</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">initial_pop</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span><span class="p"><span class="pre">[</span></span><a class="reference internal" href="src.genome.html#src.genome.chromosome.Chromosome" title="src.genome.chromosome.Chromosome"><span class="pre">Chromosome</span></a><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fit_func</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Callable</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_islands</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">select_op</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="src.operators.selection.html#src.operators.selection.select_operator.SelectionOperator" title="src.operators.selection.select_operator.SelectionOperator"><span class="pre">SelectionOperator</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mutate_op</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="src.operators.mutation.html#src.operators.mutation.mutate_operator.MutationOperator" title="src.operators.mutation.mutate_operator.MutationOperator"><span class="pre">MutationOperator</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cross_op</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="src.operators.crossover.html#src.operators.crossover.crossover_operator.CrossoverOperator" title="src.operators.crossover.crossover_operator.CrossoverOperator"><span class="pre">CrossoverOperator</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">migrate_op</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="src.operators.migration.html#src.operators.migration.clockwise_migration.ClockwiseMigration" title="src.operators.migration.clockwise_migration.ClockwiseMigration"><span class="pre">ClockwiseMigration</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/src/engines/island_model_ga.html#IslandModelGA"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#src.engines.island_model_ga.IslandModelGA" title="Link to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>In Island Model GA we run in parallel a number of “islands”, each one evolving its own
(sub)-population. Optionally we can allow “migration”, among the best individuals from
each island.</p>
<dl class="py attribute">
<dt class="sig sig-object py" id="src.engines.island_model_ga.IslandModelGA.MAX_CPUs">
<span class="sig-name descname"><span class="pre">MAX_CPUs</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">4</span></em><a class="headerlink" href="#src.engines.island_model_ga.IslandModelGA.MAX_CPUs" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="src.engines.island_model_ga.IslandModelGA.best_chromosome">
<span class="sig-name descname"><span class="pre">best_chromosome</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/src/engines/island_model_ga.html#IslandModelGA.best_chromosome"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#src.engines.island_model_ga.IslandModelGA.best_chromosome" title="Link to this definition"></a></dt>
<dd><p>Auxiliary method.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Return the chromosome with the highest fitness.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="src.engines.island_model_ga.IslandModelGA.evaluate_fitness">
<span class="sig-name descname"><span class="pre">evaluate_fitness</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">in_population</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span><span class="p"><span class="pre">[</span></span><a class="reference internal" href="src.genome.html#src.genome.chromosome.Chromosome" title="src.genome.chromosome.Chromosome"><span class="pre">Chromosome</span></a><span class="p"><span class="pre">]</span></span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/src/engines/island_model_ga.html#IslandModelGA.evaluate_fitness"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#src.engines.island_model_ga.IslandModelGA.evaluate_fitness" title="Link to this definition"></a></dt>
<dd><p>Evaluate all the chromosomes of the input population list with the
custom fitness  function. After updating all  the chromosomes with
their fitness, the method returns the average statistics mean/std.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>in_population</strong> – (list) The population of Chromosomes that we</p>
</dd>
</dl>
<p>want to evaluate their fitness.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>mean(fitness), std(fitness).</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="src.engines.island_model_ga.IslandModelGA.evolve_population">
<em class="property"><span class="pre">static</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">evolve_population</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">island</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="#src.engines.auxiliary.SubPopulation" title="src.engines.auxiliary.SubPopulation"><span class="pre">SubPopulation</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">eval_fitness</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Callable</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">epochs</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">crs_op</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="src.operators.crossover.html#src.operators.crossover.crossover_operator.CrossoverOperator" title="src.operators.crossover.crossover_operator.CrossoverOperator"><span class="pre">CrossoverOperator</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">mut_op</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="src.operators.mutation.html#src.operators.mutation.mutate_operator.MutationOperator" title="src.operators.mutation.mutate_operator.MutationOperator"><span class="pre">MutationOperator</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">sel_op</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="src.operators.selection.html#src.operators.selection.select_operator.SelectionOperator" title="src.operators.selection.select_operator.SelectionOperator"><span class="pre">SelectionOperator</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">rnd_gen</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">f_tol</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">1e-06</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">correction</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">elitism</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/src/engines/island_model_ga.html#IslandModelGA.evolve_population"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#src.engines.island_model_ga.IslandModelGA.evolve_population" title="Link to this definition"></a></dt>
<dd><p>This method is called to evolve each subpopulation independently. It is declared ‘static’ to avoid problems
when passed in the Parallel pool. The input parameters have identical meaning with the ones from the run().</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="src.engines.island_model_ga.IslandModelGA.fitness_func">
<span class="sig-name descname"><span class="pre">fitness_func</span></span><a class="headerlink" href="#src.engines.island_model_ga.IslandModelGA.fitness_func" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="src.engines.island_model_ga.IslandModelGA.num_islands">
<span class="sig-name descname"><span class="pre">num_islands</span></span><a class="headerlink" href="#src.engines.island_model_ga.IslandModelGA.num_islands" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="src.engines.island_model_ga.IslandModelGA.population">
<span class="sig-name descname"><span class="pre">population</span></span><a class="headerlink" href="#src.engines.island_model_ga.IslandModelGA.population" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="src.engines.island_model_ga.IslandModelGA.rng_GA">
<span class="sig-name descname"><span class="pre">rng_GA</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">Generator(PCG64)</span> <span class="pre">at</span> <span class="pre">0x7FD1984B7300</span></em><a class="headerlink" href="#src.engines.island_model_ga.IslandModelGA.rng_GA" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="src.engines.island_model_ga.IslandModelGA.run">
<span class="sig-name descname"><span class="pre">run</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">epochs</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">1000</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">correction</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">elitism</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">f_tol</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">1e-06</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">allow_migration</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_periods</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">10</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbose</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/src/engines/island_model_ga.html#IslandModelGA.run"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#src.engines.island_model_ga.IslandModelGA.run" title="Link to this definition"></a></dt>
<dd><p>Main method of the StandardGA class, that implements the evolutionary routine.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>epochs</strong> – (int) maximum number of iterations in the evolution process.</p></li>
<li><p><strong>correction</strong> – (bool) flag that if set to ‘True’ will check the validity of</p></li>
</ul>
</dd>
</dl>
<p>the population (at the gene level) and attempt to correct the genome by calling
the random() method of the flawed gene.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>elitism</strong> – (bool) flag that defines elitism. If ‘True’ then the chromosome</p>
</dd>
</dl>
<p>with the higher fitness will always be copied to the next generation (unaltered).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>f_tol</strong> – (float) tolerance in the difference between the average values of two</p>
</dd>
</dl>
<p>consecutive populations. It is used to determine the convergence of the population.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>allow_migration</strong> – (bool) flag that if set to ‘True’ will allow the migration</p>
</dd>
</dl>
<p>of the best individuals among the different islands.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>n_periods</strong> – (int) the number of times that we will break the main evolution</p>
</dd>
</dl>
<p>to allow for chromosomes to migrate. NB: This setting is active only when the option
allow_migration == True. Otherwise, is ignored.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>verbose</strong> – (bool) if ‘True’ it will display periodically information about the</p>
</dd>
</dl>
<p>current stats of the subpopulations. NB: This setting is active only when the option
allow_migration == True. Otherwise, is ignored.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>None.</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="src.engines.island_model_ga.IslandModelGA.stats">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">stats</span></span><a class="headerlink" href="#src.engines.island_model_ga.IslandModelGA.stats" title="Link to this definition"></a></dt>
<dd><p>Accessor method that returns the ‘stats’ dictionary.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>the dictionary with the statistics from the run.</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

</section>
<section id="module-src.engines.standard_ga">
<span id="src-engines-standard-ga-module"></span><h2>src.engines.standard_ga module<a class="headerlink" href="#module-src.engines.standard_ga" title="Link to this heading"></a></h2>
<dl class="py class">
<dt class="sig sig-object py" id="src.engines.standard_ga.StandardGA">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">src.engines.standard_ga.</span></span><span class="sig-name descname"><span class="pre">StandardGA</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">initial_pop</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span><span class="p"><span class="pre">[</span></span><a class="reference internal" href="src.genome.html#src.genome.chromosome.Chromosome" title="src.genome.chromosome.Chromosome"><span class="pre">Chromosome</span></a><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fit_func</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Callable</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">select_op</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="src.operators.selection.html#src.operators.selection.select_operator.SelectionOperator" title="src.operators.selection.select_operator.SelectionOperator"><span class="pre">SelectionOperator</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mutate_op</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="src.operators.mutation.html#src.operators.mutation.mutate_operator.MutationOperator" title="src.operators.mutation.mutate_operator.MutationOperator"><span class="pre">MutationOperator</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cross_op</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="src.operators.crossover.html#src.operators.crossover.crossover_operator.CrossoverOperator" title="src.operators.crossover.crossover_operator.CrossoverOperator"><span class="pre">CrossoverOperator</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/src/engines/standard_ga.html#StandardGA"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#src.engines.standard_ga.StandardGA" title="Link to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Standard GA model that at each iteration replaces the whole population using
the genetic operators (crossover and mutation).</p>
<dl class="py attribute">
<dt class="sig sig-object py" id="src.engines.standard_ga.StandardGA.MAX_CPUs">
<span class="sig-name descname"><span class="pre">MAX_CPUs</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">4</span></em><a class="headerlink" href="#src.engines.standard_ga.StandardGA.MAX_CPUs" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="src.engines.standard_ga.StandardGA.best_chromosome">
<span class="sig-name descname"><span class="pre">best_chromosome</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/src/engines/standard_ga.html#StandardGA.best_chromosome"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#src.engines.standard_ga.StandardGA.best_chromosome" title="Link to this definition"></a></dt>
<dd><p>Auxiliary method.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Return the chromosome with the highest fitness.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="src.engines.standard_ga.StandardGA.evaluate_fitness">
<span class="sig-name descname"><span class="pre">evaluate_fitness</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">input_population</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span><span class="p"><span class="pre">[</span></span><a class="reference internal" href="src.genome.html#src.genome.chromosome.Chromosome" title="src.genome.chromosome.Chromosome"><span class="pre">Chromosome</span></a><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="n"><span class="pre">parallel</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/src/engines/standard_ga.html#StandardGA.evaluate_fitness"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#src.engines.standard_ga.StandardGA.evaluate_fitness" title="Link to this definition"></a></dt>
<dd><p>Evaluate all the chromosomes of the input list with the custom fitness function.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>input_population</strong> – (list) The population of Chromosomes that we want to evaluate</p>
</dd>
</dl>
<p>their fitness.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>parallel</strong> – (bool) Flag that enables parallel computation of the fitness function.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>the mean and std of the fitness values.</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="src.engines.standard_ga.StandardGA.fitness_func">
<span class="sig-name descname"><span class="pre">fitness_func</span></span><a class="headerlink" href="#src.engines.standard_ga.StandardGA.fitness_func" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="src.engines.standard_ga.StandardGA.individual_fitness">
<span class="sig-name descname"><span class="pre">individual_fitness</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">index</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">float</span></span></span><a class="reference internal" href="_modules/src/engines/standard_ga.html#StandardGA.individual_fitness"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#src.engines.standard_ga.StandardGA.individual_fitness" title="Link to this definition"></a></dt>
<dd><p>Get the fitness value of an individual member of the population.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>index</strong> – Position of the individual in the population.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The fitness value (float).</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="src.engines.standard_ga.StandardGA.population">
<span class="sig-name descname"><span class="pre">population</span></span><a class="headerlink" href="#src.engines.standard_ga.StandardGA.population" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="src.engines.standard_ga.StandardGA.population_fitness">
<span class="sig-name descname"><span class="pre">population_fitness</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">float</span><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="_modules/src/engines/standard_ga.html#StandardGA.population_fitness"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#src.engines.standard_ga.StandardGA.population_fitness" title="Link to this definition"></a></dt>
<dd><p>Get the fitness of all the population.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>A list with all the fitness values.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="src.engines.standard_ga.StandardGA.print_operator_stats">
<span class="sig-name descname"><span class="pre">print_operator_stats</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/src/engines/standard_ga.html#StandardGA.print_operator_stats"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#src.engines.standard_ga.StandardGA.print_operator_stats" title="Link to this definition"></a></dt>
<dd><p>Print the genetic operators stats.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>None.</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="src.engines.standard_ga.StandardGA.rng_GA">
<span class="sig-name descname"><span class="pre">rng_GA</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">Generator(PCG64)</span> <span class="pre">at</span> <span class="pre">0x7FD196AFD620</span></em><a class="headerlink" href="#src.engines.standard_ga.StandardGA.rng_GA" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="src.engines.standard_ga.StandardGA.run">
<span class="sig-name descname"><span class="pre">run</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">epochs</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">100</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">elitism</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">correction</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">f_tol</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">1e-08</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">parallel</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbose</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/src/engines/standard_ga.html#StandardGA.run"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#src.engines.standard_ga.StandardGA.run" title="Link to this definition"></a></dt>
<dd><p>Main method of the StandardGA class, that implements the evolutionary routine.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>epochs</strong> – (int) maximum number of iterations in the evolution process.</p></li>
<li><p><strong>elitism</strong> – (bool) flag that defines elitism. If ‘True’ then the chromosome</p></li>
</ul>
</dd>
</dl>
<p>with the higher fitness will always be copied to the next generation (unaltered).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>correction</strong> – (bool) flag that if set to ‘True’ will check the validity of</p>
</dd>
</dl>
<p>the population (at the gene level) and attempt to correct the genome by calling
the random() method of the flawed gene.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>f_tol</strong> – (float) tolerance in the difference between the average values of</p>
</dd>
</dl>
<p>two consecutive populations. It is used to determine the convergence of the population.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>parallel</strong> – (bool) Flag that enables parallel computation of the fitness function.</p></li>
<li><p><strong>verbose</strong> – (bool) if ‘True’ it will display periodically information about</p></li>
</ul>
</dd>
</dl>
<p>the current average fitness and spread of the population.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>None.</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="src.engines.standard_ga.StandardGA.stats">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">stats</span></span><a class="headerlink" href="#src.engines.standard_ga.StandardGA.stats" title="Link to this definition"></a></dt>
<dd><p>Accessor method that returns the ‘stats’ dictionary.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>the dictionary with the statistics from the run.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="src.engines.standard_ga.StandardGA.update_stats">
<span class="sig-name descname"><span class="pre">update_stats</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">avg_fitness</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">std_fitness</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/src/engines/standard_ga.html#StandardGA.update_stats"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#src.engines.standard_ga.StandardGA.update_stats" title="Link to this definition"></a></dt>
<dd><p>Update the stats dictionary with the input mean/std values
of the fitness.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>avg_fitness</strong> – (float) mean fitness value of the population.</p></li>
<li><p><strong>std_fitness</strong> – (float) std fitness value of the population.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>None.</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

</section>
<section id="module-src.engines">
<span id="module-contents"></span><h2>Module contents<a class="headerlink" href="#module-src.engines" title="Link to this heading"></a></h2>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="src.html" class="btn btn-neutral float-left" title="src package" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="src.genome.html" class="btn btn-neutral float-right" title="src.genome package" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2024, Michalis Vrettas, PhD.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>